{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove redundancy by clustering the CytB sequences with `vsearch`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vsearch v1.1.0_linux_x86_64, 31.4GB RAM, 12 cores\n",
      "https://github.com/torognes/vsearch\n",
      "\n",
      "Reading file ../1-initial/CytB_cropped.fa 100%\n",
      "3436585 nt in 3745 seqs, min 501, max 1224, avg 918\n",
      "Indexing sequences 100%\n",
      "Masking 100%\n",
      "Sorting by length 100%\n",
      "Counting unique k-mers 100%\n",
      "Clustering 100%\n",
      "Writing clusters 100%\n",
      "Clusters: 1762 Size min 1, max 197, avg 2.1\n",
      "Singletons: 1344, 35.9% of seqs, 76.3% of clusters\n"
     ]
    }
   ],
   "source": [
    "!vsearch -threads 5 --cluster_fast ../1-initial/CytB_cropped.fa --strand both --uc CytB_id-1.uc --id 1 --query_cov 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj = Project('../1-initial/target_locus.csv', pickle='nr.pkl', git=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#pj = unpickle_pj('nr.pkl', git=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reprophylo/reprophylo.py:1015: UserWarning: Version control off\n",
      "  warnings.warn('Version control off')\n"
     ]
    }
   ],
   "source": [
    "pj.read_embl_genbank(['../1-initial/CytB_cropped.gb'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distribution Of Sequence Lengths\n",
      "Distribution Of Sequence Statistic \"Gc_Content\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/pymodules/python2.7/matplotlib/figure.py:371: UserWarning: matplotlib is currently using a non-GUI backend, so cannot show the figure\n",
      "  \"matplotlib is currently using a non-GUI backend, \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAFEAAAInCAYAAADga7XDAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFH9JREFUeJztnXu4FOV9xz8HFAUEDCJCq/SgAcH7pUGttVI1ijQJmjTe\nWlsNT0pDxGjbGDV5mtMYtdVcvDRGWytiG6lWctFHpIqNmPqYoOjRo3AMqEcQBRUDEUVAzukfv3e7\nc5bZPTt7frs7s/v9PM8+OzszO/uez5nLO/O+8x0QQgghhBCO3AGsBzoi464CngPagUeB/SLTrgBW\nAp3AqZHxR4dlrARurGJ5U8kJwJH0ljgsMjwHuD0MH4SJ3RVoBVYBLWHaUmBKGF4ITKtOcYszoNY/\nGOEXwG8Kxr0XGd4DeCcMzwDmA9uBLkziMcBYTPzSMN9dwBnVKW5xdqn1D5bB1cD5wBbya9jvAL+M\nzPM68LuY1Ncj49eG8TWlnmtiMb4OjAPmAjfUuSxlkUaJOe4GPhGG19L7ILMvtgauDcPR8WvjFjZ0\n6NBVQE8/X+39/7P8aaX3gWVCZHgO8O9hOHdgGQSMB14mf2D5FbZ/bKH0gaWnj7K0lVHevpZRc+YD\nbwDbgDXAF4D7MKntwAJgdGT+K7EDSidwWmR8roqzCripxO81pMRaUzWJad4n1prH6l2ALOCxKWpN\nrBaS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKID\nkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDqTxfuea\n09bW1hMZbik1bxxaEx2QxAiVrIUgiS5IogOS6IAkOiCJDkiiA5LogCQ6UE+Jcal11wMrsOS6HwMj\nItNSm1pXT4lz2TkI6GHgYOBw4NeYOLBwobPD+zTgFvLhQj8EZmLBRBNilll10pZa9wjQHYZ/RT7C\nKtWpdWneJ34Bi60CS62LptPlUusKxyu1LsLXsfiru+tdkHJI4/XEC4DpwMmRcf1OrQOYMmXKgsjw\n8lmzZq0AWLDARnd0dJwXnf+2226bvHTp0oNyn5cuXUoaaaX30Xka8CIwqmC+qqbWtbW19UQvzJYg\ndp56ronzgRMxYWuAb2JH40HYAQbgSWA2sBy4N7x/FMbl/qDZwJ3AYEziopqUPkI9JZ4bM+6OEvNf\nE16FLAMOdSlRhaT1wJIpJNEBSXRAEh2QRAck0QFJdEASHZBEByTRAUl0QBIdkEQHJNEBSXRAEh2Q\nRAck0QFJdEASHZBEByTRAUl0QBIdkEQHJNEBSXRAEh2QRAck0QFJdEASHZDECMOsq7MoQdE+2SNh\nfQ+8NNLu8Eq8DK2JDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkO\nSKIDkuhA2lLrPo81Fu0AjiqYP7WpdfXkBOBIekucBEwEfk5viblwoV2xQKJV5MOFlgJTwnBF4UJZ\nbqiKS63rxCL/ClFqnQNKrWt00phaF0dVU+s+efTRu+8YNGjYoG3bdu9Ytuz/k+uymlqX4+fYUTdH\nVVPr+ntgqSfzgTewsMk1WHLnGWF4C7AOeCgy/5XYAaUTOC0yPlfFWQXcVOL3GlJirWnIKk7DIIkO\nSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkii\nA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS\n6IAkOiCJDkiiA5LogCQ60NL3LA1DDzF/b09PT6+76VtaWko5iV2G1kQHshLpUjVaWlpaRsL6DbBx\nL9izkmVUIvFjwGhs1X6bnaOqmo5yNueBwGfJR7BsAFZg0SobsDCfu4Ezy1xejrjUupHAI1he2MP0\nXjMymVq3CzAHk9QNvIeF/twO/BNwXRh+DNgc5lkLXISJ74u41LrrgMvC8NeAfwzDqU6tK8VLwPvA\nvwFTKS1mIHASMDd8p7PM32ilt8ROYJ8wPCaynCswqTkWAcdiqXUrIuPPAW4t8lt1ycVZCOwPzMTW\nth0l5t0B/A9wIXAA9kdWwj7k/5D15IWmOrWu1IHl0gqXuQ64pMLvRukhI7FSaavirMc243XYpvpW\nGN9wqXXHYDv8e8Lr2jCuElrZ+cCS2/ddzs4Hlsyn1g3EqiXdRV53Ut5ROUdhat2FWBVnMfFVnIZI\nrfsmJuvH2JFxRHgdB/wkTGvr749UkVRIfA1bO+JowSrJr/X3R6pIKqL/RgM/K7Hwn5GvkjQVSSSu\nxI6cxRiDVdBFCc7FLjYcETPtyDDtnJqWKBlV25yT1BMnAq8AT2H7v9zp1kHAKcDzYZ6/L/jetxL8\nRsNTrGrT1ystpGJN3D/BvE1FEold1SpE1qn03HkI8Hth+DXgA5/iZJOkDVUHYwHim7DnprwIbAzj\nDvEtWnZIsiYeCSwBhrLz0flU4HjgROBZzwI2Gouxta7wiT2EcZvCPGklFefOm4Fvl5h+NdY0kFZS\nce78IfBmielvYg9eaDqSSFwIfKbE9E+HeZqOJBL/BtgLuA9rohwWXscAC7ALqpW2yzQs3VgrXnfM\ncLFXqRbBelOX0767yilZOT/S6JSSeEGtCpF11LXOgVISP9aP5fbnu5mjlMQurIVvrwTLGw1cBbza\njzI1FF/CeiBsxZpEvwgcDuwRmWcYdsr3ZeBB7DGZ64BZNS1pedTttG8E1iOri95Vma3hFR23Cvgq\nJjaNVE1iuR3fB2AV7BOxqzZ7k+8p+wLWb3FZmcuqF7Gd1sEk5robv1u62Td2GeVeCusGfhleogBV\ncRyQRAck0QFJdEASHZBEByTRgaSN9wOwzksfx86p4yqv6sBUggnAcrLTgamQVHRouhnr1HQZdpq3\nIcF3ReB97J6+rJKKduetWCdPUUASif+N9bcRBSRtdz4O+DvsziYRKHU98VV23gfsAYzC2pffoHc7\nc0uYP609autyPbGSG3vU7lzA1FoVIusk2Sf+EdYsUIy9wzxNRxKJjwGfLDH9ZKwS3nR4XoAYSJPu\nEz0lHge847i8zNDXufNXsDyH3Bp2A/FdjkcCw7GbypuOviRuIl/VacXWtLcK5unBbsV4Evi+Z+Ea\nkS5gRr0L0Q9ScSmsNcG8TYWaBxxIIjHXJ7vUazN29ftGLD2pKUgi8S4sOqUFiym4P7xeCuM6sHir\nHVhQWzvpvRhRN07Cbks7PWbadOxIfnL4/CdY3s2PKvytr2D/lBfCMFQWCxglFbelPQlcX2L69fTu\nNXYrpe/AKsYhmMDdsbOgR7AQtySxgHFbWCqaBw6ndPPAq8Bhkc/tJOuqnGMSFl31IbZrWAJ8Drub\na16YZx5wRhiegaU9bceqYavI5ynWhCQSN2FtzsU4Gfht5PPw8J2kvIAFVI7Ebk6fjgWpJY0FrBlJ\n6onzsVPA24DvYf/xHiyB5FIsDjUaR/rH2JE6KZ1Yq+LDWAtjOzvfqdVXLGDstDSk1g3BOrfnGum3\nh1fu80NhHrD92bXEH4SScjXWCb+TfLjRWPIpn5eHV45FxKfopeLAkmM68EOssIuwA8h0r4UHRof3\ncdgd/iOoLBYwSqok1oLHsYsa7dhuASqLBYzSdBKrQSouQIDFuMyidK+wkxIuM/MkkXg68FOsUrsZ\neDdmnqZf3fuiHQu9/f16F6RCUnHGMglrHng6wXeagiQS38F6hokCkl4K+1y1CpJlkhxY7sTqbPdj\np3evEB+csbr/xcoWSSRGH8rwqSLz9JAsa7shSCKxnLsCVMVpcFJRxRFFSCpxOBau8QTWpnFcGD8K\nSzWe5Fe07JBkn7g3Ji93uekAYHCYtgH4SyzKpenywpJI/DZ2Sf5YrH9OtE9OD1b1abqLD5Bsc/4U\ndjG2WGDGK9hF1KYjicRR2H6wGN1Ys0DTkUTiemw/WIwjaMKzFUgm8UHsyWlxfWyOAf6C4o8aEYGx\n2PXEdcC/YJvvPOA/sS4jXVTWWF8rUtPGMg5b2woTPh+g91PL0khq2lhWY902RgAHYm0sq2jye58r\nffbAJuwZogKdO7tQak3spsSdmQXk5tP1xAKUblwmSjd2QPtEByTRAUl0QBIdkEQHJNEBSXQgyblz\npSEZj1f4vcyQROJj4T3uVLDY6WFTnAYmkTgTu/FxAnbPXvS5fedhHdJvdi1dRkgicTjW9jwJu8Id\n5Srs3r/hwE0+RWtMVgLfKDH9G9jamFZS0RdnP0o/3PAD1O7cJ13A+cS3LQ8O07r6X6TskWSf+F3s\n5singR+Q7/Q5GXuozWTgr11L16Bcgm3ShYnG72MhlWmm7g+1ibInFhGQy3d4Gbs7fmMFy6oldX+o\nTZSNwL0VfK9hqUTieOwO/NFYpbsLu012DLY56F6XPrgO+Ih8z4dcf8QR2H4xzR08U1FPnIUlG/8z\ntk+M7hs2Yd1Lit2a0dAk2ZxnY3eZXoL1VSykA3uaWtORZE2ciN3xXoy3iZfb8CR9XPvQEtPHkf5q\nTlVIIvEpLLYljt2x074n+l2iDJJE4nXAHwD/QT6JaSwwDUtR2g/4jmvpGpS/wjbrwtO+LaS/20lq\nesqCrX1zgFuwWzIuxT9W6gos0qUDuBvYjQZJrasVrdg9MbuFz/dgd2s1RGpdIbti+8jPAwf3YzmF\n/BaLzxqC1WOHYE/cyGxq3VSszaTwysZ47M6q/8XWlOeBuU5lehe7drkak7cR24wzm1p3Aba2XVww\n/k4sLPIJrO/2qdgmtyRM6w8HYGdFrdjp5H8Bf14wT6ZS65Zj58pRJmFH5CWRcYOxbNlHK/qV3pwN\n3B75fD52JX0FKU2t62tzHsPOLXhTw3v0D92CHUUPo/90YneyDsYucpyC/TMfwNZ2wvtPw/D9wDnk\nU+smUOM7G/ranHfDBEXJ7bSXFIxfQ+9qR6U8h/UXfxpb45/B7uAahl0MnokdQM4K8y8P45djl+lm\nk7KqSCeWyhTl19itaYVchF2ESCt125wfx258PDR8PhNLrHsoZt5D2LlnhMAao97HrmK/hW1eWzFh\nUQZi1ZHCg1CaqNua+Ap2oXUhVn9bGD6/UDDfSWG6btVtcFJ52icCkuiAJDogiQ5IogOS6IAkOiCJ\nDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5I\nogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKID\naZR4IPBs5LUJS4jqb2pd0zIAeBMLuExtal3aORX4RRjuJB+yNoZ8zNUVmNQci7CEp0KaNgPiHCzW\nDzKcWldPBgGfpvdaliNTqXX1ZAa2aeboJKOpdfXkXPKbMlg6XSpT69LKUOAdLKkux0hgMfFVnCux\no3IncFqRZTbt0dmTptycM4MkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkii\nA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS\n6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkgicHFb2+h/aGubWOn3JTHC\ndktASUxaJe4J3AesAJZjYUFVT617Dw6u5HtplXgjsBCYDByGybkckzgReJR8KtNBwNnhfRpwCzX+\nu9IocQRwAnBH+PwRlqH4GWBeGDcPOCMMz8CSnLYDXVjI0JQalRVIp8TxwNvAXOAZ4F+xxCal1iVg\nF+Ao4CLgKeAGegeqgXNq3YIFNrqjo+O86PxZTq0bA7wa+fyHwIPYQaYqqXVtbW09bW1t5eSAZSbm\nah2wBjuAAJwCvAg8QEpT69K4OQPMAX6EiXkZuBAYCNwLzMQOIGeFeZeH8cuxg9Bsapwul1aJzwGf\niBl/SpH5rwmvupDGzTlzSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDog\niQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkO\nSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ6kVWIX8DzwLPmMm6qn\n1lVKWiX2AFOBI8mHpym1rgJaCj4rtS4hPcBi4Gngi2Fc1VPryswLywxjw/veQDuWp/ibgnneDe83\nA38WGX878NmYZbaTT7ur9NUeV9i0xly9Gd7fBn6CbZ7rsdS6dZjkt8I8a4H9It/dN4wr5IiqlJR0\nbs5DgGFheCh2tO3A0ulSmVqXRsZjm0078AJWfQGr4iwmvopzJXZA6QROq1lJhfBmELBHvQuRJc4C\nriVfz7wK2AZ0Y7uLvepUrszwt5iwp4DNwHexmsBlwFexOuatdStdRngJuCAMH4+tfX8amX46sLrG\nZcocH5KvU7YAW4EDI9P3xU4fE5HGemI1GYSJAzsD2UZvaR9hAcCJSOsZSzU5HNiArYkDgEPI1zlH\nVbLAwisljU53mfMl2kKbbU3cv94FaCTGEb+2tYRpogy6gdEx40cBO5IurNmOzn0xFKsGJaLZ9ok3\nR4avAT6IfN4Fu275XNKFNpvEQyPDk7F6Yo5twDLgOzUtUYaZCwz3Wliz7hOfB3atdyGyzmrs9O9e\ndCW8YgZg8u4BtgCvAd8CWutYpkwzEnsS0TPYBYjFwLlUcCGi2Tkee07gVqzBawP2nKxiTyISgTHA\n17AWwi3YM7GmhmmDgauxzVwU4QHsOmIHcDG2SRcymvKv+jQldwDH9jFPCzrQxHI61nNsRMy0PbHN\n98xaFiiLLAS+XGL6l7AnUIoSvE7vhqlCJhHfGaokzXbaN4rSB4seKmi8bzaJr2MNVcU4lArWxGbj\nRuyZp0Nipg3Fntx7U9KFNltr3z7Y6V038ANMGlin+YswH0dhHUlFCVqxo/QOTGZ3GF6I9Y1MTLOt\niVFGAh/HHKwk3wdcCCGEEJnn/wDva+G0ze3F+gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f0cf2f46e50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEUAAAInCAYAAADOLpRRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFUxJREFUeJztnXuUXVV5wH8zhJkhE0gyeQ3tCBFkMFEiigaQYhAzKmh5\ntIVWbDFqhSIFWgmPGqm3WhBhQG2KiksjKdW1QFJZZfFKBkkpVhheMRPygEACBDIhkMSQkMyE3Okf\n37nrnrnznXvuuY+zzz3z/da6a+4959wze36zzz777L2/vcEwDMMwjATQ4DoBwEZgJ7Af2AfMBm4E\nPgsMAi8AXwT+4Ch9TtgAtBVs6wIavffXe6/YaAw/JBYKc+wyIOu9fxzoiDMxSZAyBPQATwJfUfZ/\nCbgv1hQlgEO9n1OAFcDJvn0LgCWxpyhhfBO43Hs/D/gt0KId2Nrauh7JZZW8VtTmz6iMscDB3vtW\nRMIngU8DzwKTi3x3KOTcmRJ+v3qOMSV8sZZMA37tvR8D/AJYCjwPNCEFLsDvgK/GlSjXUjYAxyrb\nj4o7IX6ScPepFctdJ8AFYWVK2edwfflUnVktHZ0zm9vntjY2N+3ODgyuHujvWbl303NRzpEqKbNa\nOjq7xs2YN39K12BuW/fWZfOA26KISVWZMrO5fa5fCMD8KV2DM5rb50Y5T6qktDY2N2nbxwVsDyJV\nUnZnBwa17bsCtgeRKimrB/p7urcuG5Yrurcua1oz0N8T5TxJaGQqlyGU9M9q6eic0dw+d1xjc9Ou\n7MDgmuJ3H/Uc9UzN6impunyqRarqKTkymcyQ733ky8NyikKqpZSTSyDlUsrFpCiYFAWTomBSFEyK\ngklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWT\nomBSFEyKgklRMCkKSRjetZGRIbhtwB3A4d7+c4EdcSUoCTllCDgF+CAiBOBqJACqE3jI+xwbSZAC\nI8eyngEs9t4vBs6KMzFJkKKF4E4Dtnjvt3ifYyMJZcpJwGYkBHcZsLZgfy4yNDaSIGWz93MrEmQ5\nG8kd7UA/Erf8uvbF2bNnL/G9X33hhReuAViyRDb39fWd5z/+1ltvndHb2zsz97m3t7dqf0Q1CQrB\nvQG4ytt+NfpcB4G5J5PJDPkHGBchkeFyQSG4TwJ3Al8mf0uODddSgkJwtwGRormqSRLuPonDpCiY\nFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkK\nJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyK\ngklRSIKUA4BngHu8z7OBXm/bE8BH4k5QEqRcBqwmH7p2A3ANEqf8z97nWHEtpQM4Hfgp+djkzcB4\n7/0E4NW4E+U6XO57wBXAIb5tVwOPAt3IP+3EuBPlUspnkdDaZ5Cw/hw/Ay5FAi7PARYhK22PII0h\nuNcBryDBlZuB3cDtyGQQORoIXnu9ZiG4LsuUrwPvAt4N/BXwG+BvgPXAHO+YU4FIq01WA9dlip/c\nf+0C4BagGdjjfY6VpEj5H+8FEqh9vMO0OL8lJxKTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyK\ngklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWT\nomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKSZBSGIILcAmwBlgFfDfuBCUh3icXgptb\nufLjyCq4s5D1k6fEnSDXOUULwb0I+A4iBGTN01hxLSUXgpv1bTsK+BjwGLAc+HDciUpiCO4YYCJw\nAhLSfydwhHaC0RSCez/5aFOQ6NNJyvdHVQju3UjoLcga7E3Am3EmLAl3nxy5/9oi79UHDALnx52Q\npEjxh+DuQ3KMM1zffRKJSVEwKQomRcGkKJgUBZOiYFIUTIqCSVEwKQqpltIGWw6GZ6N+ryH8kMQy\nRED622DLm7ADYBJM2AbTopwj1TmlXEyKgklRiNrINBY4GpiKXI9bgXXA21VOl1NKkdIGzEMmtzxO\n+c4+4CngV8BtwPbqJc8Nxe4+E5DJcr+KTFm4DumLeQFpSG5AhL0H6Y7oBPYCPwS+TfBEmNWiZnef\nYryBTJB7DdLiHsaRwDeB17zv1prALow22DIE64ZgXRtsKeccQVyG5JCotAD/UMb3olIzKVZ5s8pb\naVQi5RzgAeTZYimO+2qqSblSLgYWI/WU/0Z68m4DrqpOspJNkLRVSP+vn+8BG2uamuHUrKANyylP\nIxW2QlqBTQXbXgXGhZyvLgiT8gjwf8BNwEG+7fcgl8uXkOnbLwe+wfBxa6nmeGAl8CLwKW/bQcCP\nkGeeLFKT/Tn5cWtx4KTy5udAYAEi4T+Byd72RuTh8IBKTl4mzsqUHPuAa4FjkcF7a5FxI1lkiNb+\nEs+TWhqAvwW2IfWTUp6LaoHTnPJnyCXzX8B85FL6KTATqUqvQkY4jpra8aXIJbIcWTDkbeC+gmP+\nFBnQ9zTwoRjT5qygfQH4d9/nzyCSCodwHgwsBAbK+SVl4uzymYxcHjlyfShtBce9hYynn0MKCGuO\n/B3wj8jY+e1Ig9NO5O6j8Vj1kpZcjkQqblnv9QZwptMU5anZ5ROWU15A6iZHIbXYtUjtNdWU0pqf\nRRqtRw3FCtqJFZy3ku86p5iUjUjrvBYsEMRUpHtjQwVpSjQXIc81A0jF7SvABxjeZnIwUmG7GLgX\neUbqBy6MIX3OKm/jgX9Cck3W9xrwXv5t65HqflzNB867OBqRNUfnIM88U8j3Ja8CHka6TuOkZl0c\npXawZ5GK2aionCXhyVYLwQVp4swy8pGi5iRBSuEquCARY13ASy4S5FqKFoILcDNwpZMU4V6KFoJ7\nJtJ9stJJikheCO5YJODSv+pt4B0yKAS367jjWvY3NR0M0DQ42NL31FPnQf2G4N6F1Cs2eK99SB1p\nqvJ9510cIP06xdYcnY1EiZbDHPSOtA0E332cd3EAfAFpXwniCGRsXLloCaz4P1kO1SxTWslP2hAV\nfwiuHzWcv9aESTnce+UKuxnI5AyFTEIeINdXL2nJJcPwh75ir3eIN9rcWXPk3eTHnCwCfsLI558h\nYBfQi9xN6p4wKSu8F8B0YAkyB0GqiVLQZmqViKQR9e7TCMxFRllPQq9tfqvSRNUTRyFPs2EFblw4\nK2j9LETqDVciLW2xTvQSJ1GknAz8AOiuUVoSQ5Rq/gAy7i31RJHyIHBSrRKSJKJI+RpwIjKaqak2\nyUkGUaI4NiAPfZORgX+vMXwAYANSmsf1EOe8iwOkETksksrJo361SV28z9DQ0Ih/TENDQ9DfafE+\npVJOI9Mc4JNIu+lNyECecUhHex+Oo00bGhoaCsuUWv6+A5DJLXPV+f3kpzM8CBlsvKCWCSggEW20\nVyEDjb+GtMD5r8U9yHCN0yKcL7FEkXI+Mgvo99Gfe9YiT891TxQp05HYnyB2UOfDunJEkfIWxUcA\nHImDqZdrQRQpjwJ/HfCdiUiU2MPVSJRroki5FpnP4DdIPzDIGNu/Q/qDxwHXVzV1dcJnkIF+ha1t\n/UjdJU4S0fIGMgJyOjIqIHdbfg5pVkjVHCqlchgyVCKIsd4xcZGIyttG4Kwi+88gJYOKq/lAmJqH\ny2r+Ie/Fewird8IK2i94rxwLkEjTQiYB70eef+qeMCkTGd68OAVpkvST62D/GfE+JSeCLPB514nw\nkYh6SmoK0jBGzR8ahahSPoc0H2xleDV/v+9n3RPl8rkCWb/rDWQ0k9bQlIoujihSLgYeR9pl99Qm\nOckgyuXTjjRHploIRJPyAjXuLkgKUaR0I7XZOKcYckKUMiWLVITWIOP0X0S/2/xHFdLllCh9yaWM\nZxsivvmZEjHq4NTwQ9JBFCnLa5WIpFFJNX8y+anNUkVUKX+MFKR/QELdXkdGGSz29pVLYRjujUiB\n/ntkgqzxFZw7MlGkHAY8iXSIvQj80nttQKZdfQIJnS2HwjDcpcD7kLkVnkOmFkgki5Ha7OnKvtOQ\nyWYWl3HeDqAHWRJYC5k7G5lSrZBExBBuRuKFg7gZ6RSLyq+ADxIcR3gPcJ6yPRFdHBORrBzEeqKP\nOvCH4Wp1jgXIBMC/jHjeiohyS34VyeI/Dth/MiPnqA3jo0h/0enILOqHIAX5+UiQ5unAJ4K+nIS4\n5O8gtdrrGX43GO/bd10F5/dfPp9G5pQrdstPRJnSigzHyCJRpS95r3e8bf9L8W7VME5B5ssGeN47\n9zPe64fK8c4nlclxIJKtzyY/u+iLSH/PbYiguKjZs0/qBheDLSZQE0yKgklRMCkKJkXBpCiUIuUi\n4NyQY/6SeKZGjIUwKWcDtxA+Qmk7suJCUibwrYgwKZ9H+o2Xhhy3FPgtw0c91S1hUo5n5JTwQTyA\nzMvknEszman/ksl0lvv9MClTKb054DX0WbacMQkm7Ctjpbuw9pS3kTaOUjiEhI26LvLMU5SwnLKe\n0ufCPxl55K97wqTcg9xRPhpy3AnIaOxULHoUJmUhMpTrXuACRi7W2IJM83w/0ta6sNoJdEGYlO1I\nThlE2mZ3IHM0PYK0iG0HbvX2n0kKFnuF0hque5FOqSuAPwdm+fa9jExedQPFm/3qilJb8/uRmYQv\nRwbtHIKsyfFWjdLllHIi2N8ipTJylCLlUO/nZu9nCzJSsrAl/BWkt6/uCZNyNNL/sgAZQwsSQHmj\ncuw7SCFc93WVsLvPF5E5DL6v7JuPjG46FenF2wV8uaqpc0RYTjkV6aDSlr1awfDRTXeQkiFgYTnl\nKKQ+UgprvePrnrCc0opcFn62IXWVwunNdjIyQKouCZOyg/zdJ0eW4Yur5WhHhn3VPWGXzypKj0zv\nIiXTsoZJuQsZDRDW9nqWd9xdlSfJPWFSFiEF6B3ICk+HF+yfDvyrt38N5U8RnyjCypS9yBCse5EK\n3NeRAnUn8vyTG7yz1jsuFSvPldLv8yIyM9dl5Aft/BH5gTqXevtTEdIPpT8Q7kEakFLRiBSGdZsq\nmBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMikISpBSG37YBy5DQvKU4mIkjCVIKw2+vRqR0\nAg95n2PFtRRtFdwzyIfyLqb4fJU1wbUUbRXcaeTHumyhzMF8leBSSlj4LcglFfvkVy6XBtbCb29H\nckc7MlDoUEScSlAI7pIlsrmvr29YkHc9LA3sxx9+ewOycAFIIRs0b3ZgDspkMkOZTKaUHFZxBHut\nySXweqS38TlkFEPsk4m7vHz8+FfB3YasdeiMJOWUxGBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyK\ngklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWT\nomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCi4ltICPI5M47oaWeQRZE2PXiS+\n8AngI3EmynUQ1F5kadC3vbQ8CvwJMvvgNcCDyAq7N3jHxYLrnAL59TuakGj27UhQZW62wQnIuqqx\n4TqngPxjngaORBZOehYJqHwU6Pb2nxhngpIgJQsci+SMB5HZSr+BTL/4a+AcZKLOrsIvpj0EN8c1\nyITiO33bGtAn/01tCO5k8rNeHITkhhUMX0LnVIovXl11XF8+hyKTPDR6r9uR9dgvQBZwa0Ym87wg\nzkS5ltKHTNtayJPIemVOcH35JBKTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXB\npCiYFIVUSymx8XoEqZYyGqnGPLWJ7OJIJCZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6Jg\nUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaWUoPBbgEuANcAq4LvxJ80tY72fY4DH\nkPDbjyNLAx/o7ZsS8N2a9SUnhbFIUPb7gDuQELkwUtvB3ohcPluAh5Hw207gY0jOWQ58OO5EuQ6X\n08JvxwATgROQcP47gSO0LweF4AZRjyG4ufDb+8lHmoJEnk5Sjk/l5aOF3z4D3E2+TOlEwv3fjD11\njjgGCedfAaxElh0HuevcjkSiPoVcUhqpv/uUQyovn8RiUhRMioJJUTApCiZFwaQomBQFk6JgUhRc\nNx1UnVktHZ0zm9vntjY2N+3ODgyuHujvWbl3U6Q5nVIlZVZLR2fXuBnz5k/pGsxt6966bB5wWxQx\nqbp8Zja3z/ULAZg/pWtwRnP73CjnSZWU1sbmJm37uIDtQaRKyu7swKC2fVfA9iBSJWX1QH9P99Zl\nw3JF99ZlTWsG+nuinKehusmKlSGU9M9q6eic0dw+d1xjc9Ou7MDgmuJ3H/Uc9Yy1vMWJSVEwKQom\nRcGkKJgUBZOiYFIUTIqCSVEwKQomRcGkKJgUBZOiYFIUTIqCSVEwKQomRcGkKJgUBZOiYFIUTIqC\nSVEwKQomRcG1lGJhuACXIyF1bWWc+5SKUuYYLQwX4F3AA8AGdClhow4yJfzuxI46KFwFd5v3+Wbg\nShcJSoKUwjDc1cCZwCYkjG5UMx65fE73fh7ibd+AHm26Asn+lbxWaAlJ2tCma5DEXkL+supAFpGe\nDbzuKF2xUhiG+wjwiYJjggramuF6xLW2Cu5DBcekLkzWSBBNwDjXiXDJuUhN+Czv87eBQaQm3IN+\n50o1lyMCngB2ATcBm5FK3xVIXefHzlLniHXAPO/9SUju+Avf/tOAl2NOk3P2Is9IIHWuAeBo3/4O\nYF/Ukyahml8JTYgIkFv3IMMlvIM8T0XCdT2lGnwAmV+lAfknv598hXByOSdMWjU/KtkSj4t0RdR7\nTlHnajKEw9BzQ4O3b1SSBaYq2ycD+6OerN7vPmG0IrftSNR7mbLQ9/468m0wIH/bbOD3UU9a71KO\n8b2fgdRTcgwis391x5qiBPFz8s2XFZOWMmUl+VlJDY+Xker+ncCnHKclMTQiMu4A9gAvAd8CpjtM\nU6JoQ3oDnkYeCHuAz1HGg2HaOAlYhFxS65EHxleASLNjpIF24CpgLXIJ/YJ8J/tBwLXIZTVquAdp\nR+kDLkXvJ5pK6U/VqWARMtNxMRoYJQXvacBGpB+6kAnI5XJ2nAlKAvcBFxfZfxEyxmVUsYnhDdWF\nvBfpnI9EvVfzJ1O88ByijM6wepeyCWm4DuIYysgp9c4PkJFPY5V9rch6Hv8W9aT13po/DanOZ4Fb\nEAkAM4G/R/6+DwH9TlLnkOnIXWg/Iifrvb8PeHc5J6z3nOKnDXgP8jc9T36UpWEYhmHUAf8PG3N/\nEIQIcHAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f0cf2e55dd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "pj.report_seq_stats()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 3745\n",
      "Number of species (see below for full list): 67\n",
      "['Rutilus rutilus', 'Proterorhinus semilunaris', 'Salmo salar', 'Gymnocephalus cernua', 'Leuciscus leuciscus', 'Lota lota', 'Phoxinus phoxinus', 'Thymallus thymallus', 'Lampetra fluviatilis', 'Barbus barbus', 'Hypophthalmichthys nobilis', 'Esox lucius', 'Alosa alosa', 'Oncorhynchus gorbuscha', 'Barbatula barbatula', 'Oncorhynchus mykiss', 'Sander lucioperca', 'Coregonus autumnalis', 'Petromyzon marinus', 'Squalius cephalus', 'Ponticola kessleri', 'Leuciscus idus', 'Leucaspius delineatus', 'Coregonus lavaretus', 'Misgurnus fossilis', 'Ctenopharyngodon idella', 'Lepomis gibbosus', 'Cyprinus carpio', 'Salvelinus fontinalis', 'Tinca tinca', 'Chondrostoma nasus', 'Perca fluviatilis', 'Coregonus oxyrinchus', 'Gasterosteus aculeatus', 'Vimba vimba', 'Alburnus alburnus', 'Cottus gobio', 'Platichthys flesus', 'Micropterus salmoides', 'Ameiurus melas', 'Carassius auratus', 'Acipenser sturio', 'Alburnoides bipunctatus', 'Pseudorasbora parva', 'Hypophthalmichthys molitrix', 'Blicca bjoerkna', 'Pimephales promelas', 'Umbra pygmaea', 'Silurus glanis', 'Cobitis taenia', 'Anguilla anguilla', 'Coregonus albula', 'Rhodeus sericeus', 'Carassius carassius', 'Ambloplites rupestris', 'Scardinius erythrophthalmus', 'Alosa fallax', 'Gobio gobio', 'Abramis brama', 'Ameiurus nebulosus', 'Neogobius melanostomus', 'Pungitius pungitius', 'Osmerus eperlanus', 'Aspius aspius', 'Salvelinus alpinus', 'Lampetra planeri', 'Salmo trutta']\n"
     ]
    }
   ],
   "source": [
    "species=[]\n",
    "for r in pj.records:\n",
    "    source = [f for f in r.features if f.type == 'source'][0]\n",
    "    species.append(\" \".join(source.qualifiers['organism'][0].split()[0:2]))\n",
    "    \n",
    "#print set(species)\n",
    "total=len(species)\n",
    "species=list(set(species))\n",
    "\n",
    "print \"Total number of records: %i\" %total\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract ids of representative sequences identified by vsearch -> remove all others from the object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 1983\n",
      "to keep: 1762\n",
      "TOTAL: 3745\n"
     ]
    }
   ],
   "source": [
    "ids = []\n",
    "\n",
    "uc=open('CytB_id-1.uc', 'r')\n",
    "\n",
    "for line in uc:\n",
    "    if line.startswith('C'):\n",
    "        ids.append(line.split(\"\\t\")[8])\n",
    "        \n",
    "recs_to_keep = {'CytB': ids}\n",
    "recs_to_drop = {'CytB': []}\n",
    "\n",
    "for r in pj.records:\n",
    "    if not r.id in recs_to_keep['CytB']:\n",
    "        recs_to_drop['CytB'].append(r.id)\n",
    "        \n",
    "print \"to drop: %s\" %len(recs_to_drop['CytB'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['CytB'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['CytB']) + len(recs_to_keep['CytB']))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adjust lists -> add record to exlusion list and remove record from inclusion list.\n",
    "\n",
    " - `JN995186` is a pseudogene and can not be translated.\n",
    " - `KP794942` is actually COI. misslabelled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1985\n"
     ]
    }
   ],
   "source": [
    "recs_to_drop['CytB'].append('JN995186.1')\n",
    "recs_to_drop['CytB'].append('KP794942.1')\n",
    "print len(recs_to_drop['CytB'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1760\n"
     ]
    }
   ],
   "source": [
    "for r in range(len(recs_to_keep['CytB'])):\n",
    "    if 'JN995186' in recs_to_keep['CytB'][r]:\n",
    "        del(recs_to_keep['CytB'][r])\n",
    "        break\n",
    "        \n",
    "for r in range(len(recs_to_keep['CytB'])):\n",
    "    if 'KP794942' in recs_to_keep['CytB'][r]:\n",
    "        del(recs_to_keep['CytB'][r])\n",
    "        break\n",
    "        \n",
    "print len(recs_to_keep['CytB'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude redundant records from object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.exclude(start_from_max=False, **recs_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 1760\n",
      "Number of species (see below for full list): 67\n",
      "['Rutilus rutilus', 'Proterorhinus semilunaris', 'Salmo salar', 'Gymnocephalus cernua', 'Leuciscus leuciscus', 'Lota lota', 'Phoxinus phoxinus', 'Thymallus thymallus', 'Lampetra fluviatilis', 'Barbus barbus', 'Hypophthalmichthys nobilis', 'Esox lucius', 'Alosa alosa', 'Oncorhynchus gorbuscha', 'Barbatula barbatula', 'Oncorhynchus mykiss', 'Sander lucioperca', 'Coregonus autumnalis', 'Petromyzon marinus', 'Squalius cephalus', 'Ponticola kessleri', 'Leuciscus idus', 'Leucaspius delineatus', 'Coregonus lavaretus', 'Misgurnus fossilis', 'Ctenopharyngodon idella', 'Lepomis gibbosus', 'Cyprinus carpio', 'Salvelinus fontinalis', 'Tinca tinca', 'Chondrostoma nasus', 'Perca fluviatilis', 'Coregonus oxyrinchus', 'Gasterosteus aculeatus', 'Vimba vimba', 'Alburnus alburnus', 'Cottus gobio', 'Platichthys flesus', 'Micropterus salmoides', 'Ameiurus melas', 'Carassius auratus', 'Acipenser sturio', 'Alburnoides bipunctatus', 'Pseudorasbora parva', 'Hypophthalmichthys molitrix', 'Blicca bjoerkna', 'Pimephales promelas', 'Umbra pygmaea', 'Silurus glanis', 'Cobitis taenia', 'Anguilla anguilla', 'Coregonus albula', 'Rhodeus sericeus', 'Carassius carassius', 'Ambloplites rupestris', 'Scardinius erythrophthalmus', 'Alosa fallax', 'Gobio gobio', 'Abramis brama', 'Ameiurus nebulosus', 'Neogobius melanostomus', 'Pungitius pungitius', 'Osmerus eperlanus', 'Aspius aspius', 'Salvelinus alpinus', 'Lampetra planeri', 'Salmo trutta']\n"
     ]
    }
   ],
   "source": [
    "species=[]\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['CytB']:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        species.append(\" \".join(source.qualifiers['organism'][0].split()[0:2]))\n",
    "    \n",
    "#print set(species)\n",
    "total=len(species)\n",
    "species=list(set(species))\n",
    "\n",
    "print \"Total number of records: %i\" %total\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write full Genbank records to file for later processing with `SATIVA`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seqs = []\n",
    "from Bio import SeqIO\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['CytB']:\n",
    "        seqs.append(r)\n",
    "\n",
    "out = open('CytB_nr_pre_Sativa.gb','w')\n",
    "SeqIO.write(seqs, out, 'genbank')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure Alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 921091473202105.88_CDS_proteins_CytB.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj,                              # The Project\n",
    "                    method_name='mafftLinsi',        # Any unique method name,\n",
    "                    CDSAlign=True,                   # Use this method to align\n",
    "                    codontable=2,                    # The genetic code that\n",
    "                    program_name='mafft',            # mafft or muscle.\n",
    "                    cmd='mafft',                     # The command on your machine\n",
    "                    loci=['CytB'],                 # A list of loci names to align.vel.\n",
    "                    cline_args={'localpair': True,# Program specific keywords and arguments.\n",
    "                                'maxiterate': 1000}# cine_args=={} by default, which will\n",
    "                    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlnConf named mafftLinsi with ID 921091473202105.88\n",
      "Loci: CytB \n",
      "Created on: Tue Sep  6 22:48:25 2016\n",
      "Commands:\n",
      "CytB: mafft --localpair --maxiterate 1000 921091473202105.88_CDS_proteins_CytB.fasta\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-3.13.0-54-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.66\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.0\n",
      " User: chrishah-office\n",
      " Program and version: MAFFT v7.123b\\nPal2Nal v14\n",
      " Program reference:Katoh\n",
      " Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\\nMikita Suyama\n",
      " David Torrents\n",
      " and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34\n",
      " W609-W612.\n",
      "execution time:\n",
      "1975.26133895\n",
      "\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The dataset(s) CytB were first aligned at the protein level using the program MAFFT v7.123b [1].\n",
      "The resulting alignments served as guides to codon-align the DNA sequences using Pal2Nal v14 [2].\n",
      "\n",
      "Reference:\n",
      "[1]Katoh, Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\n",
      "[2]Mikita Suyama, David Torrents, and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34, W609-W612.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['mafftLinsi']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nr.pkl'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'nr.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CytB@mafftLinsi_aln.fasta']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_alns(id=['record_id','source_organism'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alignemnt looks ok. Needs a bit of end clipping. TrimAL, the trimming method implemented in ReproPhylo doesn't seem to allow for end clipping only, so the below cell contains a custom solution for start/end-clipping the alignemnt. \n",
    "\n",
    "We only want to trim the alignment from the end because this is the alignment that should be used for phylogenetic placement and it needs to be possible to align reads to the alignment.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clipping positions are determined from start/end of the alignment by assessing the average per column completeness across 10 bp sliding windows. per column completeness is the proportion of sequences represented by `ACGT` in a given column. Average completeness is assessed across 10bp sliding windows. \n",
    "\n",
    "The clipping point is defined as the position at which the alignment, starting from the 5'/3'-ends, respectively, retained an average completeness > 0.7 (i.e. 70% of the sequences are informative) for 10 consecutive sliding windows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object CytB@mafftLinsi\n",
      "\n",
      "Finding forward clip position:\n",
      "\n",
      "1 29 0.7025\n",
      "2 30 0.70875\n",
      "3 31 0.714204545455\n",
      "4 32 0.719659090909\n",
      "5 33 0.725852272727\n",
      "6 34 0.729943181818\n",
      "7 35 0.734034090909\n",
      "8 36 0.738125\n",
      "9 37 0.740397727273\n",
      "10 38 0.742670454545\n",
      "Suggested forward clip at: 28 (10,38)\n",
      "\n",
      "Finding reverse clip position:\n",
      "\n",
      "1 1037 0.701477272727\n",
      "2 1036 0.704772727273\n",
      "3 1035 0.708011363636\n",
      "4 1034 0.711306818182\n",
      "5 1033 0.714602272727\n",
      "6 1032 0.717897727273\n",
      "7 1031 0.721136363636\n",
      "8 1030 0.721193181818\n",
      "9 1029 0.72125\n",
      "10 1028 0.722443181818\n",
      "Suggested reverse clip at: 1038 (10,1028)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "per_col=[]\n",
    "\n",
    "aln = pj.fa('CytB@mafftLinsi')\n",
    "\n",
    "\n",
    "icol=0\n",
    "for a in aln[1,:]:\n",
    "    ACGT_count=0\n",
    "    col=aln[:,icol]\n",
    "    for i in range(len(col)):\n",
    "        if col[i] in ['A','G','C','T']:\n",
    "            ACGT_count+=1\n",
    "#    print pj.fa('CytB@mafftLinsi')[:,a]\n",
    "    per_col.append(float(ACGT_count)/len(col))\n",
    "    icol+=1\n",
    "\n",
    "wsize=10\n",
    "min_prop=0.7\n",
    "    \n",
    "print \"\\nFinding forward clip position:\\n\"\n",
    "forw_count=0\n",
    "forw_clip=0\n",
    "for i in range(len(per_col)-wsize):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        forw_count+=1\n",
    "        forw_clip=i\n",
    "        print forw_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        forw_count=0\n",
    "        \n",
    "    if forw_count >= wsize:\n",
    "        break\n",
    "        \n",
    "print \"Suggested forward clip at: %s (%s,%s)\" %(forw_clip-wsize,forw_count,forw_clip)\n",
    "\n",
    "print \"\\nFinding reverse clip position:\\n\"\n",
    "reve_count=0\n",
    "reve_clip=0\n",
    "for i in reversed(range(len(per_col)-wsize)):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        reve_count+=1\n",
    "        reve_clip=i\n",
    "        print reve_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        reve_count=0\n",
    "        \n",
    "    if reve_count >= wsize:\n",
    "        break\n",
    "    \n",
    "print \"Suggested reverse clip at: %s (%s,%s)\" %(reve_clip+wsize,reve_count,reve_clip)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clip the alignemnt at the positions determined above and write to file. We'll use Biopython functions to do this."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clip alignment - the new Biopython alignemnt object is called `clipped_aln`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object CytB@mafftLinsi\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fa('CytB@mafftLinsi')\n",
    "\n",
    "clipped_aln = aln[:,forw_clip-wsize:reve_clip+wsize]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the clipped alignemnt (which is a Biopython alignment object) to a ReproPhylo trimmed alignment object, so that we can resume working with it in ReproPhylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.trimmed_alignments[\"CytB@mafftLinsi_aln_clipped\"] = clipped_aln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CytB@mafftLinsi_aln_clipped': <<class 'Bio.Align.MultipleSeqAlignment'> instance (1760 records of length 1010, SingleLetterAlphabet()) at 7f0cf2c939d0>}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.trimmed_alignments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write clipped alignment to file for downstream taxonomic curation using `SATIVA`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "out=open('CytB@mafftLinsi_aln_clipped.phy', 'w')\n",
    "AlignIO.write(clipped_aln, out, 'phylip-relaxed')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to run `SATIVA` is described in [this]() notebook. It may be a good strategy to set this going now in a separate notbook. Then contnue onwards in this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer Maximum Likelihood tree based on the clipped alignment. We are going to use [RAxML](http://sco.h-its.org/exelixis/web/software/raxml/index.html)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure RAxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 553111473204098.13_CytB@mafftLinsi_aln_clipped0 -p 564 -s 553111473204098.13_CytB@mafftLinsi_aln_clipped.fasta -T 5 -x 809 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-mafftLinsi-clipped', \n",
    "                 program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                 keepfiles=True,\n",
    "                 cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                 preset='fa',\n",
    "                 alns=['CytB@mafftLinsi_aln_clipped'],\n",
    "                 model='GAMMA',\n",
    "                 threads=5,\n",
    "                 cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run RAxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named raxml-mafftLinsi-clipped with ID 553111473204098.13\n",
      "Alignments: CytB@mafftLinsi_aln_clipped \n",
      "Created on: Tue Sep  6 23:21:38 2016\n",
      "Commands:\n",
      "CytB@mafftLinsi_aln_clipped:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 553111473204098.13_CytB@mafftLinsi_aln_clipped0 -p 564 -s 553111473204098.13_CytB@mafftLinsi_aln_clipped.fasta -T 5 -x 809 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-3.13.0-54-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.66\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.0\n",
      " User: chrishah-office\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "28669.8644731\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) CytB@mafftLinsi_aln_clipped using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-mafftLinsi-clipped']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nr.pkl'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'nr.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize the tree."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare tree annotation:\n",
    " - define outgroups\n",
    " - specify support coloration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Lampetra','Petromyzon']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='mafftLinsi-clipped-raxml.html')\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check tree out. Have a think which sequences you would consider mislabelled and should better be removed from the reference database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a static copy of the tree (the last `*.png` with the expected alignemnt id) for later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png CytB_mafftLinsi_clipped_raxml.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "##=========================================="
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In the meantime we have run `SATIVA` on the alignment that was the basis of the above tree (see [here](https://github.com/HullUni-bioinformatics/metabarcode-course-2016/blob/master/data/exercise-5/supplementary_data/reference_db/3-SATIVA/run_SATIVA.ipynb)).\n",
    "\n",
    "Now we are going to  highlight the records that `SATIVA` considers mislabelled in the tree. We are also going to add a heatmap to the tree illustrating the confidence that `SATIVA` has in a given call."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse `SATIVA` output and record putatively mislabelled sequence records and the confidence of the call."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "rec_mislabelled = {}\n",
    "sativa=open('../3-SATIVA/CytB.mis', 'r')\n",
    "\n",
    "for line in sativa:\n",
    "    if not line.startswith(\";\"):\n",
    "        rec_mislabelled[line.split(\"\\t\")[0]] = line.split(\"\\t\")[4]\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add two new qualifiers containing info about he `SATIVA` results to the sequence features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "for r in pj.records:\n",
    "    if r.id in rec_mislabelled:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['mislabeled'] = ['yes']\n",
    "            f.qualifiers['mislabeled_conf'] = [rec_mislabelled[r.id]]\n",
    "    else:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['mislabeled'] = ['no']\n",
    "            f.qualifiers['mislabeled_conf'] = [0.0001]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.propagate_metadata()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare tree annotation:\n",
    "\n",
    " - color records that were identified by SATIVA as being mislabelled.\n",
    " - add heatmap with mislabel confidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "leaf_colors = {'yes':'red', 'no': 'black'}\n",
    "bg_colors = {'yes':'red', 'no': 'white'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='mafftLinsi-clipped-raxml-SATIVA.html',\n",
    "            fsize=13,\n",
    "            heat_map_meta = ['mislabeled_conf'], heat_map_colour_scheme = 1,\n",
    "            leaf_node_color_meta = 'mislabeled', leaf_label_colors = leaf_colors)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png CytB_mafftLinsi_clipped_raxml_SATIVA_leaf_colored.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='mafftLinsi-clipped-raxml-SATIVA.html',\n",
    "            fsize=13,\n",
    "            heat_map_meta = ['mislabeled_conf'], heat_map_colour_scheme = 1,\n",
    "            node_bg_meta = 'mislabeled', node_bg_color = bg_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "target=CytB_mafftLinsi_clipped_raxml_SATIVA_bg_colored.png\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png $target"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Have a look!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nr.pkl'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'nr.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
