{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude the records that SATIVA has highligted, redo the alignment and rebuild the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from reprophylo import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj=Project('../1-initial/target_locus.csv', pickle='post_SATIVA.pkl', git=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reprophylo/reprophylo.py:1015: UserWarning: Version control off\n",
      "  warnings.warn('Version control off')\n"
     ]
    }
   ],
   "source": [
    "pj.read_embl_genbank(['../2-nr/CytB_nr_pre_Sativa.gb'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set the stage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 0\n",
      "to keep: 0\n",
      "TOTAL: 0\n"
     ]
    }
   ],
   "source": [
    "recs_to_keep = {'CytB': []}\n",
    "recs_to_drop = {'CytB': []}\n",
    "\n",
    "print \"to drop: %s\" %len(recs_to_drop['CytB'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['CytB'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['CytB']) + len(recs_to_keep['CytB']))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Record all record ids and assign them the status keep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "for r in pj.records_by_locus['CytB']:\n",
    "    recs_to_keep['CytB'].append(\"_\".join(r.id.split(\"_\")[:-1]))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 0\n",
      "to keep: 1760\n",
      "TOTAL: 1760\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print \"to drop: %s\" %len(recs_to_drop['CytB'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['CytB'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['CytB']) + len(recs_to_keep['CytB']))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many do we want to drop? Parse `SATIVA` output and extract records ids putatively mislabeled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rec_mislabelled = {}\n",
    "sativa=open('../3-SATIVA/CytB.mis', 'r')\n",
    "\n",
    "for line in sativa:\n",
    "    if not line.startswith(\";\"):\n",
    "        rec_mislabelled[line.split(\"\\t\")[0]] = line.split(\"\\t\")[4]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    }
   ],
   "source": [
    "print len(rec_mislabelled)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adjust lists accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EU492281.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "EU224046.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "AJ969128.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "AJ937943.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JF489783.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "KT275288.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "DQ185405.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "KT275289.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "EU224045.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JQ346141.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JQ661401.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "AJ937931.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "AJ937952.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "AJ937925.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JQ661398.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "AJ937951.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "KP644340.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "DQ664351.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "GQ279764.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "GU182336.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JQ661399.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "KP452507.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JQ661400.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n",
      "JQ231114.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n"
     ]
    }
   ],
   "source": [
    "for r in rec_mislabelled:\n",
    "    print r\n",
    "    if not r in recs_to_drop['CytB']:\n",
    "        print \"\\tadd to droplist\"\n",
    "        recs_to_drop['CytB'].append(r)\n",
    "    for i in reversed(range(len(recs_to_keep['CytB']))):\n",
    "#        print i\n",
    "        if r in recs_to_keep['CytB'][i]:\n",
    "            print \"\\tremove from keeplist\"\n",
    "            del(recs_to_keep['CytB'][i])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 24\n",
      "to keep: 1736\n",
      "TOTAL: 1760\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print \"to drop: %s\" %len(recs_to_drop['CytB'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['CytB'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['CytB']) + len(recs_to_keep['CytB']))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out good records in full Genbank format including all metadata. Adjust to valid species and taxids before doing it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify records with Subspecies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ameiurus melas  .. add to records\n",
      "Alburnoides bipunctatus  .. add to records\n",
      "Ambloplites rupestris  .. add to records\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. add to records\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus strymonicus\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Ambloplites rupestris  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "subspecies: Alburnoides bipunctatus tzanevi\n",
      "Alosa fallax  .. add to records\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Acipenser sturio  .. add to records\n",
      "subspecies: Alburnoides bipunctatus strymonicus\n",
      "Ameiurus nebulosus  .. add to records\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alosa fallax  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Acipenser sturio  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnus alburnus  .. add to records\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Alburnus alburnus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "subspecies: Alburnoides bipunctatus strymonicus\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Abramis brama  .. add to records\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alosa fallax  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Abramis brama  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "subspecies: Alburnoides bipunctatus tzanevi\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Abramis brama orientalis\n",
      "Anguilla anguilla  .. already covered\n",
      "Ambloplites rupestris  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnus alburnus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus strymonicus\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus strymonicus\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnus alburnus  .. already covered\n",
      "Alburnus alburnus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alosa fallax  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Ambloplites rupestris  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Abramis brama  .. already covered\n",
      "Aspius aspius  .. add to records\n",
      "Anguilla anguilla  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Ameiurus nebulosus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Ameiurus melas  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Ameiurus nebulosus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Aspius aspius  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Alburnoides bipunctatus  .. already covered\n",
      "subspecies: Alburnoides bipunctatus rossicus\n",
      "Alosa fallax  .. already covered\n",
      "subspecies: Alburnoides bipunctatus strymonicus\n",
      "Alburnus alburnus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Barbus barbus  .. add to records\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbatula barbatula  .. add to records\n",
      "Barbus barbus  .. already covered\n",
      "Blicca bjoerkna  .. add to records\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Blicca bjoerkna  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Blicca bjoerkna  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Barbus barbus barbus\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbus barbus  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Barbus barbus barbus\n",
      "Barbatula barbatula  .. already covered\n",
      "subspecies: Cyprinus carpio carpio\n",
      "Carassius auratus  .. add to records\n",
      "Coregonus lavaretus  .. add to records\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. add to records\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius carassius  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Chondrostoma nasus  .. add to records\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Chondrostoma nasus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "subspecies: Carassius auratus langsdorfii\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Cyprinus carpio  .. add to records\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "subspecies: Carassius auratus langsdorfii\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Cobitis taenia  .. add to records\n",
      "subspecies: Cyprinus carpio carpio\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Coregonus autumnalis  .. add to records\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Cobitis taenia  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Cobitis taenia  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Cyprinus carpio  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Ctenopharyngodon idella  .. add to records\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Coregonus lavaretus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "subspecies: Carassius auratus auratus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus autumnalis  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Cyprinus carpio 'xingguonensis'\n",
      "Carassius auratus  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "subspecies: Coregonus autumnalis autumnalis\n",
      "subspecies: Cyprinus carpio 'wuyuanensis'\n",
      "Cyprinus carpio  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus langsdorfii\n",
      "subspecies: Carassius auratus langsdorfii\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "subspecies: Cyprinus carpio haematopterus\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus ssp. 'Pingxiang'\n",
      "Carassius auratus  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "subspecies: Carassius auratus langsdorfii\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Carassius auratus auratus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Carassius auratus  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Cobitis taenia  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Cyprinus carpio  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Cyprinus carpio  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Cyprinus carpio 'wananensis'\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Carassius auratus grandoculis\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Cyprinus carpio  .. already covered\n",
      "Carassius carassius  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Cottus gobio  .. add to records\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "subspecies: Chondrostoma nasus nasus\n",
      "subspecies: Cyprinus carpio carpio\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Carassius auratus  .. already covered\n",
      "Coregonus albula  .. add to records\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Esox lucius  .. add to records\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Gobio gobio  .. add to records\n",
      "Gasterosteus aculeatus  .. add to records\n",
      "Gobio gobio  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gymnocephalus cernua  .. add to records\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gymnocephalus cernua  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gymnocephalus cernua  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "Hypophthalmichthys nobilis  .. add to records\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys molitrix  .. add to records\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Leuciscus leuciscus  .. add to records\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. add to records\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus idus  .. add to records\n",
      "Lota lota  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. add to records\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leucaspius delineatus  .. add to records\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. add to records\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lepomis gibbosus  .. add to records\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leucaspius delineatus  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lampetra planeri  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Lampetra fluviatilis  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Micropterus salmoides  .. add to records\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Misgurnus fossilis  .. add to records\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Misgurnus fossilis  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "subspecies: Micropterus salmoides salmoides\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Neogobius melanostomus  .. add to records\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Neogobius melanostomus  .. already covered\n",
      "Oncorhynchus gorbuscha  .. add to records\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. add to records\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Osmerus eperlanus  .. add to records\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Osmerus eperlanus  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Osmerus eperlanus  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Oncorhynchus gorbuscha  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Pseudorasbora parva  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Perca fluviatilis  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. add to records\n",
      "Proterorhinus semilunaris  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pimephales promelas  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Platichthys flesus  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Ponticola kessleri  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Ponticola kessleri  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "subspecies: Pungitius pungitius pungitius\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Ponticola kessleri  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Phoxinus phoxinus  .. add to records\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "subspecies: Phoxinus phoxinus tumensis\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Petromyzon marinus  .. add to records\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Ponticola kessleri  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Ponticola kessleri  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "subspecies: Phoxinus phoxinus tumensis\n",
      "Perca fluviatilis  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Ponticola kessleri  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Proterorhinus semilunaris  .. already covered\n",
      "Rutilus rutilus  .. add to records\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rhodeus sericeus  .. add to records\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rhodeus sericeus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rhodeus sericeus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rhodeus sericeus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Squalius cephalus  .. add to records\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. add to records\n",
      "Scardinius erythrophthalmus  .. add to records\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Sander lucioperca  .. add to records\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo salar  .. add to records\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. add to records\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "subspecies: Salmo trutta trutta\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Salvelinus fontinalis  .. add to records\n",
      "Squalius cephalus  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "subspecies: Squalius cephalus cabeda\n",
      "Squalius cephalus  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus fontinalis  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Silurus glanis  .. add to records\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "subspecies: Salmo trutta caspius\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "subspecies: Salmo trutta trutta\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus fontinalis  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus fontinalis  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Silurus glanis  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Salvelinus fontinalis  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Tinca tinca  .. add to records\n",
      "Tinca tinca  .. already covered\n",
      "Thymallus thymallus  .. add to records\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Thymallus thymallus  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Thymallus thymallus  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Umbra pygmaea  .. add to records\n",
      "Vimba vimba  .. add to records\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n",
      "Vimba vimba  .. already covered\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "Seqs = []\n",
    "taxon_to_taxid = {}\n",
    "recs_to_adjust = defaultdict(list)\n",
    "taxon_to_recs = defaultdict(list)\n",
    "\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['CytB']:\n",
    "        Seqs.append(r)\n",
    "        \n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "\n",
    "        #screen for subspecies and record taxids of valid species\n",
    "        #if valid species record the taxid\n",
    "        if (len(source.qualifiers['organism'][0].split(\" \")) == 2):\n",
    "            print source.qualifiers['organism'][0],\n",
    "            if 'db_xref' in source.qualifiers:\n",
    "#                print source.qualifiers['db_xref']\n",
    "                for t in source.qualifiers['db_xref']:\n",
    "#                    print t\n",
    "                    if 'taxon' in t:\n",
    "                        if not source.qualifiers['organism'][0] in taxon_to_taxid:\n",
    "                            print \" .. add to records\"\n",
    "                            taxon_to_taxid[source.qualifiers['organism'][0]] = t.split(\":\")[1]\n",
    "                        else:\n",
    "                            print \" .. already covered\"\n",
    "                        taxon_to_recs[source.qualifiers['organism'][0]].append(r.id)\n",
    "        #if the species status is 'sp.' but has a valid taxid\n",
    "        elif (len(source.qualifiers['organism'][0].split(\" \")) > 2 and source.qualifiers['organism'][0].split(\" \")[1] == 'sp.'):\n",
    "            print source.qualifiers['organism'][0],\n",
    "            if 'db_xref' in source.qualifiers:\n",
    "                for t in source.qualifiers['db_xref']:\n",
    "                    if 'taxon' in t:\n",
    "                        if not source.qualifiers['organism'][0] in taxon_to_taxid:\n",
    "                            print \" .. add to records\"\n",
    "                            taxon_to_taxid[source.qualifiers['organism'][0]] = t.split(\":\")[1]\n",
    "                        else:\n",
    "                            print \" .. already covered\" \n",
    "                        taxon_to_recs[source.qualifiers['organism'][0]].append(r.id)\n",
    "        #all other cases we consider dubious, e.g. supspecies\n",
    "        else:\n",
    "            print \"subspecies: %s\" %source.qualifiers['organism'][0]\n",
    "            recs_to_adjust[r.id] = []\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if any of the taxa that were considered dubious are not already covered in the list after removing subspecies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alburnoides bipunctatus strymonicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus tzanevi -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus strymonicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus strymonicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus tzanevi -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Abramis brama orientalis -> Abramis brama\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus strymonicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus strymonicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus rossicus -> Alburnoides bipunctatus\n",
      "Alburnoides bipunctatus strymonicus -> Alburnoides bipunctatus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Barbus barbus barbus -> Barbus barbus\n",
      "Cyprinus carpio carpio -> Cyprinus carpio\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus langsdorfii -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus langsdorfii -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Cyprinus carpio carpio -> Cyprinus carpio\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Cyprinus carpio 'xingguonensis' -> Cyprinus carpio\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Coregonus autumnalis autumnalis -> Coregonus autumnalis\n",
      "Cyprinus carpio 'wuyuanensis' -> Cyprinus carpio\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus langsdorfii -> Carassius auratus\n",
      "Carassius auratus langsdorfii -> Carassius auratus\n",
      "Cyprinus carpio haematopterus -> Cyprinus carpio\n",
      "Carassius auratus ssp. 'Pingxiang' -> Carassius auratus\n",
      "Carassius auratus langsdorfii -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Cyprinus carpio 'wananensis' -> Cyprinus carpio\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Chondrostoma nasus nasus -> Chondrostoma nasus\n",
      "Cyprinus carpio carpio -> Cyprinus carpio\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Micropterus salmoides salmoides -> Micropterus salmoides\n",
      "Pungitius pungitius pungitius -> Pungitius pungitius\n",
      "Phoxinus phoxinus tumensis -> Phoxinus phoxinus\n",
      "Phoxinus phoxinus tumensis -> Phoxinus phoxinus\n",
      "Salmo trutta trutta -> Salmo trutta\n",
      "Squalius cephalus cabeda -> Squalius cephalus\n",
      "Salmo trutta caspius -> Salmo trutta\n",
      "Salmo trutta trutta -> Salmo trutta\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "to_fetch = defaultdict(list)\n",
    "\n",
    "for r in Seqs:\n",
    "    if r.id in recs_to_adjust:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        adjust_from = source.qualifiers['organism'][0]\n",
    "        adjust_to = \" \".join(adjust_from.split(\" \")[:2])\n",
    "        print \"%s -> %s\" %(adjust_from,adjust_to)\n",
    "        if adjust_to in taxon_to_taxid:\n",
    "            taxon_to_recs[adjust_to].append(r.id)\n",
    "            recs_to_adjust[r.id].append(adjust_to)\n",
    "            recs_to_adjust[r.id].append(taxon_to_taxid[adjust_to])\n",
    "        else:\n",
    "            to_fetch[adjust_to].append(r.id)\n",
    "            recs_to_adjust[r.id].append(adjust_to)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print len(to_fetch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adjusting KM874614.1\n",
      "Alburnoides bipunctatus strymonicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874612.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874610.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874640.1\n",
      "Alburnoides bipunctatus tzanevi\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874618.1\n",
      "Alburnoides bipunctatus strymonicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874606.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874617.1\n",
      "Alburnoides bipunctatus strymonicus\t->\tAlburnoides bipunctatus\n",
      "adjusting HM173133.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874639.1\n",
      "Alburnoides bipunctatus tzanevi\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874608.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KC894466.1\n",
      "Abramis brama orientalis\t->\tAbramis brama\n",
      "adjusting KM874609.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874615.1\n",
      "Alburnoides bipunctatus strymonicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874613.1\n",
      "Alburnoides bipunctatus strymonicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874611.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874607.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting HM173134.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting HM173136.1\n",
      "Alburnoides bipunctatus rossicus\t->\tAlburnoides bipunctatus\n",
      "adjusting KM874616.1\n",
      "Alburnoides bipunctatus strymonicus\t->\tAlburnoides bipunctatus\n",
      "adjusting AY331019.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331018.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331020.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331021.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331017.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331022.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331023.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting AY331024.1\n",
      "Barbus barbus barbus\t->\tBarbus barbus\n",
      "adjusting JN105352.1\n",
      "Cyprinus carpio carpio\t->\tCyprinus carpio\n",
      "adjusting EF363361.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting LC097927.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852600.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852602.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting HM000036.1\n",
      "Carassius auratus langsdorfii\t->\tCarassius auratus\n",
      "adjusting LC097470.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting GU086396.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting FJ169953.1\n",
      "Carassius auratus langsdorfii\t->\tCarassius auratus\n",
      "adjusting EF363346.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting KP993139.1\n",
      "Cyprinus carpio carpio\t->\tCyprinus carpio\n",
      "adjusting EF363340.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363338.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363342.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363360.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting LC097924.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363368.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363356.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363354.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363352.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363362.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363348.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363344.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting JX183534.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting AB852597.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852591.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852593.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363366.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363364.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting AB852585.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting JX183536.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting AB852595.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting LC097926.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting JN105353.1\n",
      "Cyprinus carpio 'xingguonensis'\t->\tCyprinus carpio\n",
      "adjusting AB111951.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting AJ617503.1\n",
      "Coregonus autumnalis autumnalis\t->\tCoregonus autumnalis\n",
      "adjusting JN105357.1\n",
      "Cyprinus carpio 'wuyuanensis'\t->\tCyprinus carpio\n",
      "adjusting HQ443698.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting JN412532.1\n",
      "Carassius auratus langsdorfii\t->\tCarassius auratus\n",
      "adjusting JN412528.1\n",
      "Carassius auratus langsdorfii\t->\tCarassius auratus\n",
      "adjusting KP993138.1\n",
      "Cyprinus carpio haematopterus\t->\tCyprinus carpio\n",
      "adjusting HQ875340.1\n",
      "Carassius auratus ssp. 'Pingxiang'\t->\tCarassius auratus\n",
      "adjusting DQ399930.1\n",
      "Carassius auratus langsdorfii\t->\tCarassius auratus\n",
      "adjusting EF363339.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting LC097471.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363355.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting GU086397.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting EF363353.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting JX183535.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting JF694778.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting EF363345.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363347.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363351.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363341.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting GU086395.1\n",
      "Carassius auratus auratus\t->\tCarassius auratus\n",
      "adjusting AB852590.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363343.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting KF856964.1\n",
      "Cyprinus carpio 'wananensis'\t->\tCyprinus carpio\n",
      "adjusting EF363363.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting LC097925.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852598.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363365.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363359.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting AB852594.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852596.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting LC097923.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363358.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting EF363349.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting AB852584.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting EF363367.1\n",
      "Chondrostoma nasus nasus\t->\tChondrostoma nasus\n",
      "adjusting HQ443697.1\n",
      "Cyprinus carpio carpio\t->\tCyprinus carpio\n",
      "adjusting AB852588.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting AB852586.1\n",
      "Carassius auratus grandoculis\t->\tCarassius auratus\n",
      "adjusting HQ391896.1\n",
      "Micropterus salmoides salmoides\t->\tMicropterus salmoides\n",
      "adjusting AB094628.1\n",
      "Pungitius pungitius pungitius\t->\tPungitius pungitius\n",
      "adjusting KC992395.1\n",
      "Phoxinus phoxinus tumensis\t->\tPhoxinus phoxinus\n",
      "adjusting KC906398.1\n",
      "Phoxinus phoxinus tumensis\t->\tPhoxinus phoxinus\n",
      "adjusting EU492108.1\n",
      "Salmo trutta trutta\t->\tSalmo trutta\n",
      "adjusting AF421792.1\n",
      "Squalius cephalus cabeda\t->\tSqualius cephalus\n",
      "adjusting FJ655773.1\n",
      "Salmo trutta caspius\t->\tSalmo trutta\n",
      "adjusting AM910409.1\n",
      "Salmo trutta trutta\t->\tSalmo trutta\n"
     ]
    }
   ],
   "source": [
    "for r in Seqs:\n",
    "    if r.id in recs_to_adjust:\n",
    "        print \"adjusting %s\" %r.id\n",
    "        source = r.features[0]\n",
    "        print \"%s\\t->\\t%s\" %(source.qualifiers['organism'][0],recs_to_adjust[r.id][0])\n",
    "        source = r.features[0]\n",
    "        source.qualifiers['organism'] = [recs_to_adjust[r.id][0]]\n",
    "        for i in range(len(source.qualifiers['db_xref'])):\n",
    "            if 'taxon' in source.qualifiers['db_xref'][i]:\n",
    "                source.qualifiers['db_xref'][i] = \"taxon:\"+recs_to_adjust[r.id][1]\n",
    "        if not 'note' in source.qualifiers:\n",
    "            source.qualifiers['note'] = ['modified from original record: adjusted to species']\n",
    "        else:\n",
    "            source.qualifiers['note'].append('modified from original record: adjusted to species')\n",
    "#        print r.format('genbank')       \n",
    "#    print r,recs_to_adjust[r]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write only good records with all metadata in Genbank format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "       \n",
    "out = open('../CytB_European-fish_SATIVA_cleaned.gb','w')\n",
    "SeqIO.write(Seqs, out, \"genbank\")\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude the records that were highlighted by SATIVA. Then realign and rebuild phylogeny."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.exclude(start_from_max=False, **recs_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 1736\n",
      "Number of species (see below for full list): 65\n",
      "['Abramis brama', 'Acipenser sturio', 'Alburnoides bipunctatus', 'Alburnus alburnus', 'Alosa fallax', 'Ambloplites rupestris', 'Ameiurus melas', 'Ameiurus nebulosus', 'Anguilla anguilla', 'Aspius aspius', 'Barbatula barbatula', 'Barbus barbus', 'Blicca bjoerkna', 'Carassius auratus', 'Carassius carassius', 'Chondrostoma nasus', 'Cobitis taenia', 'Coregonus albula', 'Coregonus autumnalis', 'Coregonus lavaretus', 'Cottus gobio', 'Ctenopharyngodon idella', 'Cyprinus carpio', 'Esox lucius', 'Gasterosteus aculeatus', 'Gobio gobio', 'Gymnocephalus cernua', 'Hypophthalmichthys molitrix', 'Hypophthalmichthys nobilis', 'Lampetra fluviatilis', 'Lampetra planeri', 'Lepomis gibbosus', 'Leucaspius delineatus', 'Leuciscus idus', 'Leuciscus leuciscus', 'Lota lota', 'Micropterus salmoides', 'Misgurnus fossilis', 'Neogobius melanostomus', 'Oncorhynchus gorbuscha', 'Oncorhynchus mykiss', 'Osmerus eperlanus', 'Perca fluviatilis', 'Petromyzon marinus', 'Phoxinus phoxinus', 'Pimephales promelas', 'Platichthys flesus', 'Ponticola kessleri', 'Proterorhinus semilunaris', 'Pseudorasbora parva', 'Pungitius pungitius', 'Rhodeus sericeus', 'Rutilus rutilus', 'Salmo salar', 'Salmo trutta', 'Salvelinus alpinus', 'Salvelinus fontinalis', 'Sander lucioperca', 'Scardinius erythrophthalmus', 'Silurus glanis', 'Squalius cephalus', 'Thymallus thymallus', 'Tinca tinca', 'Umbra pygmaea', 'Vimba vimba']\n"
     ]
    }
   ],
   "source": [
    "species=[]\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['CytB']:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        species.append(\" \".join(source.qualifiers['organism'][0].split()[0:2]))\n",
    "    \n",
    "#print set(species)\n",
    "total=len(species)\n",
    "species=list(set(species))\n",
    "\n",
    "print \"Total number of records: %i\" %total\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print sorted(species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 788191473208483.95_CDS_proteins_CytB.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj,                              # The Project\n",
    "                    method_name='mafftLinsi-SATIVA', # Any unique method name,\n",
    "                    CDSAlign=True,                   # Use this method to align\n",
    "                    codontable=2,                    # The genetic code that\n",
    "                    program_name='mafft',            # mafft or muscle.\n",
    "                    cmd='mafft',                     # The command on your machine\n",
    "                    loci=['CytB'],                 # A list of loci names to align.vel.\n",
    "                    cline_args={'localpair': True,# Program specific keywords and arguments.\n",
    "                                'maxiterate': 1000}# cine_args=={} by default, which will\n",
    "                    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Align."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlnConf named mafftLinsi-SATIVA with ID 788191473208483.95\n",
      "Loci: CytB \n",
      "Created on: Wed Sep  7 00:34:43 2016\n",
      "Commands:\n",
      "CytB: mafft --localpair --maxiterate 1000 788191473208483.95_CDS_proteins_CytB.fasta\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-3.13.0-54-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.66\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.0\n",
      " User: chrishah-office\n",
      " Program and version: MAFFT v7.123b\\nPal2Nal v14\n",
      " Program reference:Katoh\n",
      " Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\\nMikita Suyama\n",
      " David Torrents\n",
      " and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34\n",
      " W609-W612.\n",
      "execution time:\n",
      "2642.52411389\n",
      "\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The dataset(s) CytB were first aligned at the protein level using the program MAFFT v7.123b [1].\n",
      "The resulting alignments served as guides to codon-align the DNA sequences using Pal2Nal v14 [2].\n",
      "\n",
      "Reference:\n",
      "[1]Katoh, Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\n",
      "[2]Mikita Suyama, David Torrents, and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34, W609-W612.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['mafftLinsi-SATIVA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'post_SATIVA.pkl'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find custom clipping points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object CytB@mafftLinsi-SATIVA\n",
      "\n",
      "Finding forward clip position:\n",
      "\n",
      "1 30 0.706221198157\n",
      "2 31 0.711693548387\n",
      "3 32 0.717165898618\n",
      "4 33 0.723387096774\n",
      "5 34 0.727476958525\n",
      "6 35 0.731566820276\n",
      "7 36 0.735656682028\n",
      "8 37 0.737960829493\n",
      "9 38 0.740264976959\n",
      "10 39 0.743260368664\n",
      "Suggested forward clip at: 29 (10,39)\n",
      "\n",
      "Finding reverse clip position:\n",
      "\n",
      "1 1036 0.702995391705\n",
      "2 1035 0.706278801843\n",
      "3 1034 0.709619815668\n",
      "4 1033 0.712960829493\n",
      "5 1032 0.716301843318\n",
      "6 1031 0.719585253456\n",
      "7 1030 0.719642857143\n",
      "8 1029 0.719700460829\n",
      "9 1028 0.720910138249\n",
      "10 1027 0.722062211982\n",
      "Suggested reverse clip at: 1037 (10,1027)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "per_col=[]\n",
    "\n",
    "aln = pj.fa('CytB@mafftLinsi-SATIVA')\n",
    "\n",
    "icol=0\n",
    "for a in aln[1,:]:\n",
    "    ACGT_count=0\n",
    "    col=aln[:,icol]\n",
    "    for i in range(len(col)):\n",
    "        if col[i] in ['A','G','C','T']:\n",
    "            ACGT_count+=1\n",
    "#    print pj.fa('CytB@mafftLinsi')[:,a]\n",
    "    per_col.append(float(ACGT_count)/len(col))\n",
    "    icol+=1\n",
    "\n",
    "wsize=10\n",
    "min_prop=0.7\n",
    "    \n",
    "print \"\\nFinding forward clip position:\\n\"\n",
    "forw_count=0\n",
    "forw_clip=0\n",
    "for i in range(len(per_col)-wsize):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        forw_count+=1\n",
    "        forw_clip=i\n",
    "        print forw_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        forw_count=0\n",
    "        \n",
    "    if forw_count >= wsize:\n",
    "        break\n",
    "        \n",
    "print \"Suggested forward clip at: %s (%s,%s)\" %(forw_clip-wsize,forw_count,forw_clip)\n",
    "\n",
    "print \"\\nFinding reverse clip position:\\n\"\n",
    "reve_count=0\n",
    "reve_clip=0\n",
    "for i in reversed(range(len(per_col)-wsize)):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        reve_count+=1\n",
    "        reve_clip=i\n",
    "        print reve_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        reve_count=0\n",
    "        \n",
    "    if reve_count >= wsize:\n",
    "        break\n",
    "    \n",
    "print \"Suggested reverse clip at: %s (%s,%s)\" %(reve_clip+wsize,reve_count,reve_clip)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object CytB@mafftLinsi-SATIVA\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fa('CytB@mafftLinsi-SATIVA')\n",
    "\n",
    "clipped_aln = aln[:,forw_clip-wsize:reve_clip+wsize]\n",
    "\n",
    "pj.trimmed_alignments[\"CytB@mafftLinsi-SATIVA_aln_clipped\"] = clipped_aln"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure tree building"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS -f a -m GTRGAMMA -n 704791473378903.3_CytB@mafftLinsi-SATIVA_aln_clipped0 -p 295 -s 704791473378903.3_CytB@mafftLinsi-SATIVA_aln_clipped.fasta -T 8 -x 188 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-mafftLinsi-SATIVA-clipped', \n",
    "                 program_name='raxmlHPC-PTHREADS',\n",
    "                 keepfiles=True,\n",
    "                 cmd='raxmlHPC-PTHREADS',\n",
    "                 preset='fa',\n",
    "                 alns=['CytB@mafftLinsi-SATIVA_aln_clipped'],\n",
    "                 model='GAMMA',\n",
    "                 threads=8,\n",
    "                 cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 255 from 'raxmlHPC-PTHREADS -f a -m GTRGAMMA -n 704791473378903.3_CytB@mafftLinsi-SATIVA_aln_clipped0 -p 295 -s 704791473378903.3_CytB@mafftLinsi-SATIVA_aln_clipped.fasta -T 8 -x 188 -N 100'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mApplicationError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-7-14dce67923aa>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mpj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mraxml\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/home/reprophylo/reprophylo.py\u001b[0m in \u001b[0;36mtree\u001b[1;34m(self, raxml_methods, bpcomp, bpcomp_burnin, bpcomp_step, bpcomp_cutoff)\u001b[0m\n\u001b[0;32m   2619\u001b[0m                 \u001b[1;32mfor\u001b[0m \u001b[0mcline\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mraxml_method\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcommand_lines\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mtrimmed_alignment\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2620\u001b[0m                     \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mraxml_method\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mRaxmlConf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2621\u001b[1;33m                         \u001b[0mstdout\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2622\u001b[0m                     \u001b[1;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mraxml_method\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mPbConf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2623\u001b[0m                         \u001b[0msub\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcline\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/Bio/Application/__init__.pyc\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[0;32m    515\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mreturn_code\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    516\u001b[0m             raise ApplicationError(return_code, str(self),\n\u001b[1;32m--> 517\u001b[1;33m                                    stdout_str, stderr_str)\n\u001b[0m\u001b[0;32m    518\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    519\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mApplicationError\u001b[0m: Non-zero return code 255 from 'raxmlHPC-PTHREADS -f a -m GTRGAMMA -n 704791473378903.3_CytB@mafftLinsi-SATIVA_aln_clipped0 -p 295 -s 704791473378903.3_CytB@mafftLinsi-SATIVA_aln_clipped.fasta -T 8 -x 188 -N 100'"
     ]
    }
   ],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named raxml-mafftLinsi-SATIVA-clipped with ID 812831473211145.19\n",
      "Alignments: CytB@mafftLinsi-SATIVA_aln_clipped \n",
      "Created on: Wed Sep  7 01:19:05 2016\n",
      "Commands:\n",
      "CytB@mafftLinsi-SATIVA_aln_clipped:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 812831473211145.19_CytB@mafftLinsi-SATIVA_aln_clipped0 -p 512 -s 812831473211145.19_CytB@mafftLinsi-SATIVA_aln_clipped.fasta -T 5 -x 748 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-3.13.0-54-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.66\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.0\n",
      " User: chrishah-office\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "28332.166764\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) CytB@mafftLinsi-SATIVA_aln_clipped using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-mafftLinsi-SATIVA-clipped']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'post_SATIVA.pkl'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Annotate final tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Lampetra','Petromyzon']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=None,\n",
    "            html='mafftLinsi-clipped-raxml-post_SATIVA.html',\n",
    "            fsize=13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'post_SATIVA.pkl'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out alignment for `pplacer` input. Need to modify record id - remove the '\\_f0' that is appended by Reprophylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning trimmed alignment object CytB@mafftLinsi-SATIVA_aln_clipped\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fta('CytB@mafftLinsi-SATIVA_aln_clipped')\n",
    "\n",
    "for r in aln:\n",
    "    r.id = \"_\".join(r.id.split(\"_\")[:-1])\n",
    "    r.description = r.id\n",
    "    \n",
    "out=open('../CytB_European-fish_SATIVA_cleaned.alignment.fasta', 'w')\n",
    "AlignIO.write(aln, out, 'fasta')\n",
    "out.close()\n",
    "\n",
    "out=open('../CytB_European-fish_SATIVA_cleaned.alignment.phylip', 'w')\n",
    "AlignIO.write(aln, out, 'phylip-relaxed')\n",
    "out.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also remove the '\\_f0' from the sequence ids in the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "sed 's/_f[0-9]//g' RAxML_bestTree.812831473211145.19_CytB@mafftLinsi-SATIVA_aln_clipped0 > RAxML_bestTree.CytB@mafftLinsi-SATIVA_aln_clipped0"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
