{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from reprophylo import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove redundancy by clustering the 12S sequences with vsearch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vsearch v1.1.0_linux_x86_64, 31.3GB RAM, 8 cores\r\n",
      "https://github.com/torognes/vsearch\r\n",
      "\r\n",
      "Reading file ../fetch_clean_align/12S_AmphibAlign.fasta 0%\r\n",
      "\r\n",
      "Fatal error: illegal character '-' on line 2 in the database file\r\n"
     ]
    }
   ],
   "source": [
    "!vsearch -threads 5 --cluster_fast ../fetch_clean_align/12S_AmphibAlign.fasta --strand both --uc 12S_id-1.uc --id 1 --query_cov 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj=Project('../fetch_clean_align/target_locus.csv', pickle='new.pkl', git=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#pj = unpickle_pj('new.pkl', git=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next code cell reads the GenBank file containing the reference sequences into memory. DO NOT RUN MORE THAN ONCE. Otherwise you must remove the pickle that has already been created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reprophylo/reprophylo.py:1015: UserWarning: Version control off\n",
      "  warnings.warn('Version control off')\n"
     ]
    }
   ],
   "source": [
    "pj.read_embl_genbank(['../fetch_and_clean/amphib12S_cropped_500bp.gb'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distribution Of Sequence Lengths\n",
      "Distribution Of Sequence Statistic \"Gc_Content\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/pymodules/python2.7/matplotlib/figure.py:371: UserWarning: matplotlib is currently using a non-GUI backend, so cannot show the figure\n",
      "  \"matplotlib is currently using a non-GUI backend, \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAFEAAAIiCAYAAACwpiRwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEsJJREFUeJztnX2UHWV9xz+bkAgJ4SVACC1gwmsItARoA1Qp4R3RipQC\nkaKUUkylFWlPGwzHU1dFi6G24EsrliLx+FKtaIHDiyUeEzyWEkEDoWGRCKsQSIJQo0Ted/vH77ln\nZ+/Ovbtz93vvnbv3+zlnzs6dmTv3yScz87zO8wNjjDHGGCPkRmAzsC6zbSZwN/Bj4L+AXTL7lgGP\nAX3AqZntR6VzPAZc18T0lpLjgCMYLnE5sDStXwFcndbnA2uBKcAcYAPQk/atARam9TuA05uW4pIy\nh+ES+4A90/rs9BniKrwic9xdwDHAXsAjme2Lgc81I6H1mNTqHxyFPYlbnPS3IvQ3gKcyxz0F/GbO\n9o1pe0spm8Qsg2mRMH369A2Zcza6rM07d9kkbiZuY4hbdUta3wjskzlub+IK3JjWs9s35p1427Zt\n+xPP0VrLh0fZ3wMcnnfuskm8FbgwrV8I/Gdm+2JgKjAXOJDIUDYBvwSOJv6R78p8pyv4KvA08Arw\nJHARUcRZSX4R50oiV+4DTstsrxRxNgCfqvN7oz0aeseQZtnjpVMZTcAiwTkmPAoBueco2zOxI7FE\nAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIs\nUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQK2a3cCykBvb+9gZr2n3rF5\n+EoUYIkZGrkKwRIlWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIA\nSxRgiQIsMcMM+N92p6Hs1Jw/cSZsHoRHZw7NrFzoHL4SBViiAEsUYIkCLFGAJQqwRAGWKMASBVii\nAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViigLJKfD8R5PXhtA4RLPZu\n8oPFLiNi3vcBp7YumeXlMELg9sBkQtz+wHJgaTrmCuDqtD6fCF49hQj/voH8i6OrxuLMA+4DXgJe\nB1YDZwNvB1akY1YA70jrZxJhjl8F+gmJC1uX3HJKfBg4jrh9pwFnEBHE92ToStmcPsPIuPdP0eK4\n92Wc0qUP+ATx3NtG3KqvVx1TiUFfi9x9CxcuvDmzvn7JkiWPAJxy1FHbvz516oypr7yy/boHHji/\ncsz1119/yJo1a+ZXPq9Zs6bov6U0fAx4LyF3dtq2V/oM8IG0VLiLCN9eTVc9EwFmpb/7An8IfIWI\ne39h2n4hQ7HtbwUWA1OBucCBQOdeMkLuIYb+rgVOSNtmAivJL+JcSWQofcBpNc7ZtCuxm+i627mj\nsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJ\nAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVY\nogBLFGCJAixRgCUKsEQBlijAEgVYooCedieghQyS8+8dHBwcNsNIT09PPSe55/CVKKCMc4W1lJ6e\nnp6ZsPk5+MVuwycsGjO+EgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijA\nEgVYogBLFGCJAhrpqNqVCPUxCDwL/J80RR3IWCROJuJAnQMcz1BgmQrPEHGk/gO4BRhQJrATqCdx\nOyIizweISDzbgPuBO4DniE7smcABRBCuxYTQvwf+hZEhkbqSRwlx/wYsIq7IWkwGTgS+kL7TV+fY\ndtGWoDb/xFBEsiLMBq5t4HvNxpGBBDgyUJlppIhzNHAWEZkM4HEictl9qkRNZCYDNxJFmLzlJupn\nPu2mFM/EDxGyvgkcA+yclmOBb6V9veP9kSZSCok/JeLl5dFDRL/96Xh/pImUImOZRdRIap38Fhor\nEnU8RSQ+xsgqX5bZRAHd1OGdRGPDgpx9R6R9i1uaomI07XYuUsQ5iCjO/IB4/j2Sts8HTgYeSsf8\nXdX3PlLgNyosAy4gMqt1wEXAdOBrwBuJYNjnAr/IHP+nRH39Mmo/u9tOraLNaEtR5hD/WW9In79G\nRNEtbbj2IlfifgWOHQ+/JEKvTyOurGnA08TVdnw6ZgWwimhhqhWu/X9alN5CEvublYgqngc+CfwM\neBH4NvH4qBeuPSusY8K1TyOeTRBlw19rkgPA/sDlxK25lWjsvaDqmI4O134ocCdx61Seea+kbYdJ\nfgHOA27IfH4X8FkiI+v4cO1HAPcCpwLfAa5Ly3fTtv9Ox4yXPqJauQNREzoZWA/cxgQI176SKFIc\nmbPvSOLWWyn6raVEuPZ1RCYyhQkSrv0F4Ko6+z9GdA2UlVLczi8RHVG1eIbITbuOIhLvIHr1avEH\n6Ziuo4jEvwZ2A75BFGZnpOVo4GbimfVX6gR2OgNEjWEgZ73WUua+5rZU+744lpSN5UcmOvUk/kmr\nEtHpuMtUQD2Ju47jvOP5bsdRT2I/0cO3W4HzzQI+CjwxjjRNKN4LbAFeJrpELwEOB3bMHDODqPL9\nBXA70TCxCVjS0pSOjbZV+3YmGkP7GV6UeTkt2W0bgL8lxJaRpkkc6+RCk4gC9vFEc/weDI2UfZho\nyXlgjOdqF7kTA0FIrEzp8nz9bt/cc4y1UXaAaD1uWZN7J+EijgBLFGCJAixRgCUKsEQBliigaOf9\nJKIL8wCiTp1XeG1kAFPXcCDR/6sewNQqSjGg6dPEoKalRDXvuQLfNYltwCfanYhxUIp+55eJcYOm\niiISvw28qVkJ6WSK9jsfC/wNMXjIJOq1Jz7ByGfAjsDuRP/y0wzvZ+5Jx7dqRG1R2tKe2MiLPe53\nrmJRqxLR6RR5Jv4+0S1Qiz3SMV1HEYmrgFPq7D+JKIR3HcoGiMl06TNRKfFY4OfC83UMo9Wd30+8\nDlG5wq4lf8jxTGAn4qXyrmM0iVsZKurMIa60LVXHDBKD1O8lZjAxdegnXgHrVErRFDanwLFdhbsH\nBBSRWBmTXW95gWj9vo54cbErKCLxi8QbTj3ENAW3puXRtG0d8V7d68D7iHeQy9oY0TZOJF5Le0vO\nvjOInPyk9PmtxIuTX25N0sZEKV5Luxe4ps7+axg+auxz1H8Dq9WUonvgcOp3DzwB/Hbm81qKDVXu\nWIpI3Er0OdfiJGLqgQo7pe9MeIpI/Cox0dr1wMFEg8MkYF7adlY6psIJRE5tMkwjBrdXOulfZfgb\n+HemYwC2J6ZFzcuE2kXbx2xnOYN4o7Qy9V8/UdQp+xumTetj6SZKkTubGhQdFfZG4kWfeqPCThxv\nojqNIhLfQswAMoWoIz+fc0xXdg8UYS2wEfiddiekQUrxTJxHdA/cX+A7XUERiT8nRoaZKoo2hZ3d\nrIR0MkUylpuIqtytRKPr4+RPnPGz8SersygiMRuU4W01jhmk3HNtN4UiEsfyVoCLOBOcUhRxTA2K\nStyJmFzj+8S828em7bsTsxrP0yWtcyjyTNyDkDcX+AkxbekOad9zxMSQu9KF84UVkXgV0dZ2DDE+\nJzsmZ5Ao+nRd4wMUu53fRgTwqjVhxuPAvuNOUQdSROLuxHOwFgNEt0DXUUTiZuI5WIsFdGFtBYpJ\nvB24mPwxNkcD76Z2qBGT2ItoT9wEfJ64fVcA/04MGemn3J31pRhGApFx3MLIGT5vA/ZW/EATKcUg\nT4hn3pnEHGIHE30sG+jyd58bjT2wlRLPqt5qXHcWUO9KHGDsPf6V49yeWIVnNx4jnt1YgJ+JAixR\ngCUKsEQBlijAEgVYooAidedGJ8m4p8HvdQxFJK5Kf/OqgrWqh11RDSwi8WLixccDiXf2slF1zydC\nwX1amroOoYjEnYi+53lEC3eWjxLv/u0EfEqTtInJY8AH6+z/IHE1lpVSjMXZh/rBDX+N+51HpZ8I\nzprXt7xD2tc//iRNbC4hGmofJgLenJCWS4kpXQaA9wh+52DgR5llKxGCfSYR5zkvluky4nHTRwSn\nzaM0vX2XE7d09YzG24hJKtVMIl4834cSh2tvhF2AcxmKq3wOw68KJacC30vrfQy9vDiboeHPywip\nFe4iBl1VU5ouU4h5IL7ewPcaYTFD71BPqHDtc4k38GcRhe5+Yo7Z2cQ/TvWuy1TileArcvZ1dLj2\n5cBrDI18qIxH3Jl4LioHeJ5J3JoV+pgA4dqXEDMbf4Z4VmXryluJ4SW1Xs1ohHcyfDqEW5kA4dof\nBL6Z1ncnrsbsyNhljKwONsp04jW4bPi6CRGu/UXgz9N6nsQ/o9zv/pXidn6JuEJqsS+Rc3cdRST+\ngJi2JY/tiWrf98edog6kiMTlwO8BX2JoJqa9gNOB1USt4h+kqZugvIe4raurfS9S/mEnpamxfJ4Y\nFftHwCFEMefHRA1GlTObElOaKzHLFOB3iXrqeqI5zOSwiOgzqZ76aS7wEMMHv3+hpSkrTtsK2zeR\n32+ympD3PeCTxFSoA5Q7c2mbxPVEXTnLPELY6sy2HYi5Zb/TyI+0iLbVWGYz8kpclP7ekNn2IvAV\nhs/k2TWMJvENhKAsC9Pf1VXbn6R5LdylZjSJTwKHVm17M/Guc/XLkNNw3TmXe4gXH38rfT6LmLHu\nzpxjD8MF7lz2I1qsXyeuvgGiueuwquMmE9HTqjOhMtG2jOVx4HhiqtPn09/jib7nLCem/X5Vd4JT\nikZZUwNLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlghc1ts768O9vQc1+n1LzPBq\ng5HSLTHDr0Z2yo0JSxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRg\niQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIF\nWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRQVom7AN8AHiGi\nEx3N+GPeN42ySryOCBZxCBFtqI+I4Xw3cBARxqkS03k+cF76ezrwz7T431VGiTsDxwE3ps+vEfGj\n3w6sSNtWAO9I62cScaBfJaKeb2AoelFLKKPEucCzRAi7HwL/SkTzrRfz/qnM9zsi5n2z2Q44EvhL\nIpLvtQwPxw7imPc33xyb161bd372+GbFvG8Fs4EnMp/fDNxOZDJNiXnf29s72NvbO5aYfB0Tj2UT\nEWCsMrPmyUSIz9soacz7Mt7OAO8DvkyI+QlwEREH6+vAxUQGcm46dn3avp7IhC5FEJu+CGWV+CAR\nbLaak2sc//G0tIUy3s4dhyUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixR\ngCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBL\nFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijA\nEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLzDDGWd9HYImmEIpZ4Dsm9kDHYYkCLFGAJQqwRAGW\nKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUUFaJ/cBDwI8YCu7a7HDtixr8Xml5gpCWZTmw\nNK1fAVyd1ucDa4EpwBwi0njexTFa90DvGNLVcd0DPVWfHa69IIPASuB+4JK0zeHaC/Im4BlgD+I5\n2Fe1v5Fw7Q+O8h2AD42y/8G8jWWV+Ez6+yzwLeL23EyEa99EhGvfko7ZCOyT+e7eaVs1C5qSUsp5\nO08DZqT16URuu44Iy17KcO1lZC6R264FHiaKLxC59UryizhXEhlKH3Bay1JqjDGdzQLgxKptFxA1\npC3A9UQGVYjJ409XR3ED8W9enT7PB+4EHkjb3k2UWFbnftsAUX48JvP5IwwvQF9MlAhMHV5ieMF8\nFXBV5vMBwK+KnrSMhe1m8ixD9ertgKOA+zL7pwIDrU5Up/El4hm4H9Gs9gKwY2b/2dSoH5sh5hI1\nmwHgNeDSqv23AP/Y6kR1IlOIok5ec9kCYLfWJscAI1uPu4Ue4DzgJGAWkcEOpu2DRCv6mClre2Kz\nWQ5cDnyXaLvMNtYqXtXoCjYD56hO1m3lxAqTiO5YCd1Wd66wK3AoUWMZN936TNwZ+GPgFGKQwKtp\neyVjuazIybo1d16VWc9mJBWJJ7Q0NR3IdsBbgd3bnZBO52ViyImEbs2dHySavSR0a+78FHANUV7c\nRtziO2SWF4ucrFszlnpthoMUvLi6sYgzlRgodSMjx/iYAmwBDlKdrFufiXsCRxIjzsZNN97OEIOm\nLiBqLA8QmQs0WGPpVonzgR+m9f0y2ysSjTHGmPbx/xk0y2HZMM1kAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fd7a54ea5d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEUAAAIiCAYAAACe4wXiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE7xJREFUeJztnXu0FdV9xz8guVzkIW9IRAMaSSAtGm3RtMugBNKkJhKb\nEhrjA00bq65qTBSNJvXWR2KRJmaRaGkMITFNFqkkqS6rEaKYpq1SH7wUpPhI4gPEByoI9yL39o/f\nzLpzD99z5sy558zMmfv7rDWLc+Z1Np87s2fP3vu3NziO4ziO4wBwEPAYcGfw/Wjgf4D1wB3A0LQT\n1D/tHxRcDDwBdAXfbwUWANOAnwOXZZSuzJgArAJOpvtK2RnZfhjweNqJyvpK+SZ2JXRG1j0OzAk+\nz8XEpEqWUj4OvITlJ/0i688FLgAeBoYAHergwYMHb8Vuud4sa9W5+6mVKfE14EzgbaAVGAasAM6K\n7DMZuA04XhzfRZn0t7W1dUU+V/o/ynNkeaVcid0ak4C/Au7DhIwJtvcHvgLcUusPxAgpS9Z5SpTw\nr3s68CSwCXgOWJZ2Qgak/YNleCBYAL4VLJmRpyslN7gUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4\nFIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSB\nSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBRBHqSUhuBOB9YE6/4X+OO0E5QHKaUhuAuBrwIf\nAP4++J4qWUuZAPw5Foschra9CBwSfB4OPJ92orKODAtDcIdF1l0B/AZYhP3RPph2orKUEg3BPSmy\n/nvARVj0+lxgKTBbnWD69OkrIp+fOO+88zYBrFhhqzds2HB6dP8lS5ZMWbNmzdTw+5o1a+rw36gv\nXwN+DzyD3TK7sXDbNyL79ANeL3N8V5n1tLW1dUXDcCsg98lbCO6ZwFZgRrDPTGBL2gnLOk+JEv7V\nPg98BxgI7Am+p0pepERDcB9GR6ynRtaP5FziUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXg\nUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIE\nLkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5FkId4n4OwGJ/ngE8Ay7GZKsGiTXdi4bipkQcpYVxy\nOC/yvMi2RfScFTcVsr59VFxySD/g08BP0k5U1lLU1MAhJwLbgadSTRH5jEsO+Qzw40on6CtxyT8M\ntg0AtgHvqnB8n4lLDudKnoXNgvtCFgnLOk+JEv2rzSODDDYkD49k6BmXDHBOVgmBfF0pucGlCFyK\nwKUIXIrApQhcisClCFyKwKUICi1lJGwfCo8nPa60tquZ6KJM+kfC9leCasxRMPxVGJfkHIW+UmrF\npQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUI\nXIrApQhcisClCFyKwKUIXIrApQhcisClCPIgpXRqYIC/w4KgNgL/mHaC8hDvUxqCezJwKjAN2AeM\nSTtBWV8pKgT3fODrmBCAHWknKmspKgT3KOBDwIPAauCP0k5UHkNwBwAjgBOwqcZ/ChyhTlAuBHf2\ncce17m9pGQrQ0tHRuuGRR06H5g3BvQ24m+5ZcMFmxR0lji8bYjsStnfBk13w5EgL+C5HNWG6mTGD\n7qfPecA/BJ8nA78rc0zDpOTh6RMSJnBpsGwAOuiOVU4N74bu3dCrw6UIXIrApQhcisClCFyKwKUI\nXIrApQhcisClCFyKwKUIXIrApQiS1rwdDLwXGItV0OwAngTeqnO6MqUaKSOB+cBc4DhxzD7gEeDf\ngGXAa/VLXjZUqo4cDnwVuAAYiF0RD2JDob4SHDsSeA/WHDEZ2AvcDFwLvN6wVBsNq46sxMvA85iY\nSVXsfyRwNTYI5stJfqhGMmniuBi7QpLSCnyhhuOS0jApXpvvtfnV0Rspc4F7sDGP7gXOrEuKckCt\nUi4EfoCVU+7AWvKWAZfXJ1n5ppy0jdgQzFG+CTzb0NT0pGEZbdyV8ihWYCtlMDYjQpTngSEx52sK\n4qT8Gvhv4J+AQZH1d2K3y7nAbOBLwFfo2W+t0BwPrAeeBv4sWDcIuAV75+nESrLfp7vfWhpk3j/l\nHcBVmIQfAaOD9f2xl8ODenPyGsksTwnZB1wPHIN13tuM9RvpxLpo7a/yPIWlH/DXwKtY+aSa96JG\nkOmV8hfYLfMz4FLsVroVmIoVpTdiPRz7TOn4IuwWWQ38HMtT/qNkn09gHfoeBY5NMW2ZZbRPAd+O\nfD8Fk1TahXMosBhor+VHaiSz22c0dnuEhGO7jizZ702sP/0M+gD3YDVuHwL+EOvou5N8lFzlX7lL\nkPQccRyJFdw6g+VlYE4tJ2oADZNSTSVTf6y//CCsfLK3imPSoGGVTNXU5ndit1CfoVJGO6IX5+3N\nsZlTScqzWO28ChYox1iseeOZXqQp15yPvde0YwW3vwGOpueTZyhWYLsQuAt7R9qGBR00mswKb4cA\nX8aums7I0h4s0XVbseJ+WtUHmTdx9AemY4WzqVhcX9iWvBG4H2s6TZNMnz5gV8KDwVJ48vBmWxqC\n24bV/z4WLB9NO0F5CIIqDcHtAr4RLJmQ9ZWiQnD7kXFzbtZSVAhuF/bGvQ74HtYlJFXyGIJ7C3BN\n8PlarHnlc+oEfSUE94cl+0zEAiwVmTdxgLXrHF9h+3QsSrQWoiG474ysv4TycyZn3sQBcDZWv1KO\nI7C+cbXQj+4ELsTqcNZhsi6p8Zw1U888ZTDdgzYkZXWwQA66dMRJeXewhI/IKVjVZCmjsBfIrfVL\nWn5po+dLX6XlbdKNNs8srP8XdPc5WQr8Cwe+/3QBu4A12NOk6YmTsjZYwB6PKyj/iCwMSTLatkYl\nIm8kffr0B2ZhvaxHod9RrhHrCstR2NtsXIabFrkYP2UxVkBbgNW0vZLg2KYiiZQTgW8BixqUltyQ\npJjfjvV7KzxJpPwS+NNGJSRPJJHyReCDWG+mlsYkJx8kqfZ7BnvpG411/HuBnh0AwzddOSZbA8i8\niQPgt5USEvmRpieJlJMalYi8kXXFdS6pRcoMrKPxd4H3BeuGYPUsTd0FoxYOwvq8hcX5/cDMYNsg\nrLPxVSmmJxd1tJdjHY2/iNXARTPcPVh3jY8lOF9uSSLlLGwU0JvQ7z2bsbfnpieJlIlY7E85dlKQ\nPCWJlDc5sFNxlCPJYOjlRpBEym+AM8ocMwKLEru/HonKmiRSrsfGM7gPawcGi//5W6w9eAhwQ11T\n1yScgnX0K61t2wZ8JOW05KLmDawH5EQsmDJ8LG/BqhUKNYZKtRyODSpTjoODfdIiF4W3Z4FPVth+\nKgXpVFzPF8LCvFzW8z/yPoKKnWYnrubt7GABq0/ZhL5HRwF/gL3//GW9EhdD2Qqvtra2LoCr29q2\nNKLmbQQ9qxfHYFWSpSfehXXaS/MtOZZRMHxfDUMhxUm5KVjAyiOXAP+a9EeyosIVUpEk5ZTCZKRx\n9Jn/aBKSSvkMVn2wg57F/P2Rf5ueJLfPZdj8XS9jvZlURVOfa+K4EHgIq5fd05jk5IMkt894rDqy\n0EIgmZSnyCB4IAuSSFmEjZuS5hBDmZAkT+nEivibsH76T6OfNqVBB01HEinfj3wuV5zvoo9JmRm/\nSzFIImV1oxKRN3pTzB9N99BmhSKplEOxPON1LNTtJWxM6x8E22pBzYILNspgJ5Ub4BpCktvncKxE\nOw4LUAqHIpqKxejMxiLHkgYtlIbgAhwWnO+3Cc9VF5JcKddihbePAx/AWgvPwAaAOAX7i16X8PdV\nCC5YTPKChOeqG0mkfASLBC0d0gxsPtKb6R5bslpUCO4cLIJ9fcJz1Y0kt88IrOGrHFtJ1utAheAe\nDFyJ3TohZeuRy4Xgrlhhqzds2HB6dP9GhOA+BSyvsP0nwT7VokJwb8dKzc8Eyz6svWmsOL5sNUVb\nW1tXWHkdQ68bw36KjWt9AzauSsgh2EzY86gsrZQrsQx1EjYK8n1YS8C4YN0k7DY6FruiUiPJ7XMd\nFrSwAHtcvhCsPxST+18kz2ijqL9aJpVWSaTsxiaNnw+cRvfoor/E2nuWYcGVtfBAsJSSVu/tHiTt\ndbAP6yr63QakJTd4bb7ApQhcisClCFyKwKUIqpFyPvDpmH3mkc7QiKkQJ+U04DvE91B6DXuDzssA\nvr0iTspnsXbje2P2uxcr5p8ds19TECfleHT9ieIebFympidOylgOnLKmHC+gX/GbjjgpbwHDqjzX\nMArS6zpOylaqHwv/ROD/epecfBAn5U7sifInMfudgPXGLsSkR3FSFmNdue4CPs+BkzW2YsM8343V\nji2udwKzIE7Ka9iV0gH8M1ZeWYtNm/VYsH1JsH0OBZjsFaqrZFqDDfh9GfApYFpk2++wwasWUjla\noqmotuZtG1Yv+yWsJW8Y8AYWV1g4ahkm8U0KKiOkGinhqJ8vBv+2Yj0lS2vaf49NJN30xEl5L9aQ\nfhXWhxYsgPJGse/bWCbc9GWVuKfPOdgYBjeJbZdivZtmAh/GIjnkCMPNRtyVMhOb0FVNe7WWnr2b\nllOQLmBxV8pRWHmkGjYH+zc9cVfKYOy2iPIqVlYpHd7sDQ4MkGpK4qTspOeY02B9STaKfcdj3b6a\nnrjbZyPVR6bPpiDDssZJuR3rUBNX9/rJYL/be5+k7ImTshTLQJdjfd7eXbJ9Itb9YjnWPb3WIeJz\nRVyeshfrhnUXVoC7EstQ38Def8LOO5uD/fIy81yvqKbd52msN9HF2BgqncC7gn//E5v/9FgKEtIP\n1b8Q7sEqkApRiRSHN5sKXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUI8iClNAT3WiyadS3w\nKyxQKlXyICUMwY3OQ3g0NqzrL7B54FMlaykqBDfaeD+EGsZp6y1Zz5cchuCWdja8HgsAfwvrOpYq\neZwFF6w++CrgCkzcOeoEeQjBrTfVzIJ7OLrhDXISgltvVAjuWfRsj55D9W3ZdSPrPCUkOgvu17F+\nMfux4O/z005MXqSsprtbR1pDt5Yl60dyLnEpApcicCkClyJwKQKXInApApcicCkClyJwKQKXInAp\nApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKX\nInApApcicCkClyJwKQKXInApApcicCkClyLIi5TS2OQbscH01gE/o+dkkA0nL1JKY5PvBd6PxSdv\nAb6cZmLyIEXFJq+ke2bch4J9UiMPUtT0wFHOpfrpdOpC1jGElWKTwWKTO4Afq4OLGJcMlWOT52Oz\nS7WWObaQcclQPjb5o9gtNYcMRi7NWkqUaGzyYmzwh5XYrXVzmgnJOk+Jspru2ORMR1XP05WSG1yK\nwKUIXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhcisCl\nCFyKwKUIXIqg0FKq7KNyAIWW0hep6Sqo5hx+pQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrA\npQhcisClCPIipTQEdy7wODYX4bFpJyYvUkpDcDcApwG/ziIxeZCiQnA3Y/HImZAHKXEhuKmTdbhc\nXAhuRcqF4JajCCG4APdTPqNtWMV1nphB99Mn5H7guDL795na/DCRp2FX0AnAXcDdmaWoyegzV0ou\ncCkClyJwKQKXInApApciyPrdp+5Ma50weerA8bMG9x/YsruzveOJ9m2r1u99LtEbd6GkTGudMHn2\nkCnzLx0zuyNct2jHyvnAsiRiCnX7TB04flZUCMClY2Z3TBk4flaS8xRKyuD+A1vU+iFl1pejUFJ2\nd7Z3qPW7yqwvR6GkPNG+bdWiHSt7XBWLdqxs2dS+bVWS8/SL3yW3dCHSP611wuQpA8fPGtJ/YMuu\nzvaOTZWfPvIczYxXHaSJSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsR\nuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBRBHqSUht+OxKba24JN/Dq8xvOeVGuC8iCl\nNPz2CkzKZOBXwfdaOKnXKcuICcAq4GS6r5TNwLjg8/jguyKuJ1NbFb+fy55MKvx2HLA9+LydbkGp\nkaWUaPhtuc54XZS/ItZFtqvl6pjtXcE5coUKv70Nu13GB/u8k/K3T+GJht8uBC4PPl8B3JBJinLA\nDOCO4PNILPPt7SPZcRwnKccAM0vWnYE90V4ClgCJwlrA3juamVux/8MDwfep2LgIjwTrzsLKYg/I\nowvK89ggESHX0LNA9jlgY6opygF7gcMi31cD10W+vwd4M+lJs3736S07gEODzwOwsVYeimxvIUfD\nGqXFj7A85AhgAbALGBLZ/ily+H7TaCYBW7Gr4W3ggpLt/w58I+1E5YF3YI/mQ8W2Y4BR6SanoBQl\nqLAfMA/4MDAWe4CEQZNdwKlJTlaUsP6FwBew0b5epGfFVO6HLWsU27FBOOtCs5dTQvpj1Zp1odnf\nfUJGAO/HSrS9pih5yiHAZ4HZwHpgX7A+zGgvSnKyojx9Vkc+RzPWUMrJqaYmBwwATgFGZ52QvNEO\nTKzXyYry9FmHVRPUhaI8fZ4DbsTKK7uxW2pQZNmT5GRFyWgr1Zl0kfCPX4RHcgvwMLCUPtzEqngJ\n689SF4qSp4zDxsJeWY+TFeH2ATgYa++ZjTVv7A7W11SiLYqUqcCjwecjIutDKY7jOE6e+H95Bx4n\nvj8sKQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fd7a53f1210>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "pj.report_seq_stats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 90\n",
      "Number of species (see below for full list): 18\n",
      "['Alytes obstetricans', 'Pelophylax ridibundus', 'Triturus cristatus', 'Epidalea calamita', 'Rana catesbeiana', 'Salamandra salamandra', 'Discoglossus pictus', 'Hyla arborea', 'Bufo bufo', 'Xenopus laevis', 'Pelophylax esculentus', 'Lissotriton vulgaris', 'Bombina bombina', 'Pelophylax lessonae', 'Triturus carnifex', 'Rana temporaria', 'Ichthyosaura alpestris', 'Bombina variegata']\n"
     ]
    }
   ],
   "source": [
    "species=[]\n",
    "for r in pj.records:\n",
    "    source = [f for f in r.features if f.type == 'source'][0]\n",
    "    species.append(\" \".join(source.qualifiers['organism'][0].split()[0:2]))\n",
    "    \n",
    "#print set(species)\n",
    "total=len(species)\n",
    "species=list(set(species))\n",
    "\n",
    "print \"Total number of records: %i\" %total\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract ids of representative sequences identified by vsearch -> remove all others from the object.\n",
    "\n",
    "NB: 'rrnS' must be used as this is the default loci name for 12S rRNA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 30\n",
      "to keep: 60\n",
      "TOTAL: 90\n"
     ]
    }
   ],
   "source": [
    "ids = []\n",
    "recs_to_keep = {}\n",
    "recs_to_drop = {}\n",
    "\n",
    "\n",
    "uc=open('12S_500bp_id-1.uc', 'r')\n",
    "\n",
    "for line in uc:\n",
    "    if line.startswith('C'):\n",
    "        ids.append(line.split(\"\\t\")[8])\n",
    "        \n",
    "recs_to_keep = {'rrnS': ids}\n",
    "recs_to_drop = {'rrnS': []}\n",
    "\n",
    "for r in pj.records:\n",
    "    if not r.id in recs_to_keep['rrnS']:\n",
    "        recs_to_drop['rrnS'].append(r.id)\n",
    "        \n",
    "print \"to drop: %s\" %len(recs_to_drop['rrnS'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['rrnS'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['rrnS']) + len(recs_to_keep['rrnS']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude redundant records from object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.exclude(start_from_max=False, **recs_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 60\n",
      "Number of species (see below for full list): 18\n",
      "['Alytes obstetricans', 'Pelophylax ridibundus', 'Pelophylax lessonae', 'Epidalea calamita', 'Rana catesbeiana', 'Salamandra salamandra', 'Triturus cristatus', 'Discoglossus pictus', 'Hyla arborea', 'Bufo bufo', 'Xenopus laevis', 'Pelophylax esculentus', 'Lissotriton vulgaris', 'Bombina bombina', 'Triturus carnifex', 'Rana temporaria', 'Ichthyosaura alpestris', 'Bombina variegata']\n"
     ]
    }
   ],
   "source": [
    "species=[]\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['rrnS']:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        species.append(\" \".join(source.qualifiers['organism'][0].split()[0:2]))\n",
    "    \n",
    "#print set(species)\n",
    "total=len(species)\n",
    "species=list(set(species))\n",
    "\n",
    "print \"Total number of records: %i\" %total\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write full Genbank records to file for later processing with SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seqs = []\n",
    "from Bio import SeqIO\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['rrnS']:\n",
    "        seqs.append(r)\n",
    "\n",
    "out = open('12S_500_pre_Sativa.gb','w')\n",
    "SeqIO.write(seqs, out, 'genbank')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure Alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "muscle -in 832571474456826.89_rrnS.fasta\n"
     ]
    }
   ],
   "source": [
    "rRNA = [\"rrnS\"]\n",
    "\n",
    "muscle = AlnConf(pj, loci=rRNA,\n",
    "                 method_name='muscleDefault', \n",
    "                 program_name='muscle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.align([muscle])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@muscleDefault']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.alignments.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlnConf named muscleDefault with ID 832571474456826.89\n",
      "Loci: rrnS \n",
      "Created on: Wed Sep 21 11:20:26 2016\n",
      "Commands:\n",
      "rrnS: muscle -in 832571474456826.89_rrnS.fasta\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: MUSCLE v3.8.31\\nPal2Nal v14\n",
      " Program reference:\\nEdgar 2004: MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Research 32(5):1792-1797\\nMikita Suyama\n",
      " David Torrents\n",
      " and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34\n",
      " W609-W612.\n",
      "execution time:\n",
      "7.93936300278\n",
      "\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The dataset(s) rrnS were first aligned at the protein level using the program MUSCLE v3.8.31 [1].\n",
      "The resulting alignments served as guides to codon-align the DNA sequences using Pal2Nal v14 [2].\n",
      "\n",
      "Reference:\n",
      "[1]Edgar 2004: MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Research 32(5):1792-1797\n",
      "[2]Mikita Suyama, David Torrents, and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34, W609-W612.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['muscleDefault']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'500.pkl'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, '500.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@muscleDefault_aln.fasta']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_alns(id=['record_id','source_organism'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alignment needs fair amount of trimming and several gaps within sequences. The below cell contains a custom solution for start/end-clipping the alignment.\n",
    "\n",
    "We only want to trim the alignment from the end because this is the alignment that should be used for phylogenetic placement and it needs to be possible to align reads to the alignment.\n",
    "\n",
    "Clipping positions are determined from start/end of the alignment by assessing the average per column completeness across 10 bp sliding windows. per column completeness is the proportion of sequences represented by ACGT in a given column. Average completeness is assessed across 10bp sliding windows.\n",
    "\n",
    "The clipping point is defined as the position at which the alignment, starting from the 5'/3'-ends, respectively, retained an average completeness > 0.7 (i.e. 70% of the sequences are informative) for 10 consecutive sliding windows.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object rrnS@muscleDefault\n",
      "\n",
      "Finding forward clip position:\n",
      "\n",
      "1 145 0.716666666667\n",
      "2 146 0.733333333333\n",
      "3 147 0.75\n",
      "4 148 0.766666666667\n",
      "5 149 0.783333333333\n",
      "6 150 0.8\n",
      "7 151 0.816666666667\n",
      "8 152 0.816666666667\n",
      "9 153 0.8\n",
      "10 154 0.783333333333\n",
      "Suggested forward clip at: 144 (10,154)\n",
      "\n",
      "Finding reverse clip position:\n",
      "\n",
      "1 982 0.711666666667\n",
      "2 981 0.778333333333\n",
      "3 980 0.84\n",
      "4 979 0.901666666667\n",
      "5 978 0.93\n",
      "6 977 0.956666666667\n",
      "7 976 0.966666666667\n",
      "8 975 0.966666666667\n",
      "9 974 0.966666666667\n",
      "10 973 0.966666666667\n",
      "Suggested reverse clip at: 983 (10,973)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "per_col=[]\n",
    "\n",
    "aln = pj.fa('rrnS@muscleDefault')\n",
    "\n",
    "\n",
    "icol=0\n",
    "for a in aln[1,:]:\n",
    "    ACGT_count=0\n",
    "    col=aln[:,icol]\n",
    "    for i in range(len(col)):\n",
    "        if col[i] in ['A','G','C','T']:\n",
    "            ACGT_count+=1\n",
    "#    print pj.fa('rrnS@muscle_defaults)[:,a]\n",
    "    per_col.append(float(ACGT_count)/len(col))\n",
    "    icol+=1\n",
    "\n",
    "wsize=10\n",
    "min_prop=0.7\n",
    "    \n",
    "print \"\\nFinding forward clip position:\\n\"\n",
    "forw_count=0\n",
    "forw_clip=0\n",
    "for i in range(len(per_col)-wsize):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        forw_count+=1\n",
    "        forw_clip=i\n",
    "        print forw_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        forw_count=0\n",
    "        \n",
    "    if forw_count >= wsize:\n",
    "        break\n",
    "        \n",
    "print \"Suggested forward clip at: %s (%s,%s)\" %(forw_clip-wsize,forw_count,forw_clip)\n",
    "\n",
    "print \"\\nFinding reverse clip position:\\n\"\n",
    "reve_count=0\n",
    "reve_clip=0\n",
    "for i in reversed(range(len(per_col)-wsize)):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        reve_count+=1\n",
    "        reve_clip=i\n",
    "        print reve_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        reve_count=0\n",
    "        \n",
    "    if reve_count >= wsize:\n",
    "        break\n",
    "    \n",
    "print \"Suggested reverse clip at: %s (%s,%s)\" %(reve_clip+wsize,reve_count,reve_clip)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clip the alignment at the positions determined above and write to file. We'll use Biopython functions to do this.\n",
    "\n",
    "Clip alignment - the new Biopython alignment object is called clipped_aln."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object rrnS@muscleDefault\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fa('rrnS@muscleDefault')\n",
    "\n",
    "clipped_aln = aln[:,forw_clip-wsize:reve_clip+wsize]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the clipped alignment (which is a Biopython alignment object) to a ReproPhylo trimmed alignment object, so that we can resume working with it in ReproPhylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.trimmed_alignments[\"rrnS@muscleDefault_aln_clipped_500bp\"] = clipped_aln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@muscleDefault_aln_clipped_500bp': <<class 'Bio.Align.MultipleSeqAlignment'> instance (60 records of length 839, SingleLetterAlphabet()) at 7fd7a4f57590>}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.trimmed_alignments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write clipped alignment to file for downstream taxonomic curation using SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "out=open('rrnS@muscleDefault_aln_clipped_500bp.phy', 'w')\n",
    "AlignIO.write(clipped_aln, out, 'phylip-relaxed')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to run SATIVA is described in a different notebook. It may be a good strategy to set this going now in a separate notebook. Then contnue onwards in this notebook.\n",
    "\n",
    "Infer Maximum Likelihood tree based on the clipped alignment. We are going to use RAxML.\n",
    "\n",
    "Configure RAxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 398211474456904.99_rrnS@muscleDefault_aln_clipped_500bp0 -p 883 -s 398211474456904.99_rrnS@muscleDefault_aln_clipped_500bp.fasta -T 5 -x 696 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-muscleDefault-clipped',\n",
    "                  program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                  keepfiles=True,\n",
    "                  cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                  preset='fa',\n",
    "                  alns=['rrnS@muscleDefault_aln_clipped_500bp'],\n",
    "                  model='GAMMA',\n",
    "                  threads=5,\n",
    "                  cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named raxml-muscleDefault-clipped with ID 398211474456904.99\n",
      "Alignments: rrnS@muscleDefault_aln_clipped_500bp \n",
      "Created on: Wed Sep 21 11:21:44 2016\n",
      "Commands:\n",
      "rrnS@muscleDefault_aln_clipped_500bp:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 398211474456904.99_rrnS@muscleDefault_aln_clipped_500bp0 -p 883 -s 398211474456904.99_rrnS@muscleDefault_aln_clipped_500bp.fasta -T 5 -x 696 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "113.423015833\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) rrnS@muscleDefault_aln_clipped_500bp using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-muscleDefault-clipped']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'500.pkl'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, '500.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize the tree.\n",
    "\n",
    "Prepare tree annotation:\n",
    "\n",
    " - define outgroups\n",
    " - specify support coloration\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Xenopus','Hyla']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='muscleDefault-clipped-500bp-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check tree out. Have a think which sequences you would consider mislabelled and should better be removed from the reference database.\n",
    "\n",
    "Create a static copy of the tree (the last *.png with the expected alignment id) for later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *rrnS@muscleDefault_aln_clipped_500bp.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png rrnS_muscleDefault_clipped_500bp_raxml.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now compare the Muscle alignment with Mafft alignment.\n",
    "\n",
    "Configure Mafft alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 267641474457297.81_rrnS.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj,                              # The Project\n",
    "                      method_name='mafftLinsi',        # Any unique method name,\n",
    "                      CDSAlign=False,                   # Specify as false as 12S rRNA is not recognised as a CDS\n",
    "                      codontable=0,                    # The genetic code that\n",
    "                      program_name='mafft',            # mafft or muscle.\n",
    "                      cmd='mafft',                     # The command on your machine\n",
    "                      loci=['rrnS'],                   # A list of loci names to align.vel.\n",
    "                      cline_args={'localpair': True,   # Program specific keywords and arguments.\n",
    "                                  'maxiterate': 1000}  # cine_args=={} by default, which will\n",
    "                     )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlnConf named mafftLinsi with ID 267641474457297.81\n",
      "Loci: rrnS \n",
      "Created on: Wed Sep 21 11:28:17 2016\n",
      "Commands:\n",
      "rrnS: mafft --localpair --maxiterate 1000 267641474457297.81_rrnS.fasta\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: MAFFT v7.123b\n",
      " Program reference:Katoh\n",
      " Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\n",
      "execution time:\n",
      "12.1976089478\n",
      "\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The dataset(s) rrnS were aligned using the program MAFFT v7.123b [1].\n",
      "\n",
      "Reference:\n",
      "Katoh, Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['mafftLinsi']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'500.pkl'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, '500.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@mafftLinsi_aln.fasta', 'rrnS@muscleDefault_aln.fasta']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_alns(id=['record_id','source_organism'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start/end-clip the alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object rrnS@mafftLinsi\n",
      "\n",
      "Finding forward clip position:\n",
      "\n",
      "Suggested forward clip at: -10 (0,0)\n",
      "\n",
      "Finding reverse clip position:\n",
      "\n",
      "Suggested reverse clip at: 10 (0,0)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "per_col=[]\n",
    "\n",
    "aln = pj.fa('rrnS@mafftLinsi')\n",
    "\n",
    "\n",
    "icol=0\n",
    "for a in aln[1,:]:\n",
    "    ACGT_count=0\n",
    "    col=aln[:,icol]\n",
    "    for i in range(len(col)):\n",
    "        if col[i] in ['A','G','C','T']:\n",
    "            ACGT_count+=1\n",
    "#    print pj.fa('rrnS@mafftDefault')[:,a]\n",
    "    per_col.append(float(ACGT_count)/len(col))\n",
    "    icol+=1\n",
    "\n",
    "wsize=10\n",
    "min_prop=0.7\n",
    "    \n",
    "print \"\\nFinding forward clip position:\\n\"\n",
    "forw_count=0\n",
    "forw_clip=0\n",
    "for i in range(len(per_col)-wsize):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        forw_count+=1\n",
    "        forw_clip=i\n",
    "        print forw_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        forw_count=0\n",
    "        \n",
    "    if forw_count >= wsize:\n",
    "        break\n",
    "        \n",
    "print \"Suggested forward clip at: %s (%s,%s)\" %(forw_clip-wsize,forw_count,forw_clip)\n",
    "\n",
    "print \"\\nFinding reverse clip position:\\n\"\n",
    "reve_count=0\n",
    "reve_clip=0\n",
    "for i in reversed(range(len(per_col)-wsize)):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        reve_count+=1\n",
    "        reve_clip=i\n",
    "        print reve_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        reve_count=0\n",
    "        \n",
    "    if reve_count >= wsize:\n",
    "        break\n",
    "    \n",
    "print \"Suggested reverse clip at: %s (%s,%s)\" %(reve_clip+wsize,reve_count,reve_clip)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object rrnS@mafftLinsi\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fa('rrnS@mafftLinsi')\n",
    "\n",
    "clipped_aln = aln[:,forw_clip-wsize:reve_clip+wsize]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the clipped alignment (which is a Biopython alignment object) to a ReproPhylo trimmed alignment object, so that we can resume working with it in ReproPhylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.trimmed_alignments[\"rrnS@mafftLinsi_aln_clipped_500bp\"] = clipped_aln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@mafftLinsi_aln_clipped_500bp': <<class 'Bio.Align.MultipleSeqAlignment'> instance (60 records of length 0, SingleLetterAlphabet()) at 7fd7a5250850>,\n",
       " 'rrnS@muscleDefault_aln_clipped_500bp': <<class 'Bio.Align.MultipleSeqAlignment'> instance (60 records of length 839, SingleLetterAlphabet()) at 7fd7a4f57590>}"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.trimmed_alignments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write clipped alignment to file for downstream taxonomic curation using SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Non-empty sequences are required",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-42-a79417a76d61>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mout\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'rrnS@mafftLinsi_aln_clipped_500bp.phy'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'w'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0mAlignIO\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mclipped_aln\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mout\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'phylip-relaxed'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[0mout\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/Bio/AlignIO/__init__.pyc\u001b[0m in \u001b[0;36mwrite\u001b[1;34m(alignments, handle, format)\u001b[0m\n\u001b[0;32m    212\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[1;32min\u001b[0m \u001b[0m_FormatToWriter\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    213\u001b[0m             \u001b[0mwriter_class\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_FormatToWriter\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 214\u001b[1;33m             \u001b[0mcount\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mwriter_class\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfp\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite_file\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malignments\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    215\u001b[0m         \u001b[1;32melif\u001b[0m \u001b[0mformat\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_FormatToWriter\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    216\u001b[0m             \u001b[1;31m# Exploit the existing SeqIO parser to do the dirty work!\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/Bio/AlignIO/Interfaces.pyc\u001b[0m in \u001b[0;36mwrite_file\u001b[1;34m(self, alignments)\u001b[0m\n\u001b[0;32m    129\u001b[0m         \u001b[0mcount\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    130\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0malignment\u001b[0m \u001b[1;32min\u001b[0m \u001b[0malignments\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 131\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite_alignment\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    132\u001b[0m             \u001b[0mcount\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    133\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite_footer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/Bio/AlignIO/PhylipIO.pyc\u001b[0m in \u001b[0;36mwrite_alignment\u001b[1;34m(self, alignment)\u001b[0m\n\u001b[0;32m    293\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    294\u001b[0m             \u001b[0mid_width\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mid\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[1;32min\u001b[0m \u001b[0malignment\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 295\u001b[1;33m         \u001b[0msuper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mRelaxedPhylipWriter\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite_alignment\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mid_width\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    296\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    297\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/Bio/AlignIO/PhylipIO.pyc\u001b[0m in \u001b[0;36mwrite_alignment\u001b[1;34m(self, alignment, id_width)\u001b[0m\n\u001b[0;32m     74\u001b[0m                 \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Sequences must all be the same length\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     75\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlength_of_seqs\u001b[0m \u001b[1;33m<=\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 76\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Non-empty sequences are required\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     77\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     78\u001b[0m         \u001b[1;31m# Check for repeated identifiers...\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Non-empty sequences are required"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "out=open('rrnS@mafftLinsi_aln_clipped_500bp.phy', 'w')\n",
    "AlignIO.write(clipped_aln, out, 'phylip-relaxed')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer Maximum Likelihood tree based on the clipped alignment. We are going to use RAxML.\n",
    "\n",
    "Configure RAxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 257301474389474.4_rrnS@mafftLinsi_aln_clipped0 -p 291 -s 257301474389474.4_rrnS@mafftLinsi_aln_clipped.fasta -T 5 -x 655 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-mafftLinsi-clipped', \n",
    "                  program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                  keepfiles=True,\n",
    "                  cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                  preset='fa',\n",
    "                  alns=['rrnS@mafftLinsi_aln_clipped'],\n",
    "                  model='GAMMA',\n",
    "                  threads=5,\n",
    "                  cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run RAxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 255 from 'raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 257301474389474.4_rrnS@mafftLinsi_aln_clipped0 -p 291 -s 257301474389474.4_rrnS@mafftLinsi_aln_clipped.fasta -T 5 -x 655 -N 100'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mApplicationError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-79-14dce67923aa>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mpj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mraxml\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/home/reprophylo/reprophylo.pyc\u001b[0m in \u001b[0;36mtree\u001b[1;34m(self, raxml_methods, bpcomp, bpcomp_burnin, bpcomp_step, bpcomp_cutoff)\u001b[0m\n\u001b[0;32m   2629\u001b[0m                 \u001b[1;32mfor\u001b[0m \u001b[0mcline\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mraxml_method\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcommand_lines\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mtrimmed_alignment\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2630\u001b[0m                     \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mraxml_method\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mRaxmlConf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2631\u001b[1;33m                         \u001b[0mstdout\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2632\u001b[0m                     \u001b[1;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mraxml_method\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mPbConf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2633\u001b[0m                         \u001b[0msub\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcline\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/Bio/Application/__init__.pyc\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[0;32m    514\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mreturn_code\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    515\u001b[0m             raise ApplicationError(return_code, str(self),\n\u001b[1;32m--> 516\u001b[1;33m                                    stdout_str, stderr_str)\n\u001b[0m\u001b[0;32m    517\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    518\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mApplicationError\u001b[0m: Non-zero return code 255 from 'raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 257301474389474.4_rrnS@mafftLinsi_aln_clipped0 -p 291 -s 257301474389474.4_rrnS@mafftLinsi_aln_clipped.fasta -T 5 -x 655 -N 100'"
     ]
    }
   ],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'raxml-mafftLinsi-clipped'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-80-f1117e8ea97d>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mprint\u001b[0m \u001b[0mpj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mused_methods\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'raxml-mafftLinsi-clipped'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m: 'raxml-mafftLinsi-clipped'"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-mafftLinsi-clipped']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pickle_pj(pj, 'nr.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize the tree constructed from the mafft alignment.\n",
    "\n",
    "Prepare tree annotation:\n",
    "\n",
    " -   define outgroups\n",
    " -   specify support coloration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Xenopus','Hyla']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='mafftLinsi-clipped-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check tree out and compare to the tree constructed from the muscle alignment.\n",
    "\n",
    "Create a static copy of the tree (the last *.png with the expected alignment id for later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png rrnS_mafftLinsi_clipped_raxml.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the meantime we have run SATIVA on the alignment that was the basis of the above tree.\n",
    "\n",
    "Now we are going to highlight the records that SATIVA considers mislabelled in the tree. We are also going to add a heatmap to the tree illustrating the confidence that SATIVA has in a given call.\n",
    "\n",
    "Parse SATIVA output and record putatively mislabelled sequence records and the confidence of the call.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rec_mislabelled = {}\n",
    "sativa=open('../3-SATIVA/CytB.mis', 'r')\n",
    "\n",
    "for line in sativa:\n",
    "    if not line.startswith(\";\"):\n",
    "        rec_mislabelled[line.split(\"\\t\")[0]] = line.split(\"\\t\")[4]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add two new qualifiers containing info about he SATIVA results to the sequence features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for r in pj.records:\n",
    "    if r.id in rec_mislabelled:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['mislabeled'] = ['yes']\n",
    "            f.qualifiers['mislabeled_conf'] = [rec_mislabelled[r.id]]\n",
    "    else:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['mislabeled'] = ['no']\n",
    "            f.qualifiers['mislabeled_conf'] = [0.0001]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.propagate_metadata()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare tree annotation:\n",
    "\n",
    "  - color records that were identified by SATIVA as being mislabelled.\n",
    "  - add heatmap with mislabel confidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "leaf_colors = {'yes':'red', 'no': 'black'}\n",
    "bg_colors = {'yes':'red', 'no': 'white'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='muscleDefault-clipped-raxml-SATIVA.html',\n",
    "            fsize=13,\n",
    "            heat_map_meta = ['mislabeled_conf'], heat_map_colour_scheme = 1,\n",
    "            leaf_node_color_meta = 'mislabeled', leaf_label_colors = leaf_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *rrnS@muscleDefault_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png rrnS_muscleDefault_clipped_raxml_SATIVA_leaf_colored.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='muscleDefault-clipped-raxml-SATIVA.html',\n",
    "            fsize=13,\n",
    "            heat_map_meta = ['mislabeled_conf'], heat_map_colour_scheme = 1,\n",
    "            node_bg_meta = 'mislabeled', node_bg_color = bg_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "target=rrnS_muscleDefault_clipped_raxml_SATIVA_bg_colored.png\n",
    "\n",
    "last_png=$(ls -hlrt *rrnS@muscleDefault_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png $target"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Have a look!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nr.pkl'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'nr.pkl')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
