{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude the records that SATIVA has highligted, redo the alignment and rebuild the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from reprophylo import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj=Project('../fetch_clean_align_tree/reptile_target_locus.csv', pickle='rep_post_SATIVA.pkl', git=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reprophylo/reprophylo.py:1015: UserWarning: Version control off\n",
      "  warnings.warn('Version control off')\n"
     ]
    }
   ],
   "source": [
    "pj.read_embl_genbank(['../fetch_clean_align_tree/12S_preSATIVA_reptiles.gb'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set the stage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 0\n",
      "to keep: 0\n",
      "TOTAL: 0\n"
     ]
    }
   ],
   "source": [
    "recs_to_keep = {'rrnS': []}\n",
    "recs_to_drop = {'rrnS': []}\n",
    "\n",
    "print \"to drop: %s\" %len(recs_to_drop['rrnS'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['rrnS'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['rrnS']) + len(recs_to_keep['rrnS']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Record all record ids and assign them the status keep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for r in pj.records_by_locus['rrnS']:\n",
    "    recs_to_keep['rrnS'].append(\"_\".join(r.id.split(\"_\")[:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 0\n",
      "to keep: 113\n",
      "TOTAL: 113\n"
     ]
    }
   ],
   "source": [
    "print \"to drop: %s\" %len(recs_to_drop['rrnS'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['rrnS'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['rrnS']) + len(recs_to_keep['rrnS']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many do we want to drop? Parse SATIVA output and extract records ids putatively mislabeled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rec_mislabelled = {}\n",
    "sativa=open('../SATIVA/12S.mis', 'r')\n",
    "\n",
    "for line in sativa:\n",
    "    if not line.startswith(\";\"):\n",
    "        rec_mislabelled[line.split(\"\\t\")[0]] = line.split(\"\\t\")[4]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "print len(rec_mislabelled)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adjust lists accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HQ442366.1\n",
      "\tadd to droplist\n",
      "\tremove from keeplist\n"
     ]
    }
   ],
   "source": [
    "for r in rec_mislabelled:\n",
    "    print r\n",
    "    if not r in recs_to_drop['rrnS']:\n",
    "        print \"\\tadd to droplist\"\n",
    "        recs_to_drop['rrnS'].append(r)\n",
    "    for i in reversed(range(len(recs_to_keep['rrnS']))):\n",
    "#        print i\n",
    "        if r in recs_to_keep['rrnS'][i]:\n",
    "            print \"\\tremove from keeplist\"\n",
    "            del(recs_to_keep['rrnS'][i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "to drop: 1\n",
      "to keep: 112\n",
      "TOTAL: 113\n"
     ]
    }
   ],
   "source": [
    "print \"to drop: %s\" %len(recs_to_drop['rrnS'])\n",
    "print \"to keep: %s\" %len(recs_to_keep['rrnS'])\n",
    "\n",
    "print \"TOTAL: %s\" %(len(recs_to_drop['rrnS']) + len(recs_to_keep['rrnS']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out good records in full Genbank format including all metadata. Adjust to valid species and taxids before doing it.\n",
    "\n",
    "Identify records with Subspecies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Podarcis muralis  .. add to records\n",
      "Podarcis muralis  .. already covered\n",
      "Emys orbicularis  .. add to records\n",
      "Podarcis muralis  .. already covered\n",
      "Chelonia mydas  .. add to records\n",
      "Zootoca vivipara  .. add to records\n",
      "Zootoca vivipara  .. already covered\n",
      "Zootoca vivipara  .. already covered\n",
      "Eretmochelys imbricata  .. add to records\n",
      "Zootoca vivipara  .. already covered\n",
      "Eretmochelys imbricata  .. already covered\n",
      "Chelonia mydas  .. already covered\n",
      "Lacerta viridis  .. add to records\n",
      "Emys orbicularis  .. already covered\n",
      "Lacerta agilis  .. add to records\n",
      "Eretmochelys imbricata  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Chelydra serpentina  .. add to records\n",
      "Trachemys scripta  .. add to records\n",
      "Podarcis muralis  .. already covered\n",
      "Chelonia mydas  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Lacerta viridis  .. already covered\n",
      "Lepidochelys kempii  .. add to records\n",
      "Chelonia mydas  .. already covered\n",
      "Emys orbicularis  .. already covered\n",
      "Caretta caretta  .. add to records\n",
      "Emys orbicularis  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Lacerta viridis  .. already covered\n",
      "subspecies: Chrysemys picta bellii\n",
      "Podarcis muralis  .. already covered\n",
      "Chelydra serpentina  .. already covered\n",
      "Lepidochelys kempii  .. already covered\n",
      "Eretmochelys imbricata  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Zootoca vivipara  .. already covered\n",
      "Emys orbicularis  .. already covered\n",
      "Chelonia mydas  .. already covered\n",
      "Natrix natrix  .. add to records\n",
      "Chelydra serpentina  .. already covered\n",
      "subspecies: Emys orbicularis orbicularis\n",
      "Caretta caretta  .. already covered\n",
      "subspecies: Trachemys scripta scripta\n",
      "Natrix natrix  .. already covered\n",
      "Lacerta viridis  .. already covered\n",
      "Chrysemys picta  .. add to records\n",
      "Caretta caretta  .. already covered\n",
      "Dermochelys coriacea  .. add to records\n",
      "Trachemys scripta  .. already covered\n",
      "Lacerta agilis  .. already covered\n",
      "Lacerta viridis  .. already covered\n",
      "Caretta caretta  .. already covered\n",
      "Eretmochelys imbricata  .. already covered\n",
      "Lacerta agilis  .. already covered\n",
      "Anguis fragilis  .. add to records\n",
      "Podarcis muralis  .. already covered\n",
      "Emys orbicularis  .. already covered\n",
      "Coronella austriaca  .. add to records\n",
      "subspecies: Emys orbicularis orbicularis\n",
      "Podarcis muralis  .. already covered\n",
      "Trachemys scripta  .. already covered\n",
      "Chelydra serpentina  .. already covered\n",
      "Vipera berus  .. add to records\n",
      "Dermochelys coriacea  .. already covered\n",
      "subspecies: Lacerta viridis viridis\n",
      "Mauremys caspica  .. add to records\n",
      "Chelonia mydas  .. already covered\n",
      "Zamenis longissimus  .. add to records\n",
      "Lacerta viridis  .. already covered\n",
      "Vipera berus  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "Chelonia mydas  .. already covered\n",
      "Mauremys caspica  .. already covered\n",
      "Coronella austriaca  .. already covered\n",
      "Caretta caretta  .. already covered\n",
      "Lacerta agilis  .. already covered\n",
      "Natrix natrix  .. already covered\n",
      "Podarcis muralis  .. already covered\n",
      "subspecies: Trachemys scripta elegans\n",
      "Podarcis muralis  .. already covered\n",
      "subspecies: Trachemys scripta elegans\n",
      "subspecies: Mus musculus castaneus\n",
      "subspecies: Mus musculus musculus\n",
      "Mus musculus  .. add to records\n",
      "Xenopus laevis  .. add to records\n",
      "Xenopus laevis  .. already covered\n",
      "subspecies: Mus musculus castaneus\n",
      "Mus musculus  .. already covered\n",
      "Mus musculus  .. already covered\n",
      "subspecies: Mus musculus castaneus\n",
      "Cottus gobio  .. add to records\n",
      "Mus musculus  .. already covered\n",
      "subspecies: Mus musculus castaneus\n",
      "subspecies: Mus musculus musculus\n",
      "Xenopus laevis  .. already covered\n",
      "subspecies: Mus musculus molossinus\n",
      "Xenopus laevis  .. already covered\n",
      "subspecies: Mus musculus musculus\n",
      "subspecies: Mus musculus domesticus\n",
      "Mus musculus  .. already covered\n",
      "Xenopus laevis  .. already covered\n",
      "subspecies: Mus musculus musculus\n",
      "subspecies: Xenopus laevis sudanensis\n",
      "Xenopus laevis  .. already covered\n",
      "Xenopus laevis  .. already covered\n",
      "Xenopus laevis  .. already covered\n",
      "subspecies: Mus musculus musculus\n",
      "subspecies: Mus musculus castaneus\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "Seqs = []\n",
    "taxon_to_taxid = {}\n",
    "recs_to_adjust = defaultdict(list)\n",
    "taxon_to_recs = defaultdict(list)\n",
    "\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['rrnS']:\n",
    "        Seqs.append(r)\n",
    "        \n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "\n",
    "        #screen for subspecies and record taxids of valid species\n",
    "        #if valid species record the taxid\n",
    "        if (len(source.qualifiers['organism'][0].split(\" \")) == 2):\n",
    "            print source.qualifiers['organism'][0],\n",
    "            if 'db_xref' in source.qualifiers:\n",
    "#                print source.qualifiers['db_xref']\n",
    "                for t in source.qualifiers['db_xref']:\n",
    "#                    print t\n",
    "                    if 'taxon' in t:\n",
    "                        if not source.qualifiers['organism'][0] in taxon_to_taxid:\n",
    "                            print \" .. add to records\"\n",
    "                            taxon_to_taxid[source.qualifiers['organism'][0]] = t.split(\":\")[1]\n",
    "                        else:\n",
    "                            print \" .. already covered\"\n",
    "                        taxon_to_recs[source.qualifiers['organism'][0]].append(r.id)\n",
    "        #if the species status is 'sp.' but has a valid taxid\n",
    "        elif (len(source.qualifiers['organism'][0].split(\" \")) > 2 and source.qualifiers['organism'][0].split(\" \")[1] == 'sp.'):\n",
    "            print source.qualifiers['organism'][0],\n",
    "            if 'db_xref' in source.qualifiers:\n",
    "                for t in source.qualifiers['db_xref']:\n",
    "                    if 'taxon' in t:\n",
    "                        if not source.qualifiers['organism'][0] in taxon_to_taxid:\n",
    "                            print \" .. add to records\"\n",
    "                            taxon_to_taxid[source.qualifiers['organism'][0]] = t.split(\":\")[1]\n",
    "                        else:\n",
    "                            print \" .. already covered\" \n",
    "                        taxon_to_recs[source.qualifiers['organism'][0]].append(r.id)\n",
    "        #all other cases we consider dubious, e.g. supspecies\n",
    "        else:\n",
    "            print \"subspecies: %s\" %source.qualifiers['organism'][0]\n",
    "            recs_to_adjust[r.id] = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if any of the taxa that were considered dubious are not already covered in the list after removing subspecies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chrysemys picta bellii -> Chrysemys picta\n",
      "Emys orbicularis orbicularis -> Emys orbicularis\n",
      "Trachemys scripta scripta -> Trachemys scripta\n",
      "Emys orbicularis orbicularis -> Emys orbicularis\n",
      "Lacerta viridis viridis -> Lacerta viridis\n",
      "Trachemys scripta elegans -> Trachemys scripta\n",
      "Trachemys scripta elegans -> Trachemys scripta\n",
      "Mus musculus castaneus -> Mus musculus\n",
      "Mus musculus musculus -> Mus musculus\n",
      "Mus musculus castaneus -> Mus musculus\n",
      "Mus musculus castaneus -> Mus musculus\n",
      "Mus musculus castaneus -> Mus musculus\n",
      "Mus musculus musculus -> Mus musculus\n",
      "Mus musculus molossinus -> Mus musculus\n",
      "Mus musculus musculus -> Mus musculus\n",
      "Mus musculus domesticus -> Mus musculus\n",
      "Mus musculus musculus -> Mus musculus\n",
      "Xenopus laevis sudanensis -> Xenopus laevis\n",
      "Mus musculus musculus -> Mus musculus\n",
      "Mus musculus castaneus -> Mus musculus\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "to_fetch = defaultdict(list)\n",
    "\n",
    "for r in Seqs:\n",
    "    if r.id in recs_to_adjust:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        adjust_from = source.qualifiers['organism'][0]\n",
    "        adjust_to = \" \".join(adjust_from.split(\" \")[:2])\n",
    "        print \"%s -> %s\" %(adjust_from,adjust_to)\n",
    "        if adjust_to in taxon_to_taxid:\n",
    "            taxon_to_recs[adjust_to].append(r.id)\n",
    "            recs_to_adjust[r.id].append(adjust_to)\n",
    "            recs_to_adjust[r.id].append(taxon_to_taxid[adjust_to])\n",
    "        else:\n",
    "            to_fetch[adjust_to].append(r.id)\n",
    "            recs_to_adjust[r.id].append(adjust_to)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "defaultdict(<type 'list'>, {'HQ681908.1': ['Emys orbicularis', '82168'], 'DQ874614.2': ['Mus musculus', '10090'], 'HQ681907.1': ['Emys orbicularis', '82168'], 'KF781660.1': ['Mus musculus', '10090'], 'KF781662.1': ['Mus musculus', '10090'], 'AY057791.1': ['Mus musculus', '10090'], 'KF781663.1': ['Mus musculus', '10090'], 'AY581640.1': ['Xenopus laevis', '8355'], 'AM176577.1': ['Lacerta viridis', '65476'], 'KF781656.1': ['Mus musculus', '10090'], 'JX945964.1': ['Mus musculus', '10090'], 'KF781653.1': ['Mus musculus', '10090'], 'KF874616.1': ['Chrysemys picta', '8479'], 'KF781661.1': ['Mus musculus', '10090'], 'AY057793.1': ['Mus musculus', '10090'], 'GU477770.1': ['Trachemys scripta', '34903'], 'KM216749.1': ['Trachemys scripta', '34903'], 'AY057792.1': ['Mus musculus', '10090'], 'KF781654.1': ['Mus musculus', '10090'], 'KM216748.1': ['Trachemys scripta', '34903']})\n"
     ]
    }
   ],
   "source": [
    "print recs_to_adjust"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print len(to_fetch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#print to_fetch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#from Bio import Entrez\n",
    "#Entrez.email = \"L.Harper@2015.hull.ac.uk\"\n",
    "\n",
    "#for binomial in to_fetch:\n",
    "    #print binomial\n",
    "    #handle = Entrez.esearch(db=\"Taxonomy\", term=binomial)\n",
    "    #record = Entrez.read(handle)\n",
    "    #print record[\"IdList\"][0]\n",
    "    #taxon_to_taxid[binomial] = record[\"IdList\"][0]\n",
    "    \n",
    "    #taxon_to_recs[binomial] = to_fetch[binomial]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update the dictionary that contains binomial, taxid for every record id with the ones that we have just fetched from NCBI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#for sp in to_fetch:\n",
    "#    print to_fetch[sp]\n",
    "#    for r in to_fetch[sp]:\n",
    "#        print r,recs_to_adjust[r]\n",
    "#        recs_to_adjust[r].append(taxon_to_taxid[recs_to_adjust[r][0]])\n",
    "#        print r,recs_to_adjust[r]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Modify the records (if necessary) - adjust subspecies to species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adjusting KF874616.1\n",
      "Chrysemys picta bellii\t->\tChrysemys picta\n",
      "adjusting HQ681907.1\n",
      "Emys orbicularis orbicularis\t->\tEmys orbicularis\n",
      "adjusting KM216749.1\n",
      "Trachemys scripta scripta\t->\tTrachemys scripta\n",
      "adjusting HQ681908.1\n",
      "Emys orbicularis orbicularis\t->\tEmys orbicularis\n",
      "adjusting AM176577.1\n",
      "Lacerta viridis viridis\t->\tLacerta viridis\n",
      "adjusting GU477770.1\n",
      "Trachemys scripta elegans\t->\tTrachemys scripta\n",
      "adjusting KM216748.1\n",
      "Trachemys scripta elegans\t->\tTrachemys scripta\n",
      "adjusting KF781663.1\n",
      "Mus musculus castaneus\t->\tMus musculus\n",
      "adjusting KF781656.1\n",
      "Mus musculus musculus\t->\tMus musculus\n",
      "adjusting KF781660.1\n",
      "Mus musculus castaneus\t->\tMus musculus\n",
      "adjusting AY057792.1\n",
      "Mus musculus castaneus\t->\tMus musculus\n",
      "adjusting KF781661.1\n",
      "Mus musculus castaneus\t->\tMus musculus\n",
      "adjusting KF781654.1\n",
      "Mus musculus musculus\t->\tMus musculus\n",
      "adjusting AY057793.1\n",
      "Mus musculus molossinus\t->\tMus musculus\n",
      "adjusting AY057791.1\n",
      "Mus musculus musculus\t->\tMus musculus\n",
      "adjusting JX945964.1\n",
      "Mus musculus domesticus\t->\tMus musculus\n",
      "adjusting DQ874614.2\n",
      "Mus musculus musculus\t->\tMus musculus\n",
      "adjusting AY581640.1\n",
      "Xenopus laevis sudanensis\t->\tXenopus laevis\n",
      "adjusting KF781653.1\n",
      "Mus musculus musculus\t->\tMus musculus\n",
      "adjusting KF781662.1\n",
      "Mus musculus castaneus\t->\tMus musculus\n"
     ]
    }
   ],
   "source": [
    "for r in Seqs:\n",
    "    if r.id in recs_to_adjust:\n",
    "        print \"adjusting %s\" %r.id\n",
    "        source = r.features[0]\n",
    "        print \"%s\\t->\\t%s\" %(source.qualifiers['organism'][0],recs_to_adjust[r.id][0])\n",
    "        source = r.features[0]\n",
    "        source.qualifiers['organism'] = [recs_to_adjust[r.id][0]]\n",
    "#        print source.qualifiers['organism']\n",
    "        for i in range(len(source.qualifiers['db_xref'])):\n",
    "            if 'taxon' in source.qualifiers['db_xref'][i]:\n",
    "                source.qualifiers['db_xref'][i] = \"taxon:\"+recs_to_adjust[r.id][1]\n",
    "        if not 'note' in source.qualifiers:\n",
    "            source.qualifiers['note'] = ['modified from original record: adjusted to species']\n",
    "        else:\n",
    "            source.qualifiers['note'].append('modified from original record: adjusted to species')\n",
    "#        print r.format('genbank')       \n",
    "#    print r,recs_to_adjust[r]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write only good records with all metadata in Genbank format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "       \n",
    "out = open('../12S_UKreptiles_SATIVA_cleaned.gb','w')\n",
    "SeqIO.write(Seqs, out, \"genbank\")\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude the records that were highlighted by SATIVA. Then realign and rebuild phylogeny."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.exclude(start_from_max=False, **recs_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 112\n",
      "Number of species (see below for full list): 22\n",
      "['Anguis fragilis', 'Caretta caretta', 'Chelonia mydas', 'Chelydra serpentina', 'Chrysemys picta', 'Coronella austriaca', 'Cottus gobio', 'Dermochelys coriacea', 'Emys orbicularis', 'Eretmochelys imbricata', 'Lacerta agilis', 'Lacerta viridis', 'Lepidochelys kempii', 'Mauremys caspica', 'Mus musculus', 'Natrix natrix', 'Podarcis muralis', 'Trachemys scripta', 'Vipera berus', 'Xenopus laevis', 'Zamenis longissimus', 'Zootoca vivipara']\n"
     ]
    }
   ],
   "source": [
    "species=[]\n",
    "for r in pj.records:\n",
    "    if r.id in recs_to_keep['rrnS']:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        species.append(\" \".join(source.qualifiers['organism'][0].split()[0:2]))\n",
    "    \n",
    "#print set(species)\n",
    "total=len(species)\n",
    "species=list(set(species))\n",
    "\n",
    "print \"Total number of records: %i\" %total\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print sorted(species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write short 12S rRNA sequences (< 400bp) to fasta file to build new alignment outside Reprophylo later in script using hmmbuild. The Hidden Markov Model alignment will include all 12S rRNA sequences - short and long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of records: 33\n"
     ]
    }
   ],
   "source": [
    "temp_seq = []\n",
    "\n",
    "for r in pj.records_by_locus['rrnS']:\n",
    "    if len(r.seq) < 400:\n",
    "        temp_seq.append(r)\n",
    "        \n",
    "total = len(temp_seq)\n",
    "print \"Total number of records: %i\" %total\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "out=open('reptile_short_to_align_SATIVA.fasta','w')\n",
    "SeqIO.write(temp_seq, out, 'fasta')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter by Length"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write cropped records (minimum length of 400bp) to fasta and Genbank file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "temp=[]\n",
    "for r in pj.records:\n",
    "    if len(r.seq) >= 400:\n",
    "        temp.append(r)\n",
    "\n",
    "out = open('reptile12S_cropped_400bp_SATIVA.gb', 'w')\n",
    "SeqIO.write(temp, out, 'genbank')\n",
    "out.close()\n",
    "\n",
    "out = open('reptile12S_cropped_400bp_SATIVA.fa', 'w')\n",
    "SeqIO.write(temp, out, 'fasta')\n",
    "out.close()\n",
    "\n",
    "\n",
    "del(temp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now view records in cropped genbank file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index 0, ID = AJ001471.1, length 544, with 2 features\n",
      "index 1, ID = JX454972.1, length 969, with 2 features\n",
      "index 2, ID = GQ142072.1, length 459, with 2 features\n",
      "index 3, ID = KM401599.1, length 953, with 2 features\n",
      "index 4, ID = KF781329.1, length 430, with 2 features\n",
      "index 5, ID = JX454980.1, length 968, with 2 features\n",
      "index 6, ID = JX454970.1, length 968, with 2 features\n",
      "index 7, ID = JX454976.1, length 969, with 2 features\n",
      "index 8, ID = AJ001480.1, length 543, with 2 features\n",
      "index 9, ID = AF149947.1, length 460, with 2 features\n",
      "index 10, ID = KP221806.1, length 969, with 2 features\n",
      "index 11, ID = L28077.1, length 966, with 2 features\n",
      "index 12, ID = JX454990.1, length 969, with 2 features\n",
      "index 13, ID = AJ001472.1, length 547, with 2 features\n",
      "index 14, ID = AJ001470.1, length 544, with 2 features\n",
      "index 15, ID = AF149962.1, length 465, with 2 features\n",
      "index 16, ID = AY770536.1, length 440, with 2 features\n",
      "index 17, ID = JX454983.1, length 970, with 2 features\n",
      "index 18, ID = AF149960.1, length 465, with 2 features\n",
      "index 19, ID = KF874616.1, length 969, with 2 features\n",
      "index 20, ID = EF122793.1, length 965, with 2 features\n",
      "index 21, ID = JX454981.1, length 968, with 2 features\n",
      "index 22, ID = HQ442366.1, length 971, with 2 features\n",
      "index 23, ID = JX454986.1, length 967, with 2 features\n",
      "index 24, ID = KX694573.1, length 455, with 2 features\n",
      "index 25, ID = GU477775.1, length 445, with 2 features\n",
      "index 26, ID = KP256531.1, length 973, with 2 features\n",
      "index 27, ID = KM216749.1, length 968, with 2 features\n",
      "index 28, ID = AY122682.1, length 590, with 2 features\n",
      "index 29, ID = AF149959.1, length 459, with 2 features\n",
      "index 30, ID = AF069423.1, length 969, with 2 features\n",
      "index 31, ID = AY770537.1, length 424, with 2 features\n",
      "index 32, ID = JX454969.1, length 968, with 2 features\n",
      "index 33, ID = FJ392294.1, length 969, with 2 features\n",
      "index 34, ID = AF149961.1, length 459, with 2 features\n",
      "index 35, ID = JX454984.1, length 969, with 2 features\n",
      "index 36, ID = DQ533485.1, length 968, with 2 features\n",
      "index 37, ID = KC990830.1, length 946, with 2 features\n",
      "index 38, ID = EU443256.1, length 941, with 2 features\n",
      "index 39, ID = FJ460597.1, length 955, with 2 features\n",
      "index 40, ID = JN999703.2, length 884, with 2 features\n",
      "index 41, ID = AJ001468.1, length 544, with 2 features\n",
      "index 42, ID = JN860636.1, length 965, with 2 features\n",
      "index 43, ID = EU543221.1, length 669, with 2 features\n",
      "index 44, ID = AM176577.1, length 943, with 2 features\n",
      "index 45, ID = KC692465.1, length 973, with 2 features\n",
      "index 46, ID = JX454971.1, length 969, with 2 features\n",
      "index 47, ID = AY122780.1, length 591, with 2 features\n",
      "index 48, ID = AF149958.1, length 460, with 2 features\n",
      "index 49, ID = AB012104.1, length 969, with 2 features\n",
      "index 50, ID = AY122836.1, length 593, with 2 features\n",
      "index 51, ID = JX454977.1, length 969, with 2 features\n",
      "index 52, ID = AJ001469.1, length 544, with 2 features\n",
      "index 53, ID = GU477770.1, length 421, with 2 features\n",
      "index 54, ID = KM216748.1, length 969, with 2 features\n",
      "index 55, ID = KF781663.1, length 955, with 2 features\n",
      "index 56, ID = KF781656.1, length 955, with 2 features\n",
      "index 57, ID = AY012114.1, length 953, with 2 features\n",
      "index 58, ID = AY581648.1, length 840, with 2 features\n",
      "index 59, ID = M10217.1, length 819, with 2 features\n",
      "index 60, ID = KF781660.1, length 955, with 2 features\n",
      "index 61, ID = V00665.1, length 956, with 2 features\n",
      "index 62, ID = X84382.1, length 956, with 2 features\n",
      "index 63, ID = AY057792.1, length 956, with 2 features\n",
      "index 64, ID = AB188189.1, length 783, with 2 features\n",
      "index 65, ID = AY533105.1, length 955, with 2 features\n",
      "index 66, ID = KF781661.1, length 955, with 2 features\n",
      "index 67, ID = KF781654.1, length 955, with 2 features\n",
      "index 68, ID = AY581639.1, length 840, with 2 features\n",
      "index 69, ID = AY057793.1, length 956, with 2 features\n",
      "index 70, ID = M27605.1, length 945, with 2 features\n",
      "index 71, ID = AY057791.1, length 956, with 2 features\n",
      "index 72, ID = JX945964.1, length 955, with 2 features\n",
      "index 73, ID = Y10943.1, length 946, with 2 features\n",
      "index 74, ID = DQ874614.2, length 955, with 2 features\n",
      "index 75, ID = AY581640.1, length 840, with 2 features\n",
      "index 76, ID = AY581638.1, length 838, with 2 features\n",
      "index 77, ID = AY581644.1, length 840, with 2 features\n",
      "index 78, ID = KF781653.1, length 955, with 2 features\n",
      "index 79, ID = KF781662.1, length 955, with 2 features\n"
     ]
    }
   ],
   "source": [
    "for index, record in enumerate(SeqIO.parse(\"reptile12S_cropped_400bp_SATIVA.gb\", \"genbank\")):\n",
    "        print(\"index %i, ID = %s, length %i, with %i features\"\n",
    "          % (index, record.id, len(record.seq), len(record.features)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now need to filter all 12S rRNA sequences within pj.records dictionary by length to remove all sequences shorter than 400bp within Reprophylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distribution Of Sequence Lengths\n",
      "Distribution Of Sequence Statistic \"Gc_Content\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/pymodules/python2.7/matplotlib/figure.py:371: UserWarning: matplotlib is currently using a non-GUI backend, so cannot show the figure\n",
      "  \"matplotlib is currently using a non-GUI backend, \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAFEAAAIiCAYAAACwpiRwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE75JREFUeJztnXu0XFV9xz83L/OABAKEpAVMeDUEKgHaAEVLEhpARJFS\nNVIQKaulYIXYVl7LlquiYvCBWh+0yEOttLa0AkvAli4I1mWJQYIXkkuJ5CIEEhAkQHilye0fvz1r\nzp2cmTt77m9m9rnn+1lr1j1z5syZfT/rPPbZe5/vASGEEEII4ch1wCagLzNvOvCfwP8C/wHskvns\nUuBRoB84PjP/iLCOR4EvtbG8SfI24DCGSlwOXBSmLwauDNPzgNXAeGA2sA7oCZ+tBBaE6duBE9tW\n4kSZzVCJ/cCeYXpmeA+2FV6cWe5O4ChgFrA2M38p8I12FLQRYzr9g8OwJ7aLE/5WhP4G8GRmuSeB\n38yZvyHM7yipScwyGF4uTJkyZV1mna2+VuetOzWJm7DdGGxXfSZMbwD2ziy3F7YFbgjT2fkb8la8\nZcuW/bDjaL3Xx4f5vAc4NG/dqUm8FTgrTJ8FfD8zfykwAZgDHICdUDYCLwJHYv/kmZnvlIKbgKeA\nN4AngLOxKs5d5FdxLsPOyv3ACZn5lSrOOuDLDX5vuENDbxNldju8FJXhBCx0WMeox0NA7jpSOyYW\nEkl0QBIdkEQHJNEBSXRAEh2QRAfGdbsAKdDb2zuYme5ptGwe2hIdkMQMrWyFIIkuSKIDkuiAJDog\niQ5IogOS6IAkOiCJGXaGh7tdhtSp21E1HTYNwiPTq0NYotahLdEBSXRAEh2QRAck0QFJdEASHZBE\nByTRAUl0QBIdkEQHJNEBSXRAEh2QRAck0QFJdEASHZBEByTRAUl0QBIdkEQHUpV4IXYP80NhGlqL\neykth2ACJwJjMXH7ERf3krdxlGoExFzgPuA1YBuwAjgNeBdwY1jmRuDdYfoU7C7+rcAAJnEBHSRF\niQ9hwUPTgcnASVhARmzcS8dI8WagfuCz2HFvC7arbqtZZri4l9zPFixYcHNmes255567FmDJEUdM\n3DZhws4T3nhjYt/9959eWeaaa645aOXKlfMq71euXBn7vyTDp4DzMLnZuJdKetMl4VXhTiydpJZS\nHRMBZoS/+wB/CHyX+LiX0nMvNuByNbAozGsl7iVL27bEMlG63blQSKIDkuiAJDogiQ5IogOS6IAk\nOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDog\niQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LoQKoSL8Vu\nGu/D7rp/E8rFiWI28BgmDuCfsbgC5eJE8CKWcTMZS0uZDDyFcnGieB74PPBLTN4L2G6sXJwI9gOW\nYbvmZuBfgDNqllEuzjC8D7g28/5M4KvAWpSL0zT9wFHAJKAH+ANgDXAbysWJ4iKqVZwbsTOvcnES\noFS7c+GQRAck0QFJdEASHZBEByTRAUl0QBIdkEQHJNEBSXRAEh2QRAck0QFJdEASHZBEByTRAUl0\nQBIdkEQHJNEBSXRAEh2QRAck0QFJdEASHZBEByTRAUl0QBIdkEQHerpdgA4ySM7/Ozg4OGQIcU9P\nTyMnuevQluhAijcDdZSenp6e6bDpOXhht6F3JDSNtkQHJNEBSXRAEh2QRAck0QFJdEASHZBEB1KU\n+FvAA5nXZuACFOnSMmOAp4G9STjSJXWOB34UpvupJpDMpJoBcSkmtcKdWPxBLaW9aXwplnkDinRp\niQnAOxm6lVVQpEuTnILtmhX6UaRLNO+nuiuDRbco0iWCKcCvgJ0z8xTpkgCl3J0LgyQ6IIkOSKID\nkuiAJDogiQ5IogOS6IAkOtBKU9iuwAzsEuhZ4NeuJSogzUgcizVLvQc4lmpzVIWngRVYlPMtwHbP\nAhaBRhLHAedhbXWzgC3AKuB24DlssON0YH8sjX0pJvQzwNeBbW0rdYF4BBP3TWAhtkXWYyywGLg+\nfKe/wbLdoitNYV+k2o8Rw0zg6ha+127UnuiA2hNTppUqzpHAqVh/BtizU74P3OdVqNHMWOA6rAqT\n97qBxiefbpPEMfFyTNa/YSMMpoXX0cC/h896R/ojbSQJiY9jvWx59GCDjR4f6Y+0kSROLDOwK5J6\nK7+F1qpEhSdG4qPseMmXZSZWQRcNeD/W2DA/57PDwmdLO1qiONq2O8dUcQ7EqjM/xY5/a8P8edjT\ne34elvnbmu99IuI3Rj31qjbDvVIhiS1x34hlS0WMxIF2FaLotDrIczLw5jD9OPCKT3GKSWwDxMHA\nHdiI/ofD64Uw7xDfohWHmC3xMKwbYAo7np2PB47Bug8e8CzgaOMubKs7POezw7Gt866OliiOJK6d\nXwauaPD5p7CugVRJ4tr5Nawjqh5PA69GrG/UECPxdqxXrx7vDMuUjhiJfwnsBvwr9hDqncPrSOBm\nrPv0I94FLDrbsb7j7TnT9V4p9zV35bLvW82UrJkfGe00kvjBThWi6KjL1IFGEncdwXpH8t3C0Uji\nANbDt1vE+mYAnwTWj6BMo4rzgGeA17Eu0T8FDgV2yiyzM3bJ9yHgB8BWYCNwbkdL2hxdu+ybht3Z\nPsDQqszr4ZWdtw74KENvakyJtklsNoRyDFbBPhZrtdmD6kjZh4C7gfubXFe3yA2QBJNYif57vnG3\nb+46mm0K2w78T3iJGlKt4uyCXV6uBdZgl5aKdInkRuBPwvQ47NisSJcIpmH927Uo0iWCOdgJ63rg\nZ8A/YF0SinSJYBxW9/wLbLTF1QxNGwFFugzLTIZe8bwVq8ivRZEuTbMReAIb1wM2zudh4DYSjXSJ\n3Z3HYP/U/tg1dV7l1WMA04eBf8TE/AI4GxvK/D3gHOwK6r1h2TVh/hrg/4DzSfgsegBW0KIMYKol\niQFNX8EGNV2EXeY9F/FdEdgCfLbbhRgBSZxYXie/Elx6YiT+EBtvI2qI7Xc+Gvhr7KwpAo3aE9ez\n4zFgJ2B3rH/5KYb2M/eE5VMdUduV9sRWbuxJtn7WThpJXNipQhSdmGPi72PdAvXYIyxTOmIk3gMs\nafD5cVglvHR4NkCMpaTHRE+JR2M5sKVjuGvnC4FlVLewq8kfcjwdmIrdVF46hpO4mWpVZza2pT1T\ns8wg1t73EyzBRDRgAEtqKipJNIXNjli2VKTYPVA4YiRWxmQ3er2MtX5/CevKLAUxEr8F9GEX4I9g\nHUS3hume8NmdmMwPY6MSUm2M6BqLsdvS3p7z2UnYmfy48P4dwBtYZ1MqJDGM5CfAVQ0+v4qho8a+\nQeM7sDpNEt0Dh9K4e2A98JbM+9XEDVUuLDESN2N9zvU4Dngx835q+M6oJ0biTVjQ2jXYY+HGhu/P\nDfNOZehDaBZhZ2qRYTI2JqbSSb81vCrv7wjLAEzEYlHzTkLdoutjtrOchN1RWon+G8CqOqnfYdq2\nPpYykcTZWdQhdlTYm7EbfRqNCls80kIVjRiJb8fGBI7HrpGfz1mmlN0DMawGNgC/0+2CtEgSx8S5\nWPfAqojvlIIYib/CRoaJGmKbwk5rV0GKTMyJ5QbsUu5WrNH1MfKDM3458mIVixiJ2YcynFxnmUHS\nztpuCzESm7krQFWcUU4SVRxRh1iJU7FwjR9j9xcfHebvjqUaz/UrWnGIOSbugcmbg93ltB8wKXz2\nHHar2K6UMC8sRuIVWFvbUdj4nOyYnEGs6lO6xgeI251Pxh7gVS8w4zFgnxGXqIDESNwdOw7WYzvW\nLVA6YiRuwo6D9ZhPCa9WIE7iD7DbZPPG2BwJfID6jxoRgVlYe+JG4O+x3fdG4J+wISMD+HXWD2AP\nhHiA6g3gI410SWIYCdiJ4xZ2TPi8DdjL4wcC6zFpWUZdpMs0LPbqSNozVGR9znpHXaTLZmw3u4/2\n3Dw+iAWfr8LS8kCRLtEcg40o2wM7DtY+G7UwkS7NJhpnl2tHuvHlwF9hIpOMdEkx3Xgy1rD7EpbM\ndDzwceyy8iwsQqE20uW7wBew3bjjkS4pMgc7267GshkvDfOnY8fJvCrOZdhZuR84oc56kzs7F5Hk\nzs4igyQ6IIkOSKIDkuiAJDoQc9nXakjGvS1+rzDESLwn/M0b/F1vQHgphpXESDwHu/HxAOyevexz\n+07HriS+4lq6ghAjcSrWqjIXa+HO8kns3r+pwJd9ijY6eRT4WIPPP4ZtjamSxGXf3jR+uOErqN95\nWAaAM8nvW54UPhsYeZGKR8wx8fPYjZCrgK9SbRQ9CHuozUHAn7uWbpSyDNula1u2t2AhlSmT1A2S\nu2CtzZV8h19g/SAvtLCuTtL1h9pkeQELBxeBViTOwe7An4FVugewjNmZ2O6ge12GYTkWTV/p2auM\nR5yGHRdTHuCZRD3xXCzZ+O+wY2L22LAZG15S79aMUU3M7nw+1k25DBurWEsf9jS10hGzJR6IdVXW\n41ny5Y56Yh/XPqXB5/uQfjWnLcRI/CkW25LHROyy78cjLlEBiZG4HPg94DtUk5hmAScCK7AGis+5\nlm6U8mfYbl172fcq6T9UO4kkT7BhxrcBf4Q1OPRgbYjfY8eGWjEKSWZLzDIe+F1sONsaLOFY5LAQ\n6zOpbdmYg43uzw7yvL6jJYuna0PrbiC/32QFJu9HWGNtX3j/wVZ+pEN0TeIa7Fo5y1xM2IrMvElY\ntux/tfIjHaJrDRAz2XFLXBj+XpuZ9yo25PctlJDhJL4JE5RlQfi7omb+EwwdAlwahpP4BHBwzby3\nYvc6194MORldO+dyL3bj42+H96diiXV35Cx7CKpw57Iv1mK9Ddv6tmPN/4fULDcWe3pa7UkoJbp2\nYnkMa2i9HYv6uz28f6hmucXhc92qO8pJoo9F1EESHZBEByTRAUl0QBIdkEQHUpY4Fot0uS28H2mk\nS9tIWeKFWHtmpYJ7CSbxQKzdshJZMA94X/h7IvA1Ovx/pSpxL+xpG9dSHTj1LizMiPD33WH6FOw5\nMFuxYX7rqDbXdYRUJX4R+CjW4FFBkS4RnIy1GD1A/aedFybSpVt8GmsMXo9l42wBvk3CkS6pcyzV\ns/NyqnFWl7BjVtgEqlGt9W7WzKW3t3ewt7d3sBud952iUvArseEq52AnkPeG+WvC/DXYUOjzaXGL\n2driQ75Tl7iCaofY89R/5N2nw2tEvLRjf1JTpHp2LhSS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiA\nJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6\nIIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDqQosSJwH3Y3fRrgM+E\n+crFieA1YBEwH4ucXoSFASsXJ5JXwt8JWFLTr1EuTjRjsN15E3A39nAI5eJEsh3bnacBP8R26Syu\nuTg332yz+/r6Ts8uX+RcnFr+BnteYNtycSqRLk2UpTDBvLtTPfNOApZgaU23AmeF+WdhzxAkzF9K\nNRfnAKCjm0yKu/Ms7MQxJry+jZ2NH6DNuTitkqLEPuDwnPltz8VplRR358IhiQ5IogOS6IAkOiCJ\nDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5I\nogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOpChx\nb6q5Dw8BF4T5inSJYCvwEeBg4CjgQ8BBKNIlio1YEgnAy8BaLF1EkS4tMhs4DEtsUqRLC+wE3Axc\nCLxU85kiXZpgPJaHsywzT5EuEfQA38TSRa7OzFekSwTHAGcAP8diXMCqMFeiSJem+W/q7yGKdBmt\nSKIDkuiAJDogiQ5IogOS6IAkOiCJDkiiA5LogCQ6IIkOSKIDkuiAJDogiQ5IogOS6IAkOiCJDkii\nA5LogCQ6IIkZmhyPswOSKKLwGJ9TmFFhhUMSHZBEByTRAUl0QBIdkEQHJNEBSXRAEh2QRAck0QFJ\ndEASHZBEByTRAUl0QBIdkEQHUpV4HZbz0JeZ1+5Il4Utfi9Z3oYFaGQlLgcuCtMXYzeRg0W5rMYi\nD2ZjaSR5G8dwHVW9TZSrozejezCboRL7qSaQzKSaAXEpJrXCnVieTi1tk5jq7pyHIl2caSXS5cFh\nvgNw+TCfP5g3s0gSN2G78UYs0uWZMH8Dli9WYa8wr5b57SpYkXbnZCNdUuUm4CngDeAJ4GysinMX\n+VWcy7Czcj9wQkdLKoQQxWU+sLhm3hnAeuxsfw12gopi7MjLVSiuxf7nFeH9POAO4P4w7wNYjWVF\n7rcFYPXH7CXhJxhagT4Hy7EVDXiNoRXze4ArMu/3Z8fAy2EpUmXbg2epXlePA47AolYrTAC2d7pQ\nReM72DFwX6xZ7WUsdrXCadS5PhZV5mBXNtupBlZmuQX4QqcLVUTGY1WdvOay+cBunS2OACwEt4z0\nYCnxxwEzsBPsYJg/iAWjN02R2hM9WY7FT98NPM3QxtrC9aN0i03Ae7xWVrZ6YoUxVOOnR0zZrp0r\n7Io9ZeMej5WV9Zg4DfhjYAkWir41zK+cWC6o871cynp2vicznT2RVCQu6mhpCsg44B3A7t0uSNF5\nHRth4UJZz84PYs1eLpT17PwkcBVWX9yC7eKTMq9XY1ZW1hNLozbDQSI3rjJWcSYAq7AxkP3DLCsa\n8Az26DoXynpM3BM4HBt5O2LKuDsDTMb6m5dg3aVbwvyWrljKKnEe8LMwvW9mfkWiEEII0T3+HwWw\nOJLJ5e/gAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7bd6ea9f10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEUAAAIiCAYAAACe4wXiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFzxJREFUeJztnXt0HNV5wH+7klarF7Yk29jG2MTYBJtAeARDkwabh6GB\nvBpK01JCTNqGQhpaHkkTSFIBeZDEefTQNIcW7IRHzimEPpKTkNjuweRFcGJskLENGHDwS5bxW5J3\nV9JO//h20Gr1rXZH2t2ZHX2/c/ZIOztzdfenuXfu3Mc3YBiGYRiGAUANsAH4ceb9ImBdZtvvgHN9\nypev3AI8Avwo834tcFnm9/cAT1Y6Q9FK/8EcZgGXA/cDkcy2PcCkzO+TgV0+5MtXHgPOAhYzVHzm\nADuA14GdwImVzpSfZ8p7gW6k7ohkbX8AuAmYDdwMrNAObmpq2gY443xtLPF3GjdfRs6I15Ai0ws8\nBBzJ2icCHM5zvFMg/Y4i8lAoDV/JLj7PZt4DXIxcgTTKJqW2iAMrhZvBjwPfAeqBY5n3RpEUOlOW\nlCCNqqMUX0hNw+92SiAxKQomRcGkKJgUBZOiYFIUTIqCSVEwKQomRcGkKJgUBZOiYFIUTIqCSVEw\nKQpB6rguCWfEZ52ysH76JU3R+lhvOpnanOxa83xi50te0giVlDPis05Z2rxg2W1Tl6bcbcv3rV4G\nfM+LmFAVn4X10y/JFgJw29SlqQX10y/xkk6opDRF62Pa9uY82/MRKim96WRK296TZ3s+QiVlc7Jr\nzfJ9q4edFcv3rY5tSXat8ZJOpPAugcVByf8Z8VmnLKiffklztD7Wk06mtox+9VHTqGZshLCSmBQF\nk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRCNWwKNpY8AhtL\nVrCxZAUbS1awsWQFG0u2sWQVG0uuJCZFwaQomBQFk6JgUhRMikIQpOSGNAP4JLAF2AR8tdIZCkLX\nwT8Am4GWzPsLgfcDZwD9wFSf8uUbs4A1iAj3THkUuKiIY0Pbov0W8CkgnbVtPnAB8FskEN47Kp0p\nP4tPdkizJVnba4FW4HwkROKjwFwtgUWLFj2e9fvm66+/fstof/C+++5bsG7duoXu+3Xr1o0172Uj\nX0izJxiK3gWwDWhXji9b8QkK2SHNrgfuzPx+ChIZUKNsUoJw9XFxM7gi8+oEUsC1vuWoCgnt1SeQ\nmBQFk6IQpIq2JNiwaQ42bKpgw6YKpRo2DVXx6U0nU139h9sPDPbOjRCpcXAG22qaXu1JJ7u9pBOq\nM+WFxJ5tDx985uyF8ZlNC+Iz4gvjM5sePvjM2ZsTe7Z5SSdUZ8pp8RnzLm459dnHDq+fW0tNzQCD\ngxe3nPrqzoFD8zqTu1YVm06opDRF62NnNczef1bD7P3Z2x8/vMGmYuTidSpGqM6UzcmuNXd3/+Tm\nK487a7Zb0T5+ZMPrW5JdP/CSTqikAHSnjk5+4uim2fWRutqk0z/QnTp6pPBRwwmVlBNrJ199RsMJ\n8+bE2uujkUg07Tg1x0Ub5vU5qaufZ2dHsemESkosUnvejNpJM89tmBOPEIk4OE4qPdBUH6k9z0s6\noapo66M1s9/VNK+ptbapZnJtY7S1tqnmXU3zmmLRmtle0gmVlBpqiEeHn/zxaC01Hr9mqIpPyhnY\nsT11YMqsusnN7rad/Yd6+p3BfJ3fKqGScmQw8fJjB9ef01rXUBOjNpJiwDnYfyx6ePDYy17SCZWU\nvsHkpOZ4ffwjreenkVmPzor9v4n3JVKTvaQTqjplVqztbTdOWXzYcZxjOBxzHOfYjVMWH54Vaz3N\nSzqhOlPqo7XReLRuIB6tO5q9PR6tq/GSTqik9KSTu3alDs2IRGiOEiGNg+PQczSd2OQlnVAVnxcT\nXb985NAzjVNrWyLttc3RqbUtkUcOPdP4UmLvr7ykE6ozZWHDzDPPbZjzygMHfj2jLhKN9jvp9LkN\nc/bsGjj09s7U7qLTCZWUpmis/sKWUw9dyKmHsrev6d0a95JOqIpPbzqV1Lb3pJMJL+mESsrmxJ4V\ny7tX1WVvW969qm5Lomull3RCVXzcftgdAweva47Wx3vSycSWRNdKL/2z1Y5NxagkJkXBpCiYFAWT\nohCqS7JLR0eHk/W759WkdqYohFrKWM4SCLmUsWJSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklR\nMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCgEQYoW0gzgViRWU1ulMxQEKW5Is+wJNCcCS4E/\n+JEhv6XMAi4H7md4BOFvAp/2JUf4L0ULafYBYCfwvC85InghzRqB25Gi45J3PDhfSLPHH5fNnZ2d\nV2fvX60hzX4I7M1sew0Jk7gdmKYcn3ciYEdHh5M9HWMUAhe96/bMCySk2W3An+Xs8xpwDnCggvny\nvU7JRvuv+RKcLigzmZ7KvHJRwyOWmyCdKYHBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBS\nFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiY\nFAWTomBSFEyKgklRMCkKXicXNwJvRebKO8A+4EWgr8T58pVipLQBy4CrkHnyucf0A+uBx4DvAQdL\nlz1/GE3KZODzwI1APXJGPAK8AuxHlpy0AfOA84FvAF8C/g24GzhctlyXmdGkbAOSyBKUh5EVFaNx\nMnAN8mD5ZcCUEuTPF0araO9GFgzcTWEhIGfQnZljvjj+rI2fNtjbAi94PW5MUTgDgkOe/LfB3v1w\nCKAdJh+A472kYZdkhfFIuQr4GXJ6rgI+UpIcBYCxSvkE8H2knfIjIIVcjv+pNNkKNvmkbQL+Imfb\nt5BFkJUi71K6NtjrwIsOvNgmCzU9pVHoTHkWabDl0oSsHc5mF9Cs7Ft1FJLyC+A3SMOsIWv7j5Hi\n8jFkDfGtwOcYGa8gtJyHrCZ/Fbgss60B+C5yz5MGEsBKoKWC+Spb8SmWOuAORMLDDLVWo8jNoaen\nt5UI3+oUl37kvuZMJGjDVuBa5CzpBgaLTCe0RIC/QVaVrwLe4lM+fD1TPoQUmf9Clt7XIaE9FiJN\n6U1IZIsJ0zq+CSkia4H/RuqUn+bs8z4kkMOzwNkVzJtvFe0rwL9mvb8CkZS73L4FuBfpaqgUvhWf\nKUjxcHFvw3PjJB0FPolEt6h6CnVHPg3cjMRMOoj0xB1Brj4avy1d1oLLyUjDLZ15vYGEBwoCZSs+\nhc6UV5C2yXykFbsVab2Wmhrg98j91PuAryPhiVKZPFxHBft8i7mMppFO642URwiMDGu2CjgNeDvw\nEvDZMv1dldGktI4jXS/HamHNVjMU0euZzD4VYzQp24F/Bto9pDeN4ju6XbSwZtl8jJFto7IyWp3y\nGaR3/nYkUz8F1iFlvCezTwtS3/wR8t++FBkTKrYHTgtrls0dSL3yA+3gfCHNlp5zTnwwFmsBiKVS\n8c7166+G0oU0m4SU5+0MXYHSSCMtmbNtG/If99J9oIU1ezDz2TLg10C+Zx2X7epT7BBHFFiENM4W\nAlMZGkveBDyJDJ2OBzes2fuAP0E6thYjzQCNsg1xFDvAnkYaZuVsnEUY+s/dC8SQChekEXljGf/2\niIxUKzYYVklCd6Y4jjOi8oxEIvm+p50pxRKUMIklIxKJRHLrFK9p2Jmi4EXKSmQMKB+LgBXjy04w\n8CLlo0j/Sj7mIq3QqqeUxacJGR+qegpVtHMyL/eytQC4QNmvHbgBuf8JPR0Mv+kb7TWAjBpWCt+6\nI/+HoTknK4B/Z+T9j4N0JaxD7nirnkJSNmZeACcBjwOd5cxQEPDSeOsoVyaChtcWbRS4BJll3Y5+\n73TXeDNVTcxHetwLVbiVwreKNpt7kQbap5Getv0ejq0qvEh5N/AvwPIy5SUweGnRJpF5b6HHi5Sf\nA+8qV0aChBcptyDjO7chncqhxUt35GvITd8UZOLfboZPAHR74yv1/AzfhzhAnt+VNyNZf6Tq8SJl\nSbkyETSsO1JhLFIWIxON/wM4NbOtGelnGc/0jaqkBniUoeb8IHBR5rMGZLLxHRXMj+/T0EGmV3wI\nuTQvYHiFewyZZ/seD+kFFi9SrgUeAr6Nft+zFbl7rnq8SDkJWfuTj0OEpE7xIuUooz989WRkvkrV\n40XKr5AV6toxrcjctCdLkalq4h3InfJaZNArjVS6f4d0bvch0zwrRcEnVlaik+n3yNXnAYaGR92+\nlW7gg4xhCX05aYfJ/fmnh+XFax/tT5AKdylDl+WXkG6FwMVQGeVGcFS8SJmNWO9DVpXmrixtRO6g\nXx9LRoKEl4p2O1JE8vF+vE0qDiylvCEMzc1lKb/IqWQ6dqqdQnXKRzMvlzuQlaa5tANvQ+5/qp5C\nUloZ3r04FemSzMYdYH+Ayt4ll41CUr6deYE01m5GglWFGi+X5NBUpIWYMF/UC16l/CXSfbCP4YPq\ng1k/qx4vxedTwFeRVu1v0TuaJtwQxyeQ9XwXId2PocVL8ZmOdEeGWgh4k/IKY5jnXo14kbIcac1W\nMsSQL3ipU9JIL9YWZJ7+q+hXmweVbVWFFykrs37P15x3mGBSLiq8SzjwImVtuTIRNMbTzJ9CFQfi\nHQ2vUk5A6ozDSA9+NxJs5vuZz0KB147rZ5Ae8ucYGs5YiIRdXYqsHKv6RQtezpS7kcbbe4GzkNHC\na5CIXVcgQ6qBCOPs0tHRMaZ7MS9SLkXiRWphO55AoqBfpnwWahLIEGk+biQgIc3Gm4aXM2UXcOEo\nn7+bkTFqQ89XkKb+PUhcFZdJWZ99uYL5KduZ4oUmZDpGGllV+ofMayCz7ZfI0GmlCIQUkGCaf4tU\ntlsyr58gd89jDRFQg4Qfcsem25C4KS8hUbzydVcERko5uAUZNvlR5v3XkDVFIJMP78lzXGilzALW\nIBW4e6ZsZWgKxXTyh2QMrZTHkIbgYoakZD8KJ0L+R+OUTYqfoUIKhTMDyXTeL58vpFk+ShXSrJxo\n4cweQorL9Mw+M5iAxcclu/h8jaHgeZ9hAla0LosZuvq0IZVvoC/JNwB/XmCfDyNPgKokvkn5U6S1\nemmB/S7N7FfJAL6+Sfkho8/Hz+aXSBj5SuHbXfJ5FB/29GdIXKaqp5CUaRTfHbA7s3/VU0hKH3Bc\nkWkdRwBnXY+FQlK2UXws/HcDL48vO9XBF5D+kncW2O/8zH5fKHuOhvDt6tOKNMEPAh9HHueZTRzp\nXzmI1CmVXBnma+NtETLbII1M2NmIPDZrQ+a9Oxvh3PHn0RO+N/OnI+GVtzN8AuD2zPYxLSEZJ75L\nyaYFGSL1e/JOoKQEBV87mWZkfu7J/IwjMyVzE9yB9KSFnrcil9rshwNMQY/alUIifFUK34rPPUiX\nYfal2JVyC9KNuATpeD5A/g6hcuCblHXIg0OycaXkTvf6bmb/SuHbXfJ8pD1SDFupbPEpG4Uq2iaG\nHkbicgA4g5HhzY4wcoFUVVJIyiGGrj4uaYY/XM1lOhV8WlM5KVR8NlG4K9JlKRMgLCvIRJxi+l4/\nmNmvYg8RwcerTxyJLJpA5rzNyfn8JGSeWxKZGJjvWTzlwNdm/lxkyoW78usgMi/lIEMNt81U/sGv\nvt/7NCBPpHwKWRHWn/n5FPD3VPYMcfFdShAJxETACYNJUTApCiZFwaQomBQFk6JgUhRMioJJUTAp\nCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRM\nioJJUTApCiZFwaQo+C0ljkQu3YjMxf1KZvsiZJnMBuB3VH4lq++4AfNqkSc3/DHyPEM3OOd70J9v\nGOopo258hBgSCO8g0MVQKMbJSNzKCUUUKT5HkXhMIGsAdiBPqtsJnKgcF8qQZi5p4EzkzPg5sibx\nc8BNyOO2rkIeBrk098AwhjTT+DxwG7LKzCWCvrgqtHXKFIaiczUgZ8NGhocouQiJ5DVhOB14FhHx\nPPIMIZCHy7qX6qeRUIq52CoOhdAWn0BiUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZF\nwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJ\nUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6Jg\nUhRMioJJUTApCiZFwU8p+cKZgTxxdwvyZO+vVj5r/qKFM7sQWA3UZT6bmufY0AeVaUSC3J0G/CcS\ncqgQoQ0q44Yz24sEuHsBOAW4ADlz1iKhiCqK3yHNtHBmtUArcD4SHvFR5JHnI9BCmt17x1cuSG3c\nvbSB2tpjDAzEzpy5+pNf+uwvoPiQZpFSfLMS8XngGHAxcA/yKHOQSF7nIY83z8YhJ/+nx0647vT6\nmd+Y3zCtpS5SQ78zyMvHuo92Jnff2pnatVL5myPSAH/PlCnAAHCIoXBmdyKRAS9CpJyChE/MFaIy\no67lrtMbT5i0rO2db37R7x34zaR96SN3dqbQpKj4KWUG8H2kXokCDwH/B/wCiQDYCaSAa4tNsL2m\neVq2EIBlbe+MPHds5zQvGfNTSidwtrK9H/jIWBKMRWpwnJEXlFikxlM14ffVp6QcSSd6dw8cjgyQ\nfvO1e+Bw5Eg60eslnVBJ2ZE4sOXB/U+n3xjoYf9AD28M9PDg/qfTOxIHRg20mYvfl+SSMr9heoPj\nOH3f3Le6sTEaoy+dojXS2De/4fj4+v4dRacTKin7B48ef3L9lMbbp1z+Zh3ynTeebHyxr2u6l3RC\nVXwm1zS23tC+ZFilekP7ksikmsbJ+Y7RCJWUeDTW15tOptOO47iv3nQy3RCNHfOSTqiKTzLd37st\nua/x1VR3fW0kyoCTZm5sWiqRTvV4SSdUUl5I7Fm9pmfzRz817bI3Gytf7/553eZE1xov6YRKymkN\nM5uvnHT26xv6Xp9ZE4lEBx0nfeWks3fvHDjUtCm1u+h0QiWlKRqrn1s/dQcSSf1NmqP1cS/phKqi\n7U2nktr2nnQy4SWdUEnZnNizYnn3qrrsbcu7V9VtSXQVfYcMISs+ncldqwB2DBy8rjlaH+9JJxNb\nEl0r3e0TgdD20QYSk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6Jg\nUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQo\nmBQFk6JgUhRMioJJUTApCiZFwaQo+C1ltLBmALciIYraxpD2knHlzGe0sGYAJwI/A15Dl1JoZVhH\nEX87sCvD+jI/Y0ANcCDz/pvAp/3IUBCk5IY12wx8ANgJPO9jvgLBJKT4XJ75eVxm+2tAu7L/RuT0\nH89ro5aRIIU0Awlr5iCxI91iNQvYBSwCun3KV0WZArixTRqQyF0X5+yTr6ItG34v1s4X1iybQAe9\nNAwjVJzJyNi11yCVczdwH9Io9ETN+PPlK/cj38GNSLoQeAJYn9l2LVKBP6UeHVJ2IXFrXe4Cnst6\n/9dInOwJRQK5cXRZC3wx6/08JJSrJ4Jw7zMe9gEnZH6vBc5BuiJcYkjXw4TiYaQOmYvcUfcAzVmf\nX8nw4jQheAsS7jmNREG+Mefz/0W6ICYcdcil+QTlszPR77ANrwSt62CsRIAPI3fY05ALiBvp3AHe\n7yUxv++SS8XXgH9Eeu72MPzOesLeZe8FripVYtXeTnGJAhtKlVi13/u4tCIPN1lbisTCUqdMAv4K\neUrD80jsfRiqaG/yklhYrj5rs37PrlhdKRdWNDcBoBa4AukEN7JIAieVKrGwXH2eQ7oJSkJYrj47\nga8j7ZVepEg1ZL08hYgPS0U7Wp+Jg8d/fhguyTHg98gzgbb6nJdA0Y08NaokhKVOOR55gNLqUiQW\nhuIDMhvqGqRFux6pbGGMLdqwSFkIPJv5PfvxfK4UwzAMI0j8P17qpQ3qg9O7AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7bd61cfa10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "pj.report_seq_stats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define minima and maxima\n",
    "\n",
    "gc_inliers = {\n",
    "    'rrnS': [40,46]\n",
    "}\n",
    "len_inliers = {\n",
    "    'rrnS': [400,1000]\n",
    "}\n",
    "\n",
    "\n",
    "# Apply to loci data\n",
    "for locus in gc_inliers:\n",
    "    # trim GC outliers\n",
    "    pj.filter_by_gc_content(locus,\n",
    "                            min_percent_gc=gc_inliers[locus][0],\n",
    "                            max_percent_gc=gc_inliers[locus][1])\n",
    "\n",
    "    # trim length outlier\n",
    "    pj.filter_by_seq_length(locus,\n",
    "                            min_length=len_inliers[locus][0],\n",
    "                            max_length=len_inliers[locus][1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Confirm that the length filter has worked."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distribution Of Sequence Lengths\n",
      "Distribution Of Sequence Statistic \"Gc_Content\"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAFEAAAIiCAYAAACwpiRwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEptJREFUeJztnX20HHV9h59LQgwJSSCBEFqgCW8NgZYALS9qSwigGKlo\nqTWlIrYcm2pbwLby4vGUq2KrUBW0VqKUF6vSUlGBY6AtPRJ6PJYQNHBjCCVAeAmQ8NJEibyZ3P7x\nnT07d+/s3jt7P3d3ZufznDMnszOzs7/7ZF5+718wxhhjjDFCrgU2AwOpbTOB/wT+F/gPYI/UvkuA\nh4H1wFtS249JzvEwcNU4preQ/BZwFEMlXg5cmKxfBHw6WV8ArAF2BeYCG4C+ZN8q4NhkfQVw2ril\nuKDMZajE9cA+yfqc5DPEVXhR6rg7gOOBfYEHU9uXAlePR0JbsUunf3AE9iFucZJ/a0J/CXgqddxT\nwC9nbN+UbO8oRZOYZjBZJEydOnVD6pztLmuyzl00iZuJ2xjiVt2SrG8C9k8dtx9xBW5K1tPbN2Wd\nePv27QcRz9Fmy8dH2N8HHJl17qJJvBU4J1k/B/huavtSYBIwDziEeKE8C/wUOI74I89OfacS3Ag8\nDbwGPAn8EZHFuZPsLM5HibfyeuCtqe21LM4G4Astfm+kR0P/KNIse7yUlZEELBKco+dRCMg8R9Ge\niaXEEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBE7udgCLQ398/mFrva3Vs\nFr4SBVhiinauQrBECZYowBJTTIOfdDsNRadpk+lM2DwID82sd6bKdQ5fiQIsUYAlCrBEAZYowBIF\nWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUUBbTYQlZZCMv3dwcHBIlX9fX18rJ5nn8JUo\noPLdSPr6+vpmwuYXYOusoaNaR42vRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkC\nLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUUFSJ5xPxBNYm69Be6KXKcgQh\ncDIwgRB3EPlCL2VdHJWaA2I+cA/wCrADWAmcCbwDuCE55gbgncn6GUREjdeBjYTEY+kgRZS4lggC\nNhOYAiwhgtXkDb3UMYrYU3Y98BniubeduFV3NBwzUuilzH3HHnvszan1dcuWLXsQ4NRjjpm8Y9Kk\naZNee23ywH33nVU7Zvny5YetWrVqQe3zqlWr8v4theFTwAcJuenQS7VIahcnS407iEhBjVTqmQgw\nO/n3AOB3gW+SP/RS5bmbmHJqDXBSsq2d0Etpxu1KrBKVu51LhSUKsEQBlijAEgVYogBLFGCJAixR\ngCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBL\nFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijA\nEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUK\nKKrES4jYAwNE8IY34PBKuZgLPEqIA/hXIuqFwyvl4KdEqKQpRNCdKcDTOLxSLl4EPgs8QcjbStzG\nDq+Ug4OAC4hbcxvwb8B7G45xeKUReA9wTerz2cCXgAdxeKVRsx44HtgN6ANOAdYBt+HwSrm4kHoW\n5wbizevwSgWgUrdz6bBEAZYowBIFWKIASxRgiQIsUYAlCrBEAe1Uhe1JxF8eBJ4D/k+aohIyGokT\niNrjdwMnUq+OqvEMsJKo97sF2KlMYBloJXEiESb9YqL+bjuwGlgBvEBUU80EDiaq7pcSQv8O+DLD\n49RXkocIcf8ELCKuyGZMABYD1yXfWd/i2G7Rlaqwz1Nvx8jDHODKNr433rg+UYDrE4tMO1mc44B3\nEe0ZEA3t3wXuUSWql5kAXEtkYbKW62n98uk2hXgmXkrI+jbRGjcjWU4AvpPs6x/rj4wjhZD4ONHK\nlkUf0Uvh8bH+yDhSiBfLbKJE0uzkt9Belqj05JH4MMOLfGnmEBl004I/ICobFmbsOyrZt7SjKcrH\nuN3OebI4hxLZmXuJ59+DyfYFRFePB5Jj/qbhe5/I8Rs9T7OszUhLUSjElXhgjmMrRR6JG8crEWWn\n3U6eU4BfSdYfB36uSU45yVsBcThwO9GD9SfJsjXZdoQ2aeUhz5V4FNEMMJXhb+e3AG8img9+rExg\nr3EncdUdnbHvaOLqvLOjKcpHIcrOLwGXtdj/KaJpoKgUouz8CtEQ1YxngJdznK9nyCNxBdGq14zf\nSY6pHHkk/iUwC/gWMWJpWrIcB9xMNJ9+WJ3AsrOTaDvembHebClyW3NXin1fG03KRvMjvU4rie/v\nVCLKjptMBbSSuOcYzjuW75aOVhI3Ei18s3KcbzbwSeCxMaSpp/ggsAV4lWgS/QBwJLB76phpRJHv\nz4DvEQO3nwWWdTSlo6Nrxb4ZxPwKGxmalXk1WdLbNgAfIcQWkXGT2DfKBOxCZLBPJGpt9qbeU3Yt\n8H3gvlGeq1sM0uTvnQmbX4Cts2CPF1s3+2aeY7RVYTuB/0kW04CzOAIsUYAlCrBEAZYowBIFWKKA\nvI33uxCdlw4mytRZmVd3YGrBIcQkP2XpwNRIITo0fZHo1HQhUcx7Icd3TcJ24DPdTsQYKES786tE\nJ0/TQB6J/070tzEN5G13PgH4a2KGOJPQqj7xMYY/A3YH9iLal59maDtzX3J8UXvUdqU+sZ2BPW53\nbmBRpxJRdvI8E3+baBZoxt7JMZUjj8S7gFNb7D+ZyIRXDmUFxAQq+kxUSjwBeF54vtIwUtn5fGLO\n69oVdiXZXY5nAtOJQeWVYySJ26hndeYSV9qWhmMGiaEYPyRmMDEt2EjM1FRWClEVNjfHsZXCzQMC\n8kis9clutbxE1H5fRUSnqAR5JH6NmMa+j5im4NZkeSjZNkAET9gB/AURaKaolRFdYzExLO1tGfuW\nEG/yk5PPbwdeA77Rxu/8KjE+sLZsA85j7OGVCjEs7YfAFS32X8HQXmNX03oE1mjYJTnH/vRIeKUj\nad088Bjw66nPa8jXVTmLUwgpT9Ij4ZW2EX9UM04m4kvVmJ58ZywsJQRBj4RXupEoAi4HPkf8jw8S\nM5B8mJiE7arU8ScRb+p2mUSMF7woY19pwytNITq31xrpX0+W2ufbk2MAJhPToma9hEbLGcTbvsZ6\nChpeqR2WEHPG3pEsVyfb1PwL9XBKEC+W2lV5McNfLLXwSo+Q3ZZSKImdYCpR2ZEeieDwSgWgEBUQ\nENO4LKN1r7DFOc9ZevJIfBsx7emuRBn5xYxjKn+5j8QaYBPwG91OSJsUosQyn2geWJ3jO5Ugj8Tn\niZ5hpoG8VWFnjldCykyeF8v1RFHuVqJ49yjZE2c8MfZklYs8EtNBGU5vcswgxZ5re1zII3E0owKc\nxelxCpHFMU3IK3E6MbnGD4g2jROS7XsRsxrP1yWtPOR5Ju5NyKtVNx1ERAOHGNNyDjGVS+XmC8sj\n8TKiSv54on9Ouk/OIJH1qVzlA+S7nU8nKmObTZjxKHDAmFNUQvJI3It4DjZjJ9EsUDnySNxMPAeb\nsZAKllYgn8TvAeeS3cfmOOB9NA81YhL2JeoTnwW+Qty+NxANSq8RDedjbawfTwrTxnIAcbU1zvB5\nG7Cf4gfGkcK0sTxBtAfPIDoe9RGtbJUe+9xu7IFtwNi7A/QILjsLaHUl7qTFyMwGase5PrEBz248\nSjy7sQA/EwVYogBLFGCJAixRgCUKsEQBecrO7U6ScXeb3ysNeSTelfybVRRsVjysRDEwj8RziYGP\nhxBj9tJx+84iOqR/UZq6kpBH4nSi7Xk+UcOd5pPE2L/pwBc0SetNHgY+1mL/x4irsagUoi/O/rQO\nbvhz3O48IhuBs8luW94t2bdx7EkqH3meiZ8lBkeuBr5EvdPnYURQm8OAP5Wmrke5gLilG2c03k5M\nUllkuh7UJs0exBQBtfkdHiGmFdjaxrk6SdeD2qTZCtzUxvd6lnYkziNG4M8mMt0biWGyc4jbwWNd\nRuBy4BfUez7U+iPOIJ6LRe7gWYh84jJiZuN/IJ6J6WfDNqJ7SbOhGT1Nntv5Q8Qo0wuIvoqNDBDR\n1CpHnivxUGLEezOeI1tuz5M3XPvUFvsPoPjZnHEhj8R7iWlbsphMFPt+MOYUlZA8Ei8H3gh8nfpM\nTPsCpwEriQqKv5emrkf5E+K2biz2vUzxu50UppPnV4hesb9HVDj0EXWINzG8otb0IIW5EtPsCvwm\nMbnZOmKGY5PBIqLNpLFmYx7wAEM7v1/X0ZTlp2ujB64nu91kJSHvv4nK2oHk8/vb+ZEO0TWJ64iy\ncpr5hLCVqW27EXPL/lc7P9IhulYBMYfhV+Ki5N9rUtteBr7J0Jk8K8NIEt9ACEpTm2p0ZcP2Jxk6\nk1xlGEnik8DhDdveTIx1bhwMOQWXnTO5mxj4+GvJ53cRM9bdnnHsETjDncmBRI31DuLq20lU/x/R\ncNwEInpa40uoSHTtxfIoUdG6gpjqb0XyeW3DcYuT/R6q2+MUoo3FNMESBViiAEsUYIkCLFGAJQoo\nqsQ9gG8RIxTWEfPujDUyUOW4AfjjZH0i0WGqsJGBisgMsiMQrafeTDGHenfnSxgas+UOYma9RipV\nYplH9Ou5DvgR8FWi+0pPRAbqFBOBo4E/J7quXMnQoDVQ4shAnWIOETSsxpuJyd4epKCRgYp4Oz9L\n1Kgfmnw+hWjTvo16pKBziL6SEDOILqUeGegQPHsUEOHt7gXuB75NvGwcGagAVOp2Lh2WKMASBVii\nAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGW\nKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWCJzX3z/74/39h458ZDaW\nmOJ1eL6d71liip8NnytyVFiiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMAS\nBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAFFlbgReAD4MfVgDIUNr1RUiYNE\nRPOjqAflvpiQeCgRFr4WPmQB8J7k39OAf6TDf1dRJQL0NXx+BxG7iuTfdybrZwA3Aq8TV/AG6uI7\nQlElDhJhQ1YDH0i2ObxSTt4EPAPsTdzC6xv2S8Mr3XxzbB4YGDgrfXyZwys1cinwV4TIcQmv1N/f\nP9jf3z+agDWliccyBZiWrE8l3rYDRBilQoZXKuLtvA/wnWR9IvANIkuzGrgJOJd4gfx+csy6ZPs6\n4BfAh+hwGKQiSnwMWJix/UUi6FcWf5ssXaGIt3PpsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijA\nEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUK\nsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJ\nAixRgCUKsEQBlijAEgVYYopRzqE4DEs0uVDMqViamTxLhyUKsEQBlijAEgVYogBLFGCJAixRgCUK\nsEQBRZY4gQivdFvyebzDKy1q83uFlng+EZShVugf7/BKi9pNaFEl7gcsAa6hHmbJ4ZVy8nngI8DO\n1LbChlcqosTTgS3E87Ax2FeNdsIr3Z/6XtZy6Qj7B5NzDKOI8VjeSNy6S4DJwHTgn4mrbw7wLBFe\naUty/CZg/9T390u2NZIV46USnEj97Xw5cFGyfjHw6WR9AbCGenilR2h+BVeSE4kYVBBZnDvJzuJ8\nlHihrAfe2skEGmNMiVkILG7Y9l4iQtsWYDnxgsrFhLGnq1RcQ/zNK5PPC4DbgfuSbe8j8s4rM79t\ngMg/Hp/6/AmGZqDPBdZ2NEUl5BWGZszvAi5LfT4Y+Fnekxax2DeePEe9XD0ROAa4J7V/EkPL6yaD\nrxPPwAOBC4GXgN1T+8+kSfnY1JlHlGx2Ug8em+YW4HOdTlQZ2ZXI6mRVly0EZnU2OQaobm1HH9Gk\ncDIwm3jBDibbB4mquFFTxPrETnA5cAHwfeAZhlbidjTAdpnZDLxbdbKq5RNr7EI0P0ioWtm5xp7A\n4USJZcxU9Zk4A/hD4FTgAaK5FeovlvPynKyqb+e7UuvpF0lN4kkdTU0JmQi8Hdir2wkpO68Cc1Un\nq+rb+X6i2ktCVd/OTwFXEPnF7cQtvltqeTnPyar6YmlVZzhIzourilmcScBq4Fqisd+0yRain6OE\nqj4T9wGOJjqNjpkq3s4AU4j25lOJ5tLtyfa2SixVlbgA+FGyfmBqe02iMcYY0z3+H7DzMkY+mwxc\nAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7bd61f1c90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEUAAAIiCAYAAACe4wXiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFJJJREFUeJztnXt0HNV5wH+7etoS2JawEcWAMWCw0wLBLaHtITbESsnj\n8GjqkpLUPNLgBk6g4R1oi3ikBUMePYQkhAAhJLQhOCHhUBLbBJMn+MRgI2IbavMw4LclG1vGeu32\nj28WjVbf7uysZmdmR9/vnD1ezc5eXf08c++d7975BgzDMAzDMACoAV4AHndt+zywFngJuD3sCtWG\n/QsVLgfWAAc4P58GnAkcD/QDkyOqV2RMBZYhInJHyiPA6ZHVCEhH+cuBrwJXAxnXtmOADwLPAsuB\nPw+7UlFK+TiwDWlPUq7ttcAk4BRE2CPal5uamtYD2VG+VgX8N42a/wDeBF4DNgM9wEPAk8Ac137r\ngVbl+1mP8jtKqINXGZEyh6E2ZSFwk/N+BrCxwHcqJiUOvU+OXAXvd16dQB+wILIaVSFeR8rcAMqo\nOoL4g9Qyou6SY4lJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhTiFI4MhOMb\np86Y1dA2ryndUN+T6e1b07tl2Yv733rFTxmJknJ849QZ7c0zL7hqcntfbtud25deAHzXj5hEnT6z\nGtrmuYUAXDW5vW9mQ9s8P+UkSkpTuqFe295cYHshEiWlJ9Pbp23fW2B7IRIlZU3vlmV3bl867Ki4\nc/vS+rW9W5b5KSflvUtsyaLU//jGqTNmNrTNa0431O/N9PatLd77qGVUMzbvEyaJGqeADd5GYIM3\nBRu8KdjgTcEGbwo2eLPBm4oN3sLEpCiYFIVEjWjBhvkjsGG+QlDD/EQdKU3phvoX3t3Yur5v+/Ra\namoGGBw8un7yq83phl1+ykmUlC39uye+0dd1wvwJs9+7Veax3atO2Ny/e6efchIlpS5dS1OqYfyS\nPWsOTKdS6Uw2m2lKNbxTl67xNdBLlJSabKpuV2Zf9kPNx7237am967K12bSvq+RESalN1xw1f+Ls\nHuTeIQDmT5zNb/atn+6rnMBrFiE1pDasfvetow6rmzQB6Vkzb/Z3764l/aqfchIlZV+2v38wm0l1\nDfaQJkWGLIPZTKon2+crnpIoKf2ZATb2d+07e/yJe3PbHtu9Kt2fGfQVHkiUlLa6CbuOqG9Z/aPd\nK4eNUw6pmzB2xyk9md6+9487fOf7xx0+bFzyYPezFo50b7NwJBaOrFg4MlFtClg8ZQRBxVMSJWVW\nQ9u8I+tap355+9KTGlK1Nb3ZgcEj61qfn9nQNm/MStk9sO/Pdg72zL1ycvt7oYNv7/z13HcG9vUU\n+14+ieqSD6wZ/xcXt57qTjvCxa2nZg6oGecr3UiipDSma7u6BnqGdbFdAz2pcem6bj/lJOr02Z8d\n6BqXrtvSNdAzCecqeVy6rvvdbL+vyFuijpQ1+zfff/eO5QMttU2bWmqb3mqpbdp0947lA2v3b3nA\nTzmJOlI6e99eAvDmQPeFzemGxr2Z3v1r9295ILd9LGAT7GGSqNMHbJg/Aps2VbDVkQq2OlLBVkcq\nWDjSwpEqNngLE5OiEAcp+emcO4C3nG0vAGeEXaE4jGjz0zlnga84r0iI+kiZCnwU+A5DvUCKiHuE\nqKVo6ZyzSDb01cB9wMSwKxXl6eNO5zzXtf2bwM3O+1uALwOf0Qo4+eSTF7ver1m4cOHaYr/wnnvu\nmblixYpZuZ9XrFhRVsUriZbO+Xt5+0xDkvVqVGycEhfc6ZwPcW3/AvBwge8kfi45xVAFFwEnOD+/\nhuS8NkrEhvlhYlIUTIqCSVEwKQomRcGkKJgUhbiMaAPDpk3zsGlTBZs2VbBpUwWbNlWwaVObNlWx\neEqYmBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBSFRE2Ggc0QjsBm\nCBVshlDBZggVbIZQwWYIbYZQxWYIw8SkKJgUBZOiYFIUTIqCSVEwKQomRcGkKJgUBZOiYFIUTIqC\nSVEwKQomRcGkKJgUhUTNEObo6OjIut77DkzbkaKQaCnlHCWQcCnlYlIUTIqCSVEwKQomRcGkKJgU\nBZOiYFIU4iAlP51zjiuRPLUtYVcoDlJy6Zzdq4oOA9qBN6KoUNRStHTOIPmtr4mkRkQvRUvnfBaS\nDf3FSGpE/NI5jweuR06dHAUv/wulc168WDZ3dnae596/WtM5Pwpsdba9BvQDrwNTlO8XXB3Z0dGR\ndUffihC7zMXXOy+QdM5XAX+Xt89rwGygK8R6Rd6muNH+1yJJzB2XwPUzziuf6WFXBOJ1pMQGk6Jg\nUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQo+A0yjQeORWKmWWA78DKwL+B6RUop\nUlqAC4D5SLw0/zv9wErgR8B3ge7gqhcNxaRMBP4NuARoQI6IHwAbgJ3I1EMLcDRwCvJkyS8B30Ce\nNLm7YrWuMMWkrAd6kamI7yOR9WIcBXwaeW7gBcBBAdQvEopJuQX4FiKmFDYANwG3A/88ynpFSjEp\n/1VmmfuBr5X53VhgXbLCaKTMB34O/BFYAvxjIDWKAeVKuRR4EBmn/AzoQ7rja4OpVrR4jVPSDF8m\nkeNzwEXA/7i2fdXZfnswVYsOryPleWTAlk8TsobEzdtAcxCVihovKb8CfocMzMa5tj+OnC4XIWtJ\nrgT+lZHr1qoSr9PnMmQUey9wDnJ6/AJpO+qArwONSJvy387+VU8pDe1zyCl0H/ATZHTbhAhqBtqc\nny8E9lSmmuFSau/Tj1zXnIgs3lsHLEAa4W3AYEVqFxF+u+RXgNOA65BR6xLgyKArFTWlSPlb5JT5\nMbIEqw5Z4jkL2AW8hKxwTMzo2OsPuQxZnDcVCRXcDDzmfLYF+Hvgk85+fwBOqkw1w8VLyuVIfGQu\n0vvMB85g+LKrx5Gj5rfA74OvYvh4STkIOT1y/NH5N3+9/B7g88gqx6rHa5zye+ALyNr5biQS9w7S\n+2g8G1zVosNLyqXI2GS583MXcD6wt4J1ihwvKRuQsckxyDB/HRJESjSlRPMzSNB6zFCsoZ00inJH\n893IKSbldeBGoNVHeVOQgLdX5D/WFJNyHdLQbkIa288CJzA8ZnIAMmC7FHgCial8liqPwBVrU74J\nPIxMhi1Ebk7K0e/8W+fa9ipyV8a3qPKrZa+Gdjfwn0iI8WRkcDYLmMzQXPJLwNPI1GkiKHWCPYMM\nzBIxOPMiMVe2QWJSFEyKgklRMCkKJkXBj5QHgA8U+fxk4P7RVSce+JFyPrJaqRDTkRVMVU+Qp08T\nQ8P/qsZrRHuE88rlG5gJfFDZrxWZMVwfXNWiw0vKhcC/u36+wXlpZJAJ96rHS8pjSFwFpBH9NiOv\nf7JIzHYFktCh6vGSssp5AUwDFgOdlaxQHPCzDL2jUpWIG37X5qeBecgq61b0hC83j7ZSUeNHyjHA\nT4HjPPYbU1LuQgZo1yCRtp0VqVEM8CPlVGQV9p0Vqkts8DOi7UWC04nHj5RfAH9dqYrECT9SrgD+\nElnN5Ov5w9WGnzbld8hF3yJk2mMTwxcAppDRbSR5lILEj5Q38H6UbiTZtoLGj5S5lapE3LBwpEI5\nUuYgC43vZWh024zEWap6CUYOP1JqgEeQ0ewXkdjJnzifDSJhhksCrV1E+JFyLbLQ+AokAuducN9F\nlmt8JLiqRYcfKQuAh5Dl59p1zzrk6tkv+emcbwFWI3Gcp5AsxqHiR8o0ZKxSiF2U16bkp3NehCwO\nOhE5JW8so8xR4UfKHoon4T4KWa/iBy2ds3vBTzOww2eZo8aPlN8gd6hr35mENLxP+/z9WjpnkN5t\nIzLXdJvPMkeNn8Hbl5D1979EbpUDOcRnIOvjmvH3B2jpnHPkZg2uQ8RdqBUQl3TOH0Pu3sjkvbYA\nH/ZZlpbO+Xt5+xzO8HsD3MQmnfMTSIPbzlC3/AoSVvCbQ0VL57wACXv+n7P9LORIChU/Ug5HGr19\nSPeZf2fpeOSuj41l1CN3hQ1yBX4sMiDcgMw8hoofKa8jDe3DBT4/E7kztaaMeixn6KaI/OTfoRPk\nBWFiLi6D/EOOQwZwVY/X6XO+88pxA/BPyn6twJ8i1z9Vj5eUSQwPL05GQpJuchPs91F4RUJV4SXl\nawxlzckgt879oKI1igF+ep/ENKRejJk/1A9+pfwDEj7YzvBh/qDr36rHz+lzNXKLyw5kNZMWaBpz\nUxyXImlDTkfCj4nFz+nThoQjEy0E/EnZgOSTTDx+pNyJjGYPqFBdYoOfNiWDPM9rLbJO/1X03iY/\nUFR1+JHygOt9oeF8ljEm5fSK1SJm+JGyvFKViBujGeYfRBUn4i2GXymHIm3GbmR6YhuSbOZB57NE\n4Ddw/RxwMDLXm0tFNAtJu9qO3DlW9Tct+JFyCzJ4+zjwv3mffQSJut3K8EhdVeLn9PkwkhQiXwjA\nk0iWr78JolJR40fKJGTiqxDrGYMrmd5GUiQW4lRG5qitSvxIeQRJfncbMMG1fQIyq3cu8MPgqhYd\nfhraW5Gj4RokKe8mZ/uhiNzfOvtUPX6OlB7k9FkILEXmlPchk+sXO58l4kElflcd9CNLRe+tQF1i\ng0XzFUyKgklRMCkKJkXBpCiUIuVzSC7rYpyLjF8SgZeUc4C78V6h1I1cQZ/lsV9V4CXlU8i88RKP\n/ZYgw/yqj6WAt5QPoMdPNH6O5GWqerykTKH0cMAmZ/+qx0vKPuDAEss6kIRcEHpJWU/pufBPZWj5\neFXjJeVxpEf5K4/9TgHOJiEPPfKScheylOsJJGbSkPd5I5K++UlkDuiuoCsYBV5SupEjpQ9J07wL\nubfvV8jdFd3APc7nZ5GAh71CaUGmFcg9fVcDnwCOd322EUletQhZppEISo28bUHislcii3YORJ7J\nUdUJvgvhNxwJIiKRMnKUIuUQ59/Nzr+NyErJ/OWhbyIPkq56vKQci0yk38DQkyibgTuUfQeQRjg2\nY5UW2NoPO/bA+4Is9zakq3V3xQch69+uQO4SnYtMb3QR7u2yBRcyt8DWLLychZdbincAZd1YeTry\nQFftQdKrGL666YckZAmY1zjlGEq/23Ods3/V43WkNDHyqU9dyFglP73ZO4y8Qaoq8ZKyi6HeJ0cG\n/QbqNmTZV9Xjdfq8ROl3preTkLSsXlIeRXoXr9jr2c5+j46+SvGnEcltsh9Z83ZE3ufTkOUXvch4\npjHEulWsSy6F6ch6/NydX91Izrduhu4MW0P4D36tmJRS5n1eRR6BdTmSQyWDJKjKAL9Gnmt6ElX+\nnDA3xbL7xZ2C2QlbYOtOZ66qFSZ2ydrfksuwaVMFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRM\nioJJUYiLlPzsxXcgMZzVwI8ZftNVxYmLlPzsxUuQWb0TkPsWvxhmZeIgRctevJShxL3POfuERhyk\nFMpenOMiSl+2GgjlLMUIkmLZi0Em9vsokNm0UObi9tmzGwfr6w8AqO/ra+xcufI8qFzm4qAplr34\nAmQVd6EZgkij+WExh6He5wxkyqRY1o1Io/lh4c5efBeyDmYpcmp9I+yKVCsWzQ8Tk6JgUhRMioJJ\nUTApCiZFwaQomBQFk6JgUhQSd+2TzWZHXPmmUqlCf6dd+5RK1JG3wEmlUqn8q2S/ZdiRomBSFEyK\ngklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWT\nomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyKgklRMCkKJkXB\npCiYFAWTomBSFEyKgklRMCkKiZRyWUfHlJs6OmaU+/1ESsnRChP7YYff7yXuvmQ3RbJ2FSXRR0q5\nmBQFk6JgUhTiIiU/nfN8JKHmIPKUqVCJi5T8dM6dwDnIE7xDJw5StHTO65Dc1pEQByle6ZxDJ+rB\nm1c656IUSue8eLFs7uzsPM+9fxLSOQM8TeGGtmDm4o6OjmxHR0cpqZpjmbn4euAw5MGOnwR+CSzI\n2yf0ZFpRS8kn9z93DnIEnQI8ATwZZiWiblPcPOO8AH7ivCIhbkdKLDApCiZFwaQomBQFk6JgUhRM\nioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQF\nk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZFwaQomBQFk6JgUhRMioJJUTApCiZF\nwaQomBQFk6JgUhRMikKipZSY72AEiZYyFinrKCilDDtSFEyKgklRMCkKJkXBpCiYFAWTomBSFEyK\ngklRMCkKJkXBpCiYFIU4SMlP5dwCLEWS9C4BJpZZ7txyKxQHKfmpnK9DpMwAnnJ+Loe5o65ZREwF\nlgGnMXSkrGPoyQhtzs8aXpG3jhJ+fywjb1oq54OBrc77rZT56IjREKUUdyrnQol4sxQ+Ila7Ptde\nN3p8nnXKiBVaKueHkNOlzdnnEAqfPolnDkNtyiLgWuf9dcBtkdQoBswBfua8b0Ea39F2yYZhGH45\nETg9b9unkR5tG3APUO+30JrR1ytSvoP8DbnU8rOQHPsrnW0LkLHYM+q3E8rbyAMHctzM8AHZZ4CX\nQq1RDNiPPLYix3LgVtfPRwN7/BYa9bXPaNkOHOq8rwVmA8+5Pq8nRo/ICYvvI23IdOAaYC/Q7Pr8\nE8Tw+qbSHAmsR46GAeCSvM9/Cnwl7ErFgTqkaz5U+exEoDXc6iSU0B8oVCFSwLnAh4ApSAeSdbZn\ngTP9FBan5/uMhkXAvyCP09rM8MBUEAsGq5KtyAMdA6Haxyk50khYMxCq/donxyTgfciIdtQkpU2Z\nAHwKaAdeBPqd7bmG9jI/hSWl91nueu9uWHNSTgu1NjGgFvgYcFDUFYkbvcC0oApLSu+zGgkTBEJS\nep+3gDuQ8UoPckqNc73e9VNYUhraYjGTLD7/85PQJdcDfwDuZwxPsWpsQ9azBEJS2pSDkecqLw2i\nsCScPgDjkfmedmR6o8fZXtaINilSZgHPO++nu7bnpBiGYRhx4v8BlO3Up+wHfI0AAAAASUVORK5C\nYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7bd6250690>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pj.report_seq_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Long Alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, need to produce a sequence alignment file based on an alignment of all 12S sequences longer than 400bp. This will be done with the AlnConf class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj = unpickle_pj('rep_post_SATIVA.pkl', git=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "muscle default configuration:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "muscle -in 815371477497805.69_rrnS.fasta\n"
     ]
    }
   ],
   "source": [
    "muscle_default = AlnConf(pj,\n",
    "                         method_name=\"muscleDefault-400bp-SATIVA\",\n",
    "                         program_name=\"muscle\",\n",
    "                         loci=['rrnS'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([muscle_default])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlnConf named muscleDefault-400bp-SATIVA with ID 815371477497805.69\n",
      "Loci: rrnS \n",
      "Created on: Wed Oct 26 16:03:25 2016\n",
      "Commands:\n",
      "rrnS: muscle -in 815371477497805.69_rrnS.fasta\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: MUSCLE v3.8.31\\nPal2Nal v14\n",
      " Program reference:\\nEdgar 2004: MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Research 32(5):1792-1797\\nMikita Suyama\n",
      " David Torrents\n",
      " and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34\n",
      " W609-W612.\n",
      "execution time:\n",
      "8.54806089401\n",
      "\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The dataset(s) rrnS were first aligned at the protein level using the program MUSCLE v3.8.31 [1].\n",
      "The resulting alignments served as guides to codon-align the DNA sequences using Pal2Nal v14 [2].\n",
      "\n",
      "Reference:\n",
      "[1]Edgar 2004: MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Research 32(5):1792-1797\n",
      "[2]Mikita Suyama, David Torrents, and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments.Nucleic Acids Res. 34, W609-W612.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['muscleDefault-400bp-SATIVA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@muscleDefault-400bp-SATIVA': <<class 'Bio.Align.MultipleSeqAlignment'> instance (62 records of length 1043, IUPACProtein()) at 7f7bd5d29f50>}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@muscleDefault-400bp-SATIVA_aln.fasta']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# record_id and source_organism are feature qualifiers in the SeqRecord object\n",
    "files = pj.write_alns(id=['record_id','source_organism'],\n",
    "                      format='fasta')\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build tree from long alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim the alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimal -in 339891477497877.55_rrnS@muscleDefault-400bp-SATIVA.fasta -gappyout\n"
     ]
    }
   ],
   "source": [
    "muscle_gappyout = TrimalConf(pj,                       # The Project\n",
    "\n",
    "                            method_name='gappyout',  # Any unique string \n",
    "                                                     # ('gappyout' is default)\n",
    "                            \n",
    "                            program_name='trimal',   # No alternatives in \n",
    "                                                     # this ReproPhylo version\n",
    "                            \n",
    "                            cmd='default',           # the default is trimal. \n",
    "                                                     # Change it here\n",
    "                                                     # or in pj.defaults['trimal']\n",
    "                            \n",
    "                            alns=['rrnS@muscleDefault-400bp-SATIVA'],  # 'all' by default\n",
    "                            \n",
    "                            trimal_commands={'gappyout': True} # By default, the gappyout algorithm is used.\n",
    "                           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.trim([muscle_gappyout])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimalConf named gappyout with ID 339891477497877.55\n",
      "Alignments: rrnS@muscleDefault-400bp-SATIVA \n",
      "Created on: Wed Oct 26 16:04:37 2016\n",
      "Commands:\n",
      "rrnS@muscleDefault-400bp-SATIVA@gappyout: trimal -in 339891477497877.55_rrnS@muscleDefault-400bp-SATIVA.fasta -gappyout\n",
      "\n",
      "Environment:Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: trimAl 1.2rev59\n",
      " Program reference: Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics 2009 25: 1972-1973.\n",
      "execution time:\n",
      "0.460474967957\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The alignment(s) rrnS@muscleDefault-400bp-SATIVA were trimmed using the program trimAl 1.2rev59 [1].\n",
      "\n",
      "Reference:\n",
      "Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics 2009 25: 1972-1973.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['gappyout']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@muscleDefault-400bp-SATIVA@gappyout': <<class 'Bio.Align.MultipleSeqAlignment'> instance (62 records of length 757, IUPACAmbiguousDNA()) at 7f7bd5a99410>}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.trimmed_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@muscleDefault-400bp-SATIVA@gappyout_trimmed_aln.fasta']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = pj.write_trimmed_alns(id=['record_id','source_organism'],\n",
    "                                   format='fasta')\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer Maximum Likelihood tree based on the clipped alignment. We are going to use RAxML. Configure RAxML for trimmed muscle alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 387341477497898.72_rrnS@muscleDefault-400bp-SATIVA@gappyout0 -p 206 -s 387341477497898.72_rrnS@muscleDefault-400bp-SATIVA@gappyout.fasta -T 5 -x 328 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-muscleDefault-400bp-SATIVA-trimmed',\n",
    "                  program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                  keepfiles=True,\n",
    "                  cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                  preset='fa',\n",
    "                  alns=['rrnS@muscleDefault-400bp-SATIVA@gappyout'],\n",
    "                  model='GAMMA',\n",
    "                  threads=5,\n",
    "                  cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named raxml-muscleDefault-400bp-SATIVA-trimmed with ID 387341477497898.72\n",
      "Alignments: rrnS@muscleDefault-400bp-SATIVA@gappyout \n",
      "Created on: Wed Oct 26 16:04:58 2016\n",
      "Commands:\n",
      "rrnS@muscleDefault-400bp-SATIVA@gappyout:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 387341477497898.72_rrnS@muscleDefault-400bp-SATIVA@gappyout0 -p 206 -s 387341477497898.72_rrnS@muscleDefault-400bp-SATIVA@gappyout.fasta -T 5 -x 328 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "134.670024157\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) rrnS@muscleDefault-400bp-SATIVA@gappyout using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-muscleDefault-400bp-SATIVA-trimmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualise Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Annotate final tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Cottus','Mus','Xenopus']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=None,\n",
    "            html='muscleDefault-trimmed-400bp-reptile-raxml-post_SATIVA.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a static copy of the tree (the last *.png with the expected alignment id) for later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *rrnS@muscleDefault-400bp-SATIVA@gappyout.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png rrnS_muscleDefault_trimmed_reptile_400bp_SATIVA_raxml.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**=======================================================================================================**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from reprophylo import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj = unpickle_pj('rep_post_SATIVA.pkl', git=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Identify the record IDs of the sequences shorter than 500bp\n",
    "- Write these sequences to new fasta file (make sure the sequence IDs are as Reprophylo expects them: 'id_f0')\n",
    "- Prepare a hmm profile for the alignment the tree is built on (either muscle 'or mafft)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@muscleDefault-400bp-SATIVA': <<class 'Bio.Align.MultipleSeqAlignment'> instance (62 records of length 1043, IUPACProtein()) at 7f7bd26668d0>}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning trimmed alignment object rrnS@muscleDefault-400bp-SATIVA@gappyout\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "out=open('400bp_reptile_align_SATIVA.fasta','w')\n",
    "\n",
    "AlignIO.write(pj.fta('rrnS@muscleDefault-400bp-SATIVA'), out, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check number of records in file containing short 12S rRNA sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total records: 33\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "records = list(SeqIO.parse(\"reptile_short_to_align_SATIVA.fasta\", \"fasta\"))\n",
    "print(\"Total records: %i\" % len(records))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that record IDs are in format that reprophylo expects for alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AY151988.1_f0\n",
      "KX080575.1_f0\n",
      "HG802830.1_f0\n",
      "AY151993.1_f0\n",
      "HG802772.1_f0\n",
      "DQ081074.1_f0\n",
      "U81335.1_f0\n",
      "KT030712.1_f0\n",
      "EF027342.1_f0\n",
      "AY151990.1_f0\n",
      "FJ039991.1_f0\n",
      "AB090021.1_f0\n",
      "HG802786.1_f0\n",
      "AY151992.1_f0\n",
      "AF080277.1_f0\n",
      "DQ081073.1_f0\n",
      "AF080334.1_f0\n",
      "HG802805.1_f0\n",
      "U81329.1_f0\n",
      "HQ681907.1_f0\n",
      "AF080298.1_f0\n",
      "KX694589.1_f0\n",
      "HQ681908.1_f0\n",
      "U81328.1_f0\n",
      "U81340.1_f0\n",
      "EU624267.1_f0\n",
      "KT030713.1_f0\n",
      "AB090043.1_f0\n",
      "KX080601.1_f0\n",
      "AF158461.1_f0\n",
      "KX080576.1_f0\n",
      "JQ287761.1_f0\n",
      "M13046.1_f0\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "short_seq = open(\"reptile_short_to_align_SATIVA.fasta\", \"rU\")\n",
    "sequences = []\n",
    "\n",
    "for r in SeqIO.parse(short_seq, \"fasta\"):\n",
    "    print(r.id)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If record IDs were not in the format Reprophylo expects for tree construction, the code cells below could be used to format them to be 'id_f0'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#temp_seq = []\n",
    "\n",
    "#for r in pj.records:\n",
    "    #if len(r.seq) < 500:\n",
    "        #temp_seq.append(r)\n",
    "\n",
    "#from Bio import SeqIO\n",
    "\n",
    "#out=open('short_to_align.fasta','w')\n",
    "#SeqIO.write(temp_seq, out, 'fasta')\n",
    "#out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#from Bio import SeqIO\n",
    "\n",
    "#short_seq = open(\"short_to_align.fasta\", \"rU\")\n",
    "#sequences = []\n",
    "\n",
    "#for r in SeqIO.parse(short_seq, \"fasta\"):\n",
    "    #print(r.id)\n",
    "    #r.description = r.id+'_f0'\n",
    "    #r.id = r.description\n",
    "    #print r.description\n",
    "    #sequences.append(r)\n",
    "\n",
    "\n",
    "#out = open(\"seq_to_align.fasta\", \"w\")\n",
    "#SeqIO.write(sequences, out, \"fasta\")\n",
    "#out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that sequences in reprophylo project have not been altered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AY151988.1\n",
      "Podarcis muralis isolate E2106124 12S ribosomal RNA gene, partial sequence; mitochondrial gene for mitochondrial product.\n",
      "KX080575.1\n",
      "Podarcis muralis isolate 509 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "HG802830.1\n",
      "Emys orbicularis partial 12S rRNA gene, specimen voucher MTD:T:1488.\n",
      "AY151993.1\n",
      "Zootoca vivipara 12S ribosomal RNA gene, partial sequence; mitochondrial gene for mitochondrial product.\n",
      "HG802772.1\n",
      "Emys orbicularis partial 12S rRNA gene, specimen voucher MTD:T:3244.\n",
      "DQ081074.1\n",
      "Podarcis muralis isolate MTA1 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "U81335.1\n",
      "Chelydra serpentina 12S mitochondrial ribosomal RNA, small subunit, mitochondrial gene, partial sequence.\n",
      "KT030712.1\n",
      "Podarcis muralis isolate Mur01 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "EF027342.1\n",
      "Podarcis muralis isolate 1 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "AY151990.1\n",
      "Podarcis muralis isolate E2106126 12S ribosomal RNA gene, partial sequence; mitochondrial gene for mitochondrial product.\n",
      "FJ039991.1\n",
      "Lepidochelys kempii isolate LK_Atlantic 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "AB090021.1\n",
      "Emys orbicularis mitochondrial gene for 12S rRNA, partial sequence.\n",
      "HG802786.1\n",
      "Emys orbicularis partial 12S rRNA gene, specimen voucher MTD:T:6420.\n",
      "AY151992.1\n",
      "Podarcis muralis isolate E2106128 12S ribosomal RNA gene, partial sequence; mitochondrial gene for mitochondrial product.\n",
      "AF080277.1\n",
      "Podarcis muralis 12S ribosomal RNA gene, mitochondrial gene for mitochondrial RNA, partial sequence.\n",
      "DQ081073.1\n",
      "Podarcis muralis isolate Gua1 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "AF080334.1\n",
      "Lacerta vivipara 12S ribosomal RNA gene, mitochondrial gene for mitochondrial RNA, partial sequence.\n",
      "HG802805.1\n",
      "Emys orbicularis partial 12S rRNA gene, specimen voucher MTD:T:3816.\n",
      "U81329.1\n",
      "Chelonia mydas 12S mitochondrial ribosomal RNA, small subunit, mitochondrial gene, partial sequence.\n",
      "HQ681907.1\n",
      "Emys orbicularis orbicularis isolate 219 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "AF080298.1\n",
      "Lacerta agilis 12S ribosomal RNA gene, mitochondrial gene for mitochondrial RNA, partial sequence.\n",
      "KX694589.1\n",
      "Coronella austriaca voucher ROM_26536 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "HQ681908.1\n",
      "Emys orbicularis orbicularis isolate 255 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "U81328.1\n",
      "Trachemys scripta 12S mitochondrial ribosomal RNA, small subunit, mitochondrial gene, partial sequence.\n",
      "U81340.1\n",
      "Dermochelys coriacea 12S mitochondrial ribosomal RNA, small subunit, mitochondrial gene, partial sequence.\n",
      "EU624267.1\n",
      "Vipera berus isolate 199 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "KT030713.1\n",
      "Podarcis muralis isolate Mur02 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "AB090043.1\n",
      "Mauremys caspica mitochondrial gene for 12S rRNA, partial sequence.\n",
      "KX080601.1\n",
      "Lacerta agilis isolate S10401 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "AF158461.1\n",
      "Natrix natrix 12S ribosomal RNA gene, partial sequence; mitochondrial gene for mitochondrial product.\n",
      "KX080576.1\n",
      "Podarcis muralis isolate 5937 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "JQ287761.1\n",
      "Mus musculus isolate ID5.4 12S ribosomal RNA gene, partial sequence; mitochondrial.\n",
      "M13046.1\n",
      "X.laevis mitochondrial origin of replication region containing the D-loop, Phe-tRNA, Pro-tRNA, 12S rRNA genes.\n"
     ]
    }
   ],
   "source": [
    "for r in pj.records:\n",
    "    if len(r.seq) < 400:\n",
    "        print r.id\n",
    "        print r.description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add short sequences to reference alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b1 (May 2013); http://hmmer.org/\n",
      "# Copyright (C) 2013 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             400bp_reptile_align_SATIVA.fasta\n",
      "# output HMM file:                  12S_reptile_ref_SATIVA.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen     W eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- ----- -------- ------ -----------\n",
      "1     400bp_reptile_align_SATIVA    62   757   757   934     2.52  0.450 \n",
      "\n",
      "# CPU time: 0.29u 0.00s 00:00:00.28 Elapsed: 00:00:00.30\n"
     ]
    }
   ],
   "source": [
    "!hmmbuild 12S_reptile_ref_SATIVA.hmm 400bp_reptile_align_SATIVA.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now use hmmalign to align the short 12S rRNA sequences to the longer reference alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "hmmalign -o 12S_reptile400bpref_plus_short_query_SATIVA.sto \\\n",
    "--mapali 400bp_reptile_align_SATIVA.fasta \\\n",
    "12S_reptile_ref_SATIVA.hmm reptile_short_to_align_SATIVA.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The hmmalign command has produced a file containing a new alignment of all 12S rRNA sequences (long and short) in stockholm output format. This must be converted to a more useful file format such as fasta or GenBank using the Seq command in BioPython."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 95 records\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "input_file = open(\"12S_reptile400bpref_plus_short_query_SATIVA.sto\", \"rU\")\n",
    "output_file = open(\"12S_ReptileAlign_ALL_SATIVA.fasta\", \"w\")\n",
    "\n",
    "alignment = AlignIO.read(input_file, \"stockholm\")\n",
    "count = len(alignment)\n",
    "AlignIO.write(alignment, output_file, \"fasta\")\n",
    "\n",
    "output_file.close()\n",
    "input_file.close()\n",
    "print(\"Converted %i records\" % count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tree Building"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The fasta file containing the 12S rRNA amphibian sequence alignment must now be converted into a format which can be used for alignment trimming. The alignment must be parsed to an alignment object, aln."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj=unpickle_pj('rep_post_SATIVA.pkl', git=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EU543221.1_f0 is 954bp\n",
      "KX694573.1_f0 is 954bp\n",
      "AY122682.1_f0 is 954bp\n",
      "AY122780.1_f0 is 954bp\n",
      "AY122836.1_f0 is 954bp\n",
      "FJ460597.1_f0 is 954bp\n",
      "AJ001471.1_f0 is 954bp\n",
      "AJ001469.1_f0 is 954bp\n",
      "AJ001472.1_f0 is 954bp\n",
      "AJ001470.1_f0 is 954bp\n",
      "AJ001468.1_f0 is 954bp\n",
      "GQ142072.1_f0 is 954bp\n",
      "KM401599.1_f0 is 954bp\n",
      "KF781329.1_f0 is 954bp\n",
      "AF149947.1_f0 is 954bp\n",
      "KC990830.1_f0 is 954bp\n",
      "AM176577.1_f0 is 954bp\n",
      "AJ001480.1_f0 is 954bp\n",
      "AF149959.1_f0 is 954bp\n",
      "AF149961.1_f0 is 954bp\n",
      "AF149958.1_f0 is 954bp\n",
      "AF149962.1_f0 is 954bp\n",
      "AF149960.1_f0 is 954bp\n",
      "EU443256.1_f0 is 954bp\n",
      "M10217.1_f0 is 954bp\n",
      "Y10943.1_f0 is 954bp\n",
      "AY581639.1_f0 is 954bp\n",
      "M27605.1_f0 is 954bp\n",
      "AY581648.1_f0 is 954bp\n",
      "AY581640.1_f0 is 954bp\n",
      "AY581644.1_f0 is 954bp\n",
      "AY581638.1_f0 is 954bp\n",
      "JN999703.2_f0 is 954bp\n",
      "KF874616.1_f0 is 954bp\n",
      "AF069423.1_f0 is 954bp\n",
      "GU477770.1_f0 is 954bp\n",
      "L28077.1_f0 is 954bp\n",
      "KM216749.1_f0 is 954bp\n",
      "KM216748.1_f0 is 954bp\n",
      "FJ392294.1_f0 is 954bp\n",
      "KC692465.1_f0 is 954bp\n",
      "EF122793.1_f0 is 954bp\n",
      "JN860636.1_f0 is 954bp\n",
      "GU477775.1_f0 is 954bp\n",
      "JX454969.1_f0 is 954bp\n",
      "JX454972.1_f0 is 954bp\n",
      "AB012104.1_f0 is 954bp\n",
      "JX454990.1_f0 is 954bp\n",
      "AY770536.1_f0 is 954bp\n",
      "JX454976.1_f0 is 954bp\n",
      "JX454971.1_f0 is 954bp\n",
      "DQ533485.1_f0 is 954bp\n",
      "JX454986.1_f0 is 954bp\n",
      "JX454980.1_f0 is 954bp\n",
      "JX454970.1_f0 is 954bp\n",
      "KP221806.1_f0 is 954bp\n",
      "JX454981.1_f0 is 954bp\n",
      "KP256531.1_f0 is 954bp\n",
      "JX454984.1_f0 is 954bp\n",
      "JX454983.1_f0 is 954bp\n",
      "AY770537.1_f0 is 954bp\n",
      "JX454977.1_f0 is 954bp\n",
      "AY151988.1_f0 is 954bp\n",
      "KX080575.1_f0 is 954bp\n",
      "HG802830.1_f0 is 954bp\n",
      "AY151993.1_f0 is 954bp\n",
      "HG802772.1_f0 is 954bp\n",
      "DQ081074.1_f0 is 954bp\n",
      "U81335.1_f0 is 954bp\n",
      "KT030712.1_f0 is 954bp\n",
      "EF027342.1_f0 is 954bp\n",
      "AY151990.1_f0 is 954bp\n",
      "FJ039991.1_f0 is 954bp\n",
      "AB090021.1_f0 is 954bp\n",
      "HG802786.1_f0 is 954bp\n",
      "AY151992.1_f0 is 954bp\n",
      "AF080277.1_f0 is 954bp\n",
      "DQ081073.1_f0 is 954bp\n",
      "AF080334.1_f0 is 954bp\n",
      "HG802805.1_f0 is 954bp\n",
      "U81329.1_f0 is 954bp\n",
      "HQ681907.1_f0 is 954bp\n",
      "AF080298.1_f0 is 954bp\n",
      "KX694589.1_f0 is 954bp\n",
      "HQ681908.1_f0 is 954bp\n",
      "U81328.1_f0 is 954bp\n",
      "U81340.1_f0 is 954bp\n",
      "EU624267.1_f0 is 954bp\n",
      "KT030713.1_f0 is 954bp\n",
      "AB090043.1_f0 is 954bp\n",
      "KX080601.1_f0 is 954bp\n",
      "AF158461.1_f0 is 954bp\n",
      "KX080576.1_f0 is 954bp\n",
      "JQ287761.1_f0 is 954bp\n",
      "M13046.1_f0 is 954bp\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "aln = AlignIO.read(open(\"12S_reptile400bpref_plus_short_query_SATIVA.sto\"), \"stockholm\", alphabet=generic_dna)\n",
    "for r in aln:\n",
    "    print(r.id + \" is \" + str(len(r)) + \"bp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.alignments['rrnS@reptile_align_allseq_SATIVA'] = aln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@muscleDefault-400bp-SATIVA': <<class 'Bio.Align.MultipleSeqAlignment'> instance (62 records of length 1043, IUPACProtein()) at 7f1f5e9455d0>,\n",
       " 'rrnS@reptile_align_allseq_SATIVA': <<class 'Bio.Align.MultipleSeqAlignment'> instance (95 records of length 954, DNAAlphabet()) at 7f1f5e5d5d10>}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@muscleDefault-400bp-SATIVA', 'rrnS@reptile_align_allseq_SATIVA']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.alignments.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim the alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimal -in 237211477498428.39_rrnS@reptile_align_allseq_SATIVA.fasta -gappyout\n"
     ]
    }
   ],
   "source": [
    "gappyout = TrimalConf(pj,                      # The Project\n",
    "\n",
    "                      method_name='gappyout',  # Any unique string \n",
    "                                               # ('gappyout' is default)\n",
    "\n",
    "                      program_name='trimal',   # No alternatives in \n",
    "                                               # this ReproPhylo version\n",
    "\n",
    "                      cmd='default',           # the default is trimal. \n",
    "                                               # Change it here\n",
    "                                               # or in pj.defaults['trimal']\n",
    "\n",
    "                      alns=['rrnS@reptile_align_allseq_SATIVA'],# 'all' by default\n",
    "\n",
    "                      trimal_commands={'gappyout': True} # By default, the gappyout algorithm is used.\n",
    "                      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.trim([gappyout])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimalConf named gappyout with ID 237211477498428.39\n",
      "Alignments: rrnS@reptile_align_allseq_SATIVA \n",
      "Created on: Wed Oct 26 16:13:48 2016\n",
      "Commands:\n",
      "rrnS@reptile_align_allseq_SATIVA@gappyout: trimal -in 237211477498428.39_rrnS@reptile_align_allseq_SATIVA.fasta -gappyout\n",
      "\n",
      "Environment:Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: trimAl 1.2rev59\n",
      " Program reference: Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics 2009 25: 1972-1973.\n",
      "execution time:\n",
      "0.650861024857\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The alignment(s) rrnS@reptile_align_allseq_SATIVA were trimmed using the program trimAl 1.2rev59 [1].\n",
      "\n",
      "Reference:\n",
      "Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics 2009 25: 1972-1973.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['gappyout']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rrnS@muscleDefault-400bp-SATIVA@gappyout': <<class 'Bio.Align.MultipleSeqAlignment'> instance (62 records of length 757, IUPACAmbiguousDNA()) at 7f1f5ef47510>,\n",
       " 'rrnS@reptile_align_allseq_SATIVA@gappyout': <<class 'Bio.Align.MultipleSeqAlignment'> instance (95 records of length 757, IUPACAmbiguousDNA()) at 7f1f5e8fd390>}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.trimmed_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@reptile_align_allseq_SATIVA@gappyout_trimmed_aln.fasta',\n",
       " 'rrnS@muscleDefault-400bp-SATIVA@gappyout_trimmed_aln.fasta']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = pj.write_trimmed_alns(id=['record_id','source_organism'],\n",
    "                                   format='fasta')\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer Maximum Likelihood tree based on the clipped alignment. We are going to use RAxML.\n",
    "\n",
    "Configure RAxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 637231477498452.88_rrnS@reptile_align_allseq_SATIVA@gappyout0 -p 722 -s 637231477498452.88_rrnS@reptile_align_allseq_SATIVA@gappyout.fasta -T 5 -x 911 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-muscleDefault-trimmed',\n",
    "                  program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                  keepfiles=True,\n",
    "                  cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                  preset='fa',\n",
    "                  alns=['rrnS@reptile_align_allseq_SATIVA@gappyout'],\n",
    "                  model='GAMMA',\n",
    "                  threads=5,\n",
    "                  cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named raxml-muscleDefault-trimmed with ID 637231477498452.88\n",
      "Alignments: rrnS@reptile_align_allseq_SATIVA@gappyout \n",
      "Created on: Wed Oct 26 16:14:12 2016\n",
      "Commands:\n",
      "rrnS@reptile_align_allseq_SATIVA@gappyout:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 637231477498452.88_rrnS@reptile_align_allseq_SATIVA@gappyout0 -p 722 -s 637231477498452.88_rrnS@reptile_align_allseq_SATIVA@gappyout.fasta -T 5 -x 911 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-38-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: biopc4\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "294.886958838\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) rrnS@reptile_align_allseq_SATIVA@gappyout using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-muscleDefault-trimmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rrnS@reptile_align_allseq_SATIVA@gappyout@raxml-muscleDefault-trimmed',\n",
       " 'rrnS@muscleDefault-400bp-SATIVA@gappyout@raxml-muscleDefault-400bp-SATIVA-trimmed']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.trees.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualise tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize the tree containing all 12S rRNA sequences with outliers identified by SATIVA removed.\n",
    "\n",
    "- Prepare tree annotation:\n",
    "- define outgroups\n",
    "- specify support coloration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Cottus','Mus','Xenopus']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='muscleDefault-trimmed-reptile-all-SATIVA-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a static copy of the tree (the last *.png with the expected alignment id) for later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *rrnS@reptile_align_allseq_SATIVA@gappyout.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png rrnS_muscleDefault-trimmed-reptile-all-SATIVA-raxml.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rep_post_SATIVA.pkl'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'rep_post_SATIVA.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out alignment for pplacer input. Need to modify record id - remove the '_f0' that is appended by Reprophylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning trimmed alignment object rrnS@reptile_align_allseq_SATIVA@gappyout\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fta('rrnS@reptile_align_allseq_SATIVA@gappyout')\n",
    "\n",
    "for r in aln:\n",
    "    r.id = \"_\".join(r.id.split(\"_\")[:-1])\n",
    "    r.description = r.id\n",
    "    \n",
    "out=open('../12S_UKreptiles_SATIVA_cleaned.alignment.fasta', 'w')\n",
    "AlignIO.write(aln, out, 'fasta')\n",
    "out.close()\n",
    "\n",
    "out=open('../12S_UKreptiles_SATIVA_cleaned.alignment.phylip', 'w')\n",
    "AlignIO.write(aln, out, 'phylip-relaxed')\n",
    "out.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
