{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare input files.\n",
    "\n",
    "The ailgnment has been produced using Reprophylo (see notebook). We'll just cleanup the sequence headers and create a local copy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!cat ../fetch_clean_align_tree/trimmed_reptile_align_ALL.phy | sed 's/_f[0-9] / /' > alignment.phy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tax file has been generated from a Genbank file using this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GQ142072.1\tEukaryota;Chordata;unknown;Squamata;Lacertidae;Zootoca;Zootoca vivipara\r\n",
      "KM401599.1\tEukaryota;Chordata;unknown;Squamata;Lacertidae;Zootoca;Zootoca vivipara\r\n",
      "AY151993.1\tEukaryota;Chordata;unknown;Squamata;Lacertidae;Zootoca;Zootoca vivipara\r\n",
      "AF080334.1\tEukaryota;Chordata;unknown;Squamata;Lacertidae;Zootoca;Zootoca vivipara\r\n",
      "KF781329.1\tEukaryota;Chordata;unknown;Squamata;Lacertidae;Zootoca;Zootoca vivipara\r\n",
      "AY122780.1\tEukaryota;Chordata;unknown;Squamata;Colubridae;Zamenis;Zamenis longissimus\r\n",
      "AB188189.1\tEukaryota;Chordata;Actinopteri;Perciformes;Cottidae;Cottus;Cottus gobio\r\n",
      "HG802830.1\tEukaryota;Chordata;unknown;Testudines;Emydidae;Emys;Emys orbicularis\r\n",
      "AB090021.1\tEukaryota;Chordata;unknown;Testudines;Emydidae;Emys;Emys orbicularis\r\n",
      "HG802786.1\tEukaryota;Chordata;unknown;Testudines;Emydidae;Emys;Emys orbicularis\r\n"
     ]
    }
   ],
   "source": [
    "!head tax_for_SATIVA.tax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now run SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SATIVA v0.9-55-g0cbb090, released on 2016-06-28. Last version: https://github.com/amkozlov/sativa \n",
      "By A.Kozlov and J.Zhang, the Exelixis Lab. Based on RAxML 8.2.3 by A.Stamatakis.\n",
      "\n",
      "SATIVA was called as follows:\n",
      "\n",
      "./sativa/sativa.py -s alignment.phy -t tax_for_SATIVA.tax -x zoo -n 12S -o ./ -T 5 -v\n",
      "\n",
      "Mislabels search is running with the following parameters:\n",
      " Alignment:                        alignment.phy\n",
      " Taxonomy:                         tax_for_SATIVA.tax\n",
      " Output directory:                 /home/working/Reference_Alignment/Reptiles/SATIVA\n",
      " Job name / output files prefix:   12S\n",
      " Model of rate heterogeneity:      AUTO\n",
      " Confidence cut-off:               0.000000\n",
      " Number of threads:                5\n",
      "\n",
      "*** STEP 1: Building the reference tree using provided alignment and taxonomic annotations ***\n",
      "\n",
      "=> Loading taxonomy from file: tax_for_SATIVA.tax ...\n",
      "\n",
      "==> Loading reference alignment from file: alignment.phy ...\n",
      "\n",
      "Guessing input format: not fasta\n",
      "Guessing input format: not phylip_relaxed\n",
      "===> Validating taxonomy and alignment ...\n",
      "\n",
      "NOTE: Following sequences are identical: KF781654.1, AY057791.1\n",
      "NOTE: Following sequences are identical: DQ874614.2, KF781653.1\n",
      "\n",
      "NOTE: Found 2 sequence duplicates\n",
      "====> Building a multifurcating tree from taxonomy with 113 seqs ...\n",
      "\n",
      "Total nodes in resulting tree: 113\n",
      "=====> Building the reference alignment ...\n",
      "\n",
      "======> Saving the outgroup for later re-rooting ...\n",
      "\n",
      "Outgroup for rooting was saved to: /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S_outgr.tre, outgroup size: 1\n",
      "=======> Resolving multifurcation: choosing the best topology from 1 independent RAxML runs ...\n",
      "\n",
      "\n",
      "Reducing the alignment: \n",
      "\n",
      "/home/working/Reference_Alignment/Reptiles/SATIVA/sativa/raxml//run_raxml.sh -w /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302 -T 5 -f c -s /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S_matrix.afa --no-dup-check -m GTRGAMMA -n reduce --verbose -p 13332\n",
      "\n",
      "\n",
      "Constrained ML inference: \n",
      "\n",
      "/home/working/Reference_Alignment/Reptiles/SATIVA/sativa/raxml//run_raxml.sh -w /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302 -T 5 -s /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S_matrix.afa -g /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S_mfu.tre --no-seq-check -N 1 -m GTRGAMMA -n mfresolv_12S --verbose -p 326\n",
      "\n",
      "\n",
      "GAMMA-based logLH of the reference tree: -11697.690917\n",
      "\n",
      "========> Calling RAxML-EPA to obtain branch labels ...\n",
      "\n",
      "/home/working/Reference_Alignment/Reptiles/SATIVA/sativa/raxml//run_raxml.sh -w /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302 -T 5 -s /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S_lblq.fa -t /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S_bfu.tre --no-seq-check -f y --epa-accumulated-threshold 0.999 -R /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S.opt -m GTRGAMMA -n epalbl_12S --verbose -p 26407\n",
      "\n",
      "=========> Post-processing the EPA tree (re-rooting, taxonomic labeling etc.) ...\n",
      "\n",
      "\n",
      "==========> Checking branch labels ...\n",
      "shared rank names before training: set(['Eukaryota', 'Chordata'])\n",
      "shared rank names after  training: set(['Eukaryota', 'Chordata'])\n",
      "\n",
      "==========> Saving the reference JSON file: ./12S.refjson\n",
      "\n",
      "Calculating the speciation rate...\n",
      "\n",
      "Writing down the reference file...\n",
      "\n",
      "*** STEP 2: Searching for mislabels ***\n",
      "\n",
      "Running the leave-one-sequence-out test...\n",
      "\n",
      "/home/working/Reference_Alignment/Reptiles/SATIVA/sativa/raxml//run_raxml.sh -w /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302 -T 5 -s /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S.refaln -t /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/ref_12S.tre --no-seq-check -f O --epa-accumulated-threshold 0.999 -R /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S.opt -m GTRGAMMA -n l1out_seq_12S --verbose -p 13332\n",
      "\n",
      "Leave-one-out test identified 1 suspicious sequences; running final EPA test to check them...\n",
      "\n",
      "/home/working/Reference_Alignment/Reptiles/SATIVA/sativa/raxml//run_raxml.sh -w /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302 -T 5 -s /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/12S.refaln -t /home/working/Reference_Alignment/Reptiles/SATIVA/sativa/tmp/12S_1477497435302/final_ref_12S.tre --no-seq-check -f v --epa-accumulated-threshold 0.999 -m GTRGAMMA -n final_epa_12S --verbose -p 326\n",
      "\n",
      "HQ442366.1\tFamily\tEmydidae\tGeoemydidae\t0.516\tEukaryota;Chordata;unknown;Testudines;Emydidae;Trachemys;Trachemys scripta\tEukaryota;Chordata;unknown;Testudines;Geoemydidae;Mauremys;Mauremys caspica\t1.000;1.000;1.000;1.000;0.516;0.516;0.516\n",
      "\n",
      "Mislabeled sequences by rank:\n",
      "Family      :\t1\n",
      "\n",
      "Total mislabels: 1 / 0.88 %\n",
      "\n",
      "Results were saved to: /home/working/Reference_Alignment/Reptiles/SATIVA/12S.mis\n",
      "Execution log was saved to: /home/working/Reference_Alignment/Reptiles/SATIVA/12S.log\n",
      "\n",
      "Analysis completed successfully, elapsed time: 25 seconds (12s reftree, 13s leave-one-out)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!./sativa/sativa.py -s alignment.phy -t tax_for_SATIVA.tax -x zoo -n 12S -o ./ -T 5 -v"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SATIVA (among other things) writes a report to a text file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ";WARNING: The revised taxon name suggested here is not necessarily the one that has priority in nomenclature. \r\n",
      ";Our suggestion should only be taken as indicative of an affiliation to the same group, whose correct name must be determined \r\n",
      ";in an additional step according to the specific rules of nomenclature that apply to the studied organisms.\r\n",
      ";\r\n",
      ";SeqID\tMislabeledLevel\tOriginalLabel\tProposedLabel\tConfidence\tOriginalTaxonomyPath\tProposedTaxonomyPath\tPerRankConfidence\r\n",
      "HQ442366.1\tFamily\tEmydidae\tGeoemydidae\t0.516\tEukaryota;Chordata;unknown;Testudines;Emydidae;Trachemys;Trachemys scripta\tEukaryota;Chordata;unknown;Testudines;Geoemydidae;Mauremys;Mauremys caspica\t1.000;1.000;1.000;1.000;0.516;0.516;0.516\r\n"
     ]
    }
   ],
   "source": [
    "!cat 12S.mis"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
